<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.18">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Sam Foreman">
<meta name="author" content="Huihuo Zheng">
<meta name="dcterms.date" content="2025-07-25">
<meta name="description" content="Statistical learning is a framework for understanding and modeling complex data patterns, combining statistical methods with machine learning algorithms to build predictive models.">

<title>Statistical Learning – Intro to HPC Bootcamp 2025</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.min.js" integrity="sha384-ZvpUoO/+PpLXR1lu4jmpXWu80pZlYUAfxl5NsBMWOEPSjUn/6Z/hRTt8+pR6L4N2" crossorigin="anonymous"></script><script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../00-intro-AI-HPC/8-clustering/index.html" rel="next">
<link href="../../00-intro-AI-HPC/6-linear-regression/index.html" rel="prev">
<link href="../../assets/favicon-sf.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-9bf6f933b760cbb4c99ada293d62cb55.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-22c6c940514060718923bc7c0517e30f.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-1a95a006fa96163f90bf11620203e68d.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark-174353d2fa111ba48747fab321614398.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=GTM-MRCG68M5"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'GTM-MRCG68M5', { 'anonymize_ip': true});
</script>
<script src="https://cdn.jsdelivr.net/npm/requirejs@2.3.6/require.min.js" integrity="sha384-6V1/AdqZRWk1KAlWbKBlGhN7VG4iE/yAZcO6NZPMF8od0vukrvr0tg4qY6NSrItx" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
<script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.22/dist/katex.min.css" integrity="sha384-5TcZemv2l/9On385z///+d7MSYlvIEw9FuZTIdZ14vJLqWphw7e7ZPuOiCHJcFCP" crossorigin="anonymous">
<script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@0.16.22/dist/katex.min.js" integrity="sha384-cMkvdD8LoxVzGF/RPUKAcvmm49FQ0oxwDF3BGKtDXcEc+T1b2N+teh/OJfpU0jr6" crossorigin="anonymous"></script>
<script defer="" src="https://cdn.jsdelivr.net/npm/katex@0.16.22/dist/contrib/auto-render.min.js" integrity="sha384-hCXGrW6PitJEwbkoStFjeJxv+fSOOQKOPbJxSfM6G5sWZjAyWhXiTIIAmQqnlLlh" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=IBM+Plex+Sans:ital,wght@0,100;0,200;0,300;0,400;0,500;0,600;0,700;1,100;1,200;1,300;1,400;1,500;1,600;1,700&amp;family=IBM+Plex+Sans+Condensed:ital,wght@0,400;0,500;0,600;0,700&amp;family=IBM+Plex+Mono:ital,wght@0,100;0,200;0,300;0,400;0,500;0,600;0,700&amp;display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css?family=IBM+Plex+Sans&amp;family=IBM+Plex+Sans+Condensed&amp;family=IBM+Plex+Mono&amp;display=swap" rel="stylesheet">
<!-- Google Tag Manager -->
<script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src='https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);})(window,document,'script','dataLayer','GTM-TC329HJ');</script>
<!-- End Google Tag Manager -->
<link rel="preconnect" href="https://fonts.googleapis.com">
<link href="https://iosevka-webfonts.github.io/iosevka/iosevka.css" rel="stylesheet">

  <script defer="" src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN" && texText && texText.data) {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<meta property="og:title" content="Statistical Learning – Intro to HPC Bootcamp 2025">
<meta property="og:description" content="Statistical learning is a framework for understanding and modeling complex data patterns, combining statistical methods with machine learning algorithms to build predictive models.">
<meta property="og:image" content="https://saforem2.github.io/hpc-bootcamp-2025/00-intro-AI-HPC/7-statistical-learning/data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==">
<meta property="og:site_name" content="Intro to HPC Bootcamp 2025">
<meta name="twitter:title" content="Statistical Learning – Intro to HPC Bootcamp 2025">
<meta name="twitter:description" content="Statistical learning is a framework for understanding and modeling complex data patterns, combining statistical methods with machine learning algorithms to build predictive models.">
<meta name="twitter:image" content="https://saforem2.github.io/hpc-bootcamp-2025/00-intro-AI-HPC/7-statistical-learning/data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==">
<meta name="twitter:site" content="saforem2">
<meta name="twitter:card" content="summary_large_image">
<meta name="citation_title" content="Statistical Learning">
<meta name="citation_author" content="Sam Foreman">
<meta name="citation_author" content="Huihuo Zheng">
<meta name="citation_publication_date" content="2025-07-25">
<meta name="citation_cover_date" content="2025-07-25">
<meta name="citation_year" content="2025">
<meta name="citation_online_date" content="2025-07-25">
<meta name="citation_fulltext_html_url" content="https://saforem2.github.io/hpc-bootcamp-2025/00-intro-AI-HPC/7-statistical-learning/">
<meta name="citation_language" content="en">
<meta name="citation_reference" content="citation_title=The hugging face course, 2022;,citation_author=Hugging Face;,citation_publication_date=2022;,citation_cover_date=2022;,citation_year=2022;,citation_publisher=https://huggingface.co/course;">
<meta name="citation_reference" content="citation_title=HiPerRAG: High-performance retrieval augmented generation for scientific insights;,citation_author=Ozan Gokdemir;,citation_author=Carlo Siebenschuh;,citation_author=Alexander Brace;,citation_author=Azton Wells;,citation_author=Brian Hsu;,citation_author=Kyle Hippe;,citation_author=Priyanka V. Setty;,citation_author=Aswathy Ajith;,citation_author=J. Gregory Pauloski;,citation_author=Varuni Sastry;,citation_author=Sam Foreman;,citation_author=Huihuo Zheng;,citation_author=Heng Ma;,citation_author=Bharat Kale;,citation_author=Nicholas Chia;,citation_author=Thomas Gibbs;,citation_author=Michael E. Papka;,citation_author=Thomas Brettin;,citation_author=Francis J. Alexander;,citation_author=Anima Anandkumar;,citation_author=Ian Foster;,citation_author=Rick Stevens;,citation_author=Venkatram Vishwanath;,citation_author=Arvind Ramanathan;,citation_publication_date=2025;,citation_cover_date=2025;,citation_year=2025;,citation_fulltext_html_url=https://arxiv.org/abs/2505.04846;">
<meta name="citation_reference" content="citation_title=MOFA: Discovering materials for carbon capture with a GenAI- and simulation-based workflow;,citation_author=Xiaoli Yan;,citation_author=Nathaniel Hudson;,citation_author=Hyun Park;,citation_author=Daniel Grzenda;,citation_author=J. Gregory Pauloski;,citation_author=Marcus Schwarting;,citation_author=Haochen Pan;,citation_author=Hassan Harb;,citation_author=Samuel Foreman;,citation_author=Chris Knight;,citation_author=Tom Gibbs;,citation_author=Kyle Chard;,citation_author=Santanu Chaudhuri;,citation_author=Emad Tajkhorshid;,citation_author=Ian Foster;,citation_author=Mohamad Moosavi;,citation_author=Logan Ward;,citation_author=E. A. Huerta;,citation_publication_date=2025;,citation_cover_date=2025;,citation_year=2025;,citation_fulltext_html_url=https://arxiv.org/abs/2501.10651;">
<meta name="citation_reference" content="citation_title=MProt-DPO: Breaking the ExaFLOPS barrier for multimodal protein design workflows with direct preference optimization;,citation_abstract=We present a scalable, end-to-end workflow for protein design. By augmenting protein sequences with natural language descriptions of their biochemical properties, we train generative models that can be preferentially aligned with protein fitness landscapes. Through complex experimental- and simulation-based observations, we integrate these measures as preferred parameters for generating new protein variants and demonstrate our workflow on five diverse supercomputers. We achieve &amp;amp;amp;gt;1 ExaFLOPS sustained performance in mixed precision on each supercomputer and a maximum sustained performance of 4.11 ExaFLOPS and peak performance of 5.57 ExaFLOPS. We establish the scientific performance of our model on two tasks: (1) across a predetermined benchmark dataset of deep mutational scanning experiments to optimize the fitness-determining mutations in the yeast protein HIS7, and (2) in optimizing the design of the enzyme malate dehydrogenase to achieve lower activation barriers (and therefore increased catalytic rates) using simulation data. Our implementation thus sets high watermarks for multimodal protein design workflows.;,citation_author=Gautham Dharuman;,citation_author=Kyle Hippe;,citation_author=Alexander Brace;,citation_author=Sam Foreman;,citation_author=Väinö Hatanpää;,citation_author=Varuni K. Sastry;,citation_author=Huihuo Zheng;,citation_author=Logan Ward;,citation_author=Servesh Muralidharan;,citation_author=Archit Vasan;,citation_author=Bharat Kale;,citation_author=Carla M. Mann;,citation_author=Heng Ma;,citation_author=Yun-Hsuan Cheng;,citation_author=Yuliana Zamora;,citation_author=Shengchao Liu;,citation_author=Chaowei Xiao;,citation_author=Murali Emani;,citation_author=Tom Gibbs;,citation_author=Mahidhar Tatineni;,citation_author=Deepak Canchi;,citation_author=Jerome Mitchell;,citation_author=Koichi Yamada;,citation_author=Maria Garzaran;,citation_author=Michael E. Papka;,citation_author=Ian Foster;,citation_author=Rick Stevens;,citation_author=Anima Anandkumar;,citation_author=Venkatram Vishwanath;,citation_author=Arvind Ramanathan;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_fulltext_html_url=https://doi.org/10.1109/SC41406.2024.00013;,citation_doi=10.1109/SC41406.2024.00013;,citation_isbn=9798350352917;,citation_conference_title=Proceedings of the international conference for high performance computing, networking, storage, and analysis;,citation_conference=IEEE Press;,citation_series_title=SC ’24;">
<meta name="citation_reference" content="citation_title=Quality measures for dynamic graph generative models;,citation_author=Ryien Hosseini;,citation_author=Filippo Simini;,citation_author=Venkatram Vishwanath;,citation_author=Rebecca Willett;,citation_author=Henry Hoffmann;,citation_publication_date=2025;,citation_cover_date=2025;,citation_year=2025;,citation_fulltext_html_url=https://openreview.net/forum?id=8bjspmAMBk;,citation_conference_title=The thirteenth international conference on learning representations;">
<meta name="citation_reference" content="citation_title=Superconductivity of in and sn samples;,citation_author=George Deamont;,citation_author=Sam Foreman;,citation_publication_date=2014;,citation_cover_date=2014;,citation_year=2014;">
<meta name="citation_reference" content="citation_title=RG-inspired machine learning for lattice field theory;,citation_author=Sam Foreman;,citation_author=Joel Giedt;,citation_author=Yannick Meurice;,citation_author=Judah Unmuth-Yockey;,citation_publication_date=2018;,citation_cover_date=2018;,citation_year=2018;,citation_volume=175;,citation_conference_title=EPJ web of conferences;,citation_conference=EDP Sciences;">
<meta name="citation_reference" content="citation_title=Large energy density in three-plate nanocapacitors due to coulomb blockade;,citation_author=A Hubler;,citation_author=S Foreman;,citation_author=J Liu;,citation_author=L Wortsmann;,citation_publication_date=2018;,citation_cover_date=2018;,citation_year=2018;,citation_issue=10;,citation_volume=123;,citation_journal_title=Journal of Applied Physics;,citation_publisher=AIP Publishing;">
<meta name="citation_reference" content="citation_title=Examples of renormalization group transformations for image sets;,citation_author=Samuel Foreman;,citation_author=Joel Giedt;,citation_author=Yannick Meurice;,citation_author=Judah Unmuth-Yockey;,citation_publication_date=2018;,citation_cover_date=2018;,citation_year=2018;,citation_issue=5;,citation_volume=98;,citation_journal_title=Physical Review E;,citation_publisher=American Physical Society;">
<meta name="citation_reference" content="citation_title=Machine learning inspired analysis of the Ising model transition;,citation_author=Samuel Foreman;,citation_author=Joel Giedt;,citation_author=Yannick Meurice;,citation_author=Judah Unmuth-Yockey;,citation_publication_date=2018;,citation_cover_date=2018;,citation_year=2018;,citation_doi=10.22323/1.334.0245;,citation_volume=LATTICE2018;,citation_journal_title=PoS;">
<meta name="citation_reference" content="citation_title=Machine learning inspired analysis of the ising model transition;,citation_author=Samuel Foreman;,citation_author=Joel Giedt;,citation_author=Yannick Meurice;,citation_author=Judah Unmuth-Yockey;,citation_publication_date=2018;,citation_cover_date=2018;,citation_year=2018;,citation_conference_title=Lattice 2018;">
<meta name="citation_reference" content="citation_title=Learning better physics: A machine learning approach to lattice gauge theory;,citation_author=Samuel Alfred Foreman;,citation_publication_date=2019;,citation_cover_date=2019;,citation_year=2019;,citation_dissertation_institution=University of Iowa;">
<meta name="citation_reference" content="citation_title=Machine learning and neural networks for field theory;,citation_author=Sam Foreman;,citation_author=Xiao-Yong Jin;,citation_author=James C Osborn;,citation_publication_date=2020;,citation_cover_date=2020;,citation_year=2020;">
<meta name="citation_reference" content="citation_title=HMC with normalizing flows;,citation_author=Sam Foreman;,citation_author=Taku Izubuchi;,citation_author=Luchang Jin;,citation_author=Xiao-Yong Jin;,citation_author=James C Osborn;,citation_author=Akio Tomiya;,citation_publication_date=2021;,citation_cover_date=2021;,citation_year=2021;,citation_fulltext_html_url=https://arxiv.org/abs/2112.01586;,citation_journal_title=arXiv preprint arXiv:2112.01586;">
<meta name="citation_reference" content="citation_title=LeapfrogLayers: A trainable framework for effective topological sampling;,citation_author=Sam Foreman;,citation_author=Xiao-Yong Jin;,citation_author=James C Osborn;,citation_publication_date=2021;,citation_cover_date=2021;,citation_year=2021;,citation_fulltext_html_url=https://arxiv.org/abs/2112.01582;,citation_journal_title=arXiv preprint arXiv:2112.01582;">
<meta name="citation_reference" content="citation_title=Energy storage in quantum resonators;,citation_author=Jiaqi Liu;,citation_author=Alfred W Hubler;,citation_author=Samuel Alfred Foreman;,citation_author=Katharina Ott;,citation_publication_date=2017;,citation_cover_date=2017;,citation_year=2017;">
<meta name="citation_reference" content="citation_title=Applications of machine learning to lattice quantum field theory;,citation_author=Denis Boyda;,citation_author=Salvatore Calı̀;,citation_author=Sam Foreman;,citation_author=Lena Funcke;,citation_author=Daniel C Hackett;,citation_author=Yin Lin;,citation_author=Gert Aarts;,citation_author=Andrei Alexandru;,citation_author=Xiao-Yong Jin;,citation_author=Biagio Lucini;,citation_author=others;,citation_publication_date=2022;,citation_cover_date=2022;,citation_year=2022;,citation_fulltext_html_url=https://arxiv.org/abs/2202.05838;,citation_journal_title=arXiv preprint arXiv:2202.05838;">
<meta name="citation_reference" content="citation_title=Lattice QCD and particle physics;,citation_author=Andreas S Kronfeld;,citation_author=Tanmoy Bhattacharya;,citation_author=Thomas Blum;,citation_author=Norman H Christ;,citation_author=Carleton DeTar;,citation_author=William Detmold;,citation_author=Robert Edwards;,citation_author=Anna Hasenfratz;,citation_author=Huey-Wen Lin;,citation_author=Swagato Mukherjee;,citation_author=others;,citation_publication_date=2022;,citation_cover_date=2022;,citation_year=2022;,citation_fulltext_html_url=https://arxiv.org/abs/2207.07641;,citation_journal_title=arXiv preprint arXiv:2207.07641;">
<meta name="citation_reference" content="citation_title=GenSLMs: Genome-scale language models reveal SARS-CoV-2 evolutionary dynamics;,citation_author=Maxim Zvyagin;,citation_author=Alexander Brace;,citation_author=Kyle Hippe;,citation_author=Yuntian Deng;,citation_author=Bin Zhang;,citation_author=Cindy Orozco Bohorquez;,citation_author=Austin Clyde;,citation_author=Bharat Kale;,citation_author=Danilo Perez-Rivera;,citation_author=Heng Ma;,citation_author=others;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_issue=6;,citation_volume=37;,citation_journal_title=The International Journal of High Performance Computing Applications;,citation_publisher=SAGE Publications Sage UK: London, England;">
<meta name="citation_reference" content="citation_title=MLMC: Machine learning monte carlo;,citation_author=Sam Foreman;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_conference_title=The international symposium on lattice field theory;">
<meta name="citation_reference" content="citation_title=Superconductivity of in and sn samples;,citation_author=George Deamont;,citation_author=Sam Foreman;,citation_publication_date=2014;,citation_cover_date=2014;,citation_year=2014;">
<meta name="citation_reference" content="citation_title=A comprehensive performance study of large language models on novel AI accelerators;,citation_author=Murali Emani;,citation_author=Sam Foreman;,citation_author=Varuni Sastry;,citation_author=Zhen Xie;,citation_author=Siddhisanket Raskar;,citation_author=William Arnold;,citation_author=Rajeev Thakur;,citation_author=Venkatram Vishwanath;,citation_author=Michael E Papka;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=https://arxiv.org/abs/2310.04607;,citation_journal_title=arXiv preprint arXiv:2310.04607;">
<meta name="citation_reference" content="citation_title=DeepSpeed4Science initiative: Enabling large-scale scientific discovery through sophisticated AI system technologies;,citation_author=Shuaiwen Leon Song;,citation_author=Bonnie Kruft;,citation_author=Minjia Zhang;,citation_author=Conglong Li;,citation_author=Shiyang Chen;,citation_author=Chengming Zhang;,citation_author=Masahiro Tanaka;,citation_author=Xiaoxia Wu;,citation_author=Jeff Rasley;,citation_author=Ammar Ahmad Awan;,citation_author=others;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=https://arxiv.org/abs/2310.04610;,citation_journal_title=arXiv preprint arXiv:2310.04610;">
<meta name="citation_reference" content="citation_title=Protein generation via genome-scale language models with bio-physical scoring;,citation_author=Gautham Dharuman;,citation_author=Logan Ward;,citation_author=Heng Ma;,citation_author=Priyanka V Setty;,citation_author=Ozan Gokdemir;,citation_author=Sam Foreman;,citation_author=Murali Emani;,citation_author=Kyle Hippe;,citation_author=Alexander Brace;,citation_author=Kristopher Keipert;,citation_author=others;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_conference_title=Proceedings of the SC’23 workshops of the international conference on high performance computing, network, storage, and analysis;">
<meta name="citation_reference" content="citation_title=MLMC: Machine learning monte carlo for lattice gauge theory;,citation_author=Sam Foreman;,citation_author=Xiao-Yong Jin;,citation_author=James C Osborn;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=https://arxiv.org/abs/2312.08936;,citation_journal_title=arXiv preprint arXiv:2312.08936;">
<meta name="citation_reference" content="citation_title=Snowmass 2021 computational frontier CompF03 topical group report: Machine learning;,citation_author=Phiala Shanahan;,citation_author=Kazuhiro Terao;,citation_author=Daniel Whiteson;,citation_publication_date=2022;,citation_cover_date=2022;,citation_year=2022;,citation_fulltext_html_url=https://arxiv.org/abs/2209.07559;,citation_journal_title=arXiv preprint arXiv:2209.07559;">
<meta name="citation_reference" content="citation_title=Thorough characterization and analysis of large transformer model training at-scale;,citation_author=Scott Cheng;,citation_author=Jun-Liang Lin;,citation_author=Murali Emani;,citation_author=Siddhisanket Raskar;,citation_author=Sam Foreman;,citation_author=Zhen Xie;,citation_author=Venkatram Vishwanath;,citation_author=Mahmut Taylan Kandemir;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_issue=1;,citation_volume=8;,citation_journal_title=Proceedings of the ACM on Measurement and Analysis of Computing Systems;,citation_publisher=ACM New York, NY, USA;">
<meta name="citation_reference" content="citation_title=Communities through energy justice projects;,citation_author=Mary Ann Leung;,citation_author=Katharine Cahill;,citation_author=Rebecca Hartman-Baker;,citation_author=Paige Kinsley;,citation_author=Lois Curfman McInnes;,citation_author=Suzanne Parete-Koon;,citation_author=Subil Abraham;,citation_author=Lacy Beach Barrier;,citation_author=Gladys Chen;,citation_author=Lizanne DeStefano;,citation_author=others;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_issue=1;,citation_volume=15;,citation_journal_title=Journal of Computational Science;">
<meta name="citation_reference" content="citation_title=Applications of a foundation model approach for weather and climate;,citation_author=Troy Arcomano;,citation_author=Alexander Wikner;,citation_author=Romit Maulik;,citation_author=Veerabhadra Rao Kotamarthi;,citation_author=Sam Foreman;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_volume=2023;,citation_conference_title=AGU fall meeting abstracts;">
<meta name="citation_reference" content="citation_title=Toward a holistic performance evaluation of large language models across diverse ai accelerators;,citation_author=Murali Emani;,citation_author=Sam Foreman;,citation_author=Varuni Sastry;,citation_author=Zhen Xie;,citation_author=Siddhisanket Raskar;,citation_author=William Arnold;,citation_author=Rajeev Thakur;,citation_author=Venkatram Vishwanath;,citation_author=Michael E Papka;,citation_author=Sanjif Shanmugavelu;,citation_author=others;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_conference_title=2024 IEEE international parallel and distributed processing symposium workshops (IPDPSW);,citation_conference=IEEE;">
<meta name="citation_reference" content="citation_title=Intro to HPC bootcamp: Engaging new communities through energy justice projects;,citation_author=Suzanne Parete-Koon;,citation_author=Michael Sandoval;,citation_author=Kellen Leland;,citation_author=Subil Abraham;,citation_author=Mary Ann Leung;,citation_author=Rebecca Hartman-Baker;,citation_author=Paige Kinsley;,citation_author=Lois McInnes;,citation_author=Sreeranjani Ramprakash;,citation_author=Lacy Beach Barrier;,citation_author=others;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_issue=1;,citation_volume=15;,citation_journal_title=Journal of Computational Science Education;,citation_publisher=Oak Ridge National Laboratory (ORNL), Oak Ridge, TN (United States);">
<meta name="citation_reference" content="citation_title=MProt-DPO: Breaking the ExaFLOPS barrier for multimodal protein design workflows with direct preference optimization;,citation_author=Gautham Dharuman;,citation_author=Kyle Hippe;,citation_author=Alexander Brace;,citation_author=Sam Foreman;,citation_author=Väinä Hatanpää;,citation_author=Varuni K Sastry;,citation_author=Huihuo Zheng;,citation_author=Logan Ward;,citation_author=Servesh Muralidharan;,citation_author=Archit Vasan;,citation_author=others;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_conference_title=2024 SC24: International conference for high performance computing, networking, storage and analysis SC;,citation_conference=IEEE Computer Society;">
<meta name="citation_reference" content="citation_title=Emergent abilities of large language models;,citation_author=Jason Wei;,citation_author=Yi Tay;,citation_author=Rishi Bommasani;,citation_author=Colin Raffel;,citation_author=Barret Zoph;,citation_author=Sebastian Borgeaud;,citation_author=Dani Yogatama;,citation_author=Maarten Bosma;,citation_author=Denny Zhou;,citation_author=Donald Metzler;,citation_author=Ed H. Chi;,citation_author=Tatsunori Hashimoto;,citation_author=Oriol Vinyals;,citation_author=Percy Liang;,citation_author=Jeff Dean;,citation_author=William Fedus;,citation_publication_date=2022;,citation_cover_date=2022;,citation_year=2022;,citation_fulltext_html_url=https://arxiv.org/abs/2206.07682;">
<meta name="citation_reference" content="citation_title=DeepSpeed4Science initiative: Enabling large-scale scientific discovery through sophisticated AI system technologies;,citation_author=Shuaiwen Leon Song;,citation_author=Bonnie Kruft;,citation_author=Minjia Zhang;,citation_author=Conglong Li;,citation_author=Shiyang Chen;,citation_author=Chengming Zhang;,citation_author=Masahiro Tanaka;,citation_author=Xiaoxia Wu;,citation_author=Jeff Rasley;,citation_author=Ammar Ahmad Awan;,citation_author=Connor Holmes;,citation_author=Martin Cai;,citation_author=Adam Ghanem;,citation_author=Zhongzhu Zhou;,citation_author=Yuxiong He;,citation_author=Pete Luferenko;,citation_author=Divya Kumar;,citation_author=Jonathan Weyn;,citation_author=Ruixiong Zhang;,citation_author=Sylwester Klocek;,citation_author=Volodymyr Vragov;,citation_author=Mohammed AlQuraishi;,citation_author=Gustaf Ahdritz;,citation_author=Christina Floristean;,citation_author=Cristina Negri;,citation_author=Rao Kotamarthi;,citation_author=Venkatram Vishwanath;,citation_author=Arvind Ramanathan;,citation_author=Sam Foreman;,citation_author=Kyle Hippe;,citation_author=Troy Arcomano;,citation_author=Romit Maulik;,citation_author=Maxim Zvyagin;,citation_author=Alexander Brace;,citation_author=Bin Zhang;,citation_author=Cindy Orozco Bohorquez;,citation_author=Austin Clyde;,citation_author=Bharat Kale;,citation_author=Danilo Perez-Rivera;,citation_author=Heng Ma;,citation_author=Carla M. Mann;,citation_author=Michael Irvin;,citation_author=J. Gregory Pauloski;,citation_author=Logan Ward;,citation_author=Valerie Hayot;,citation_author=Murali Emani;,citation_author=Zhen Xie;,citation_author=Diangen Lin;,citation_author=Maulik Shukla;,citation_author=Ian Foster;,citation_author=James J. Davis;,citation_author=Michael E. Papka;,citation_author=Thomas Brettin;,citation_author=Prasanna Balaprakash;,citation_author=Gina Tourassi;,citation_author=John Gounley;,citation_author=Heidi Hanson;,citation_author=Thomas E Potok;,citation_author=Massimiliano Lupo Pasini;,citation_author=Kate Evans;,citation_author=Dan Lu;,citation_author=Dalton Lunga;,citation_author=Junqi Yin;,citation_author=Sajal Dash;,citation_author=Feiyi Wang;,citation_author=Mallikarjun Shankar;,citation_author=Isaac Lyngaas;,citation_author=Xiao Wang;,citation_author=Guojing Cong;,citation_author=Pei Zhang;,citation_author=Ming Fan;,citation_author=Siyan Liu;,citation_author=Adolfy Hoisie;,citation_author=Shinjae Yoo;,citation_author=Yihui Ren;,citation_author=William Tang;,citation_author=Kyle Felker;,citation_author=Alexey Svyatkovskiy;,citation_author=Hang Liu;,citation_author=Ashwin Aji;,citation_author=Angela Dalton;,citation_author=Michael Schulte;,citation_author=Karl Schulz;,citation_author=Yuntian Deng;,citation_author=Weili Nie;,citation_author=Josh Romero;,citation_author=Christian Dallago;,citation_author=Arash Vahdat;,citation_author=Chaowei Xiao;,citation_author=Thomas Gibbs;,citation_author=Anima Anandkumar;,citation_author=Rick Stevens;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=https://arxiv.org/abs/2310.04610;">
<meta name="citation_reference" content="citation_title=Emergent abilities of large language models;,citation_author=Jason Wei;,citation_author=Yi Tay;,citation_author=Rishi Bommasani;,citation_author=Colin Raffel;,citation_author=Barret Zoph;,citation_author=Sebastian Borgeaud;,citation_author=Dani Yogatama;,citation_author=Maarten Bosma;,citation_author=Denny Zhou;,citation_author=Donald Metzler;,citation_author=Ed H. Chi;,citation_author=Tatsunori Hashimoto;,citation_author=Oriol Vinyals;,citation_author=Percy Liang;,citation_author=Jeff Dean;,citation_author=William Fedus;,citation_publication_date=2022;,citation_cover_date=2022;,citation_year=2022;,citation_fulltext_html_url=https://arxiv.org/abs/2206.07682;">
<meta name="citation_reference" content="citation_title=The climate risk &amp;amp;amp; resilience portal (ClimRR) metadata and data dictionary;,citation_author=C. Burdi;,citation_author=Wall. T Branham;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=https://dub.sh/ClimRR-Metadata;">
<meta name="citation_reference" content="citation_title=Progress on $(g-2)_\mu$ from lattice QCD;,citation_author=Hartmut Wittig;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=https://arxiv.org/abs/2306.04165;">
<meta name="citation_reference" content="citation_title=Hybrid Monte Carlo;,citation_author=S. Duane;,citation_author=A. D. Kennedy;,citation_author=B. J. Pendleton;,citation_author=D. Roweth;,citation_publication_date=1987;,citation_cover_date=1987;,citation_year=1987;,citation_doi=10.1016/0370-2693(87)91197-X;,citation_volume=195;,citation_journal_title=Phys. Lett. B;">
<meta name="citation_reference" content="citation_title=Snowmass 2021 Computational Frontier CompF03 Topical Group Report: Machine Learning;,citation_author=Phiala Shanahan;,citation_author=others;,citation_publication_date=2022-09;,citation_cover_date=2022-09;,citation_year=2022;,citation_fulltext_html_url=https://arxiv.org/abs/2209.07559;">
<meta name="citation_reference" content="citation_title=Applications of Machine Learning to Lattice Quantum Field Theory;,citation_author=Denis Boyda;,citation_author=others;,citation_publication_date=2022-02;,citation_cover_date=2022-02;,citation_year=2022;,citation_fulltext_html_url=https://arxiv.org/abs/2202.05838;,citation_conference_title=Snowmass 2021;">
<meta name="citation_reference" content="citation_title=HMC with Normalizing Flows;,citation_author=Sam Foreman;,citation_author=Taku Izubuchi;,citation_author=Luchang Jin;,citation_author=Xiao-Yong Jin;,citation_author=James C. Osborn;,citation_author=Akio Tomiya;,citation_publication_date=2022;,citation_cover_date=2022;,citation_year=2022;,citation_fulltext_html_url=https://arxiv.org/abs/2112.01586;,citation_doi=10.22323/1.396.0073;,citation_volume=LATTICE2021;,citation_journal_title=PoS;">
<meta name="citation_reference" content="citation_title=LeapfrogLayers: A Trainable Framework for Effective Topological Sampling;,citation_author=Sam Foreman;,citation_author=Xiao-Yong Jin;,citation_author=James C. Osborn;,citation_publication_date=2022;,citation_cover_date=2022;,citation_year=2022;,citation_fulltext_html_url=https://arxiv.org/abs/2112.01582;,citation_doi=10.22323/1.396.0508;,citation_volume=LATTICE2021;,citation_journal_title=PoS;">
<meta name="citation_reference" content="citation_title=Deep Learning Hamiltonian Monte Carlo;,citation_author=Sam Foreman;,citation_author=Xiao-Yong Jin;,citation_author=James C. Osborn;,citation_publication_date=2021-05;,citation_cover_date=2021-05;,citation_year=2021;,citation_fulltext_html_url=https://arxiv.org/abs/2105.03418;,citation_conference_title=9th International Conference on Learning Representations;">
<meta name="citation_reference" content="citation_title=Deep learning hamiltonian monte carlo;,citation_author=Sam Foreman;,citation_author=Xiao-Yong Jin;,citation_author=James C. Osborn;,citation_publication_date=2021;,citation_cover_date=2021;,citation_year=2021;,citation_fulltext_html_url=https://arxiv.org/abs/2105.03418;">
<meta name="citation_reference" content="citation_title=Deep Learning Hamiltonian Monte Carlo;,citation_author=Sam Foreman;,citation_author=Xiao-Yong Jin;,citation_author=James C. Osborn;,citation_publication_date=2021-05;,citation_cover_date=2021-05;,citation_year=2021;,citation_fulltext_html_url=https://arxiv.org/abs/2105.03418;,citation_conference_title=9th International Conference on Learning Representations;">
<meta name="citation_reference" content="citation_title=Deep learning hamiltonian monte carlo;,citation_author=Sam Foreman;,citation_author=Xiao-Yong Jin;,citation_author=James C Osborn;,citation_publication_date=2021;,citation_cover_date=2021;,citation_year=2021;,citation_journal_title=arXiv preprint arXiv:2105.03418;">
<meta name="citation_reference" content="citation_title=LeapfrogLayers: A trainable framework for effective topological sampling;,citation_author=Sam Foreman;,citation_author=Xiao-Yong Jin;,citation_author=James C Osborn;,citation_publication_date=2021;,citation_cover_date=2021;,citation_year=2021;,citation_journal_title=arXiv preprint arXiv:2112.01582;">
<meta name="citation_reference" content="citation_title=Energy Justice Analysis of Climate Data with ClimRR;,citation_author=Sam Foreman;,citation_publication_date=2023-08-07;,citation_cover_date=2023-08-07;,citation_year=2023;,citation_fulltext_html_url=https://saforem2.github.io/climate-analysis;,citation_language=en;">
<meta name="citation_reference" content="citation_author=Sam Foreman;,citation_publication_date=2023-08-19;,citation_cover_date=2023-08-19;,citation_year=2023;,citation_fulltext_html_url=https://saforem2.github.io/l2hmc-qcd;,citation_language=en;">
<meta name="citation_reference" content="citation_title=MLMC: Machine learning monte carlo for lattice gauge theory;,citation_author=Sam Foreman;,citation_author=Xiao-Yong Jin;,citation_author=James Osborn;,citation_publication_date=00;,citation_cover_date=00;,citation_year=0;,citation_conference_title=40th international symposium on lattice field theory (lattice 2023) (batavia, IL, united states, 07/31/2023 - 08/04/2023);">
<meta name="citation_reference" content="citation_title=Progress on $(g-2)_\mu$ from lattice QCD;,citation_author=Hartmut Wittig;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=https://arxiv.org/abs/2306.04165;">
<meta name="citation_reference" content="citation_title=Hybrid Monte Carlo;,citation_author=S. Duane;,citation_author=A. D. Kennedy;,citation_author=B. J. Pendleton;,citation_author=D. Roweth;,citation_publication_date=1987;,citation_cover_date=1987;,citation_year=1987;,citation_doi=10.1016/0370-2693(87)91197-X;,citation_volume=195;,citation_journal_title=Phys. Lett. B;">
<meta name="citation_reference" content="citation_title=Snowmass 2021 Computational Frontier CompF03 Topical Group Report: Machine Learning;,citation_author=Phiala Shanahan;,citation_author=others;,citation_publication_date=2022-09;,citation_cover_date=2022-09;,citation_year=2022;,citation_fulltext_html_url=https://arxiv.org/abs/2209.07559;">
<meta name="citation_reference" content="citation_title=Applications of Machine Learning to Lattice Quantum Field Theory;,citation_author=Denis Boyda;,citation_author=others;,citation_publication_date=2022-02;,citation_cover_date=2022-02;,citation_year=2022;,citation_fulltext_html_url=https://arxiv.org/abs/2202.05838;,citation_conference_title=Snowmass 2021;">
<meta name="citation_reference" content="citation_title=LeapfrogLayers: A Trainable Framework for Effective Topological Sampling;,citation_author=S. Foreman;,citation_author=X. Jin;,citation_author=J. Osborn;,citation_publication_date=2022-07;,citation_cover_date=2022-07;,citation_year=2022;,citation_fulltext_html_url=https://arxiv.org/abs/2112.01582;,citation_doi=10.22323/1.396.0508;,citation_conference_title=The 38th international symposium on lattice field theory;">
<meta name="citation_reference" content="citation_title=HMC with Normalizing Flows;,citation_author=Sam Foreman;,citation_author=Taku Izubuchi;,citation_author=Luchang Jin;,citation_author=Xiao-Yong Jin;,citation_author=James C. Osborn;,citation_author=Akio Tomiya;,citation_publication_date=2022;,citation_cover_date=2022;,citation_year=2022;,citation_fulltext_html_url=https://arxiv.org/abs/2112.01586;,citation_doi=10.22323/1.396.0073;,citation_volume=LATTICE2021;,citation_journal_title=PoS;">
<meta name="citation_reference" content="citation_title=Mastering language models;,citation_author=Samuel Montgomery;,citation_publication_date=2023-10;,citation_cover_date=2023-10;,citation_year=2023;,citation_fulltext_html_url=https://towardsdatascience.com/mastering-language-models-32e1d891511a
           ;,citation_journal_title=Medium;,citation_publisher=Towards Data Science;">
<meta name="citation_reference" content="citation_title=Harnessing the power of LLMs in practice: A survey on ChatGPT and beyond;,citation_author=Jingfeng Yang;,citation_author=Hongye Jin;,citation_author=Ruixiang Tang;,citation_author=Xiaotian Han;,citation_author=Qizhang Feng;,citation_author=Haoming Jiang;,citation_author=Bing Yin;,citation_author=Xia Hu;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=https://arxiv.org/abs/2304.13712;">
<meta name="citation_reference" content="citation_title=Training tips for the transformer model;,citation_author=Martin Popel;,citation_author=Ondřej Bojar;,citation_publication_date=2018-04;,citation_cover_date=2018-04;,citation_year=2018;,citation_fulltext_html_url=https://doi.org/10.2478%2Fpralin-2018-0002;,citation_issue=1;,citation_doi=10.2478/pralin-2018-0002;,citation_volume=110;,citation_journal_title=The Prague Bulletin of Mathematical Linguistics;,citation_publisher=Charles University in Prague, Karolinum Press;">
<meta name="citation_reference" content="citation_title=Attention is all you need;,citation_author=Ashish Vaswani;,citation_author=Noam Shazeer;,citation_author=Niki Parmar;,citation_author=Jakob Uszkoreit;,citation_author=Llion Jones;,citation_author=Aidan N. Gomez;,citation_author=Lukasz Kaiser;,citation_author=Illia Polosukhin;,citation_publication_date=2017;,citation_cover_date=2017;,citation_year=2017;,citation_fulltext_html_url=https://arxiv.org/abs/1706.03762;">
<meta name="citation_reference" content="citation_title=Tree of thoughts: Deliberate problem solving with large language models;,citation_author=Shunyu Yao;,citation_author=Dian Yu;,citation_author=Jeffrey Zhao;,citation_author=Izhak Shafran;,citation_author=Thomas L. Griffiths;,citation_author=Yuan Cao;,citation_author=Karthik Narasimhan;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=https://arxiv.org/abs/2305.10601;">
<meta name="citation_reference" content="citation_title=GenSLMs: Genome-scale language models reveal SARS-CoV-2 evolutionary dynamics;,citation_abstract=We seek to transform how new and emergent variants of pandemiccausing viruses, specifically SARS-CoV-2, are identified and classified. By adapting large language models (LLMs) for genomic data, we build genome-scale language models (GenSLMs) which can learn the evolutionary landscape of SARS-CoV-2 genomes. By pretraining on over 110 million prokaryotic gene sequences and finetuning a SARS-CoV-2-specific model on 1.5 million genomes, we show that GenSLMs can accurately and rapidly identify variants of concern. Thus, to our knowledge, GenSLMs represents one of the first whole genome scale foundation models which can generalize to other prediction tasks. We demonstrate scaling of GenSLMs on GPU-based supercomputers and AI-hardware accelerators utilizing 1.63 Zettaflops in training runs with a sustained performance of 121 PFLOPS in mixed precision and peak of 850 PFLOPS. We present initial scientific insights from examining GenSLMs in tracking evolutionary dynamics of SARS-CoV-2, paving the path to realizing this on large biological data.Competing Interest StatementThe authors have declared no competing interest.;,citation_author=Maxim Zvyagin;,citation_author=Alexander Brace;,citation_author=Kyle Hippe;,citation_author=Yuntian Deng;,citation_author=Bin Zhang;,citation_author=Cindy Orozco Bohorquez;,citation_author=Austin Clyde;,citation_author=Bharat Kale;,citation_author=Danilo Perez-Rivera;,citation_author=Heng Ma;,citation_author=Carla M. Mann;,citation_author=Michael Irvin;,citation_author=J. Gregory Pauloski;,citation_author=Logan Ward;,citation_author=Valerie Hayot-Sasson;,citation_author=Murali Emani;,citation_author=Sam Foreman;,citation_author=Zhen Xie;,citation_author=Diangen Lin;,citation_author=Maulik Shukla;,citation_author=Weili Nie;,citation_author=Josh Romero;,citation_author=Christian Dallago;,citation_author=Arash Vahdat;,citation_author=Chaowei Xiao;,citation_author=Thomas Gibbs;,citation_author=Ian Foster;,citation_author=James J. Davis;,citation_author=Michael E. Papka;,citation_author=Thomas Brettin;,citation_author=Rick Stevens;,citation_author=Anima Anandkumar;,citation_author=Venkatram Vishwanath;,citation_author=Arvind Ramanathan;,citation_publication_date=2022;,citation_cover_date=2022;,citation_year=2022;,citation_fulltext_html_url=https://www.biorxiv.org/content/early/2022/11/23/2022.10.10.511571;,citation_doi=10.1101/2022.10.10.511571;,citation_journal_title=bioRxiv;,citation_publisher=Cold Spring Harbor Laboratory;">
</head>

<body class="nav-sidebar floating quarto-dark"><script id="quarto-html-before-body" type="application/javascript">
    const toggleBodyColorMode = (bsSheetEl) => {
      const mode = bsSheetEl.getAttribute("data-mode");
      const bodyEl = window.document.querySelector("body");
      if (mode === "dark") {
        bodyEl.classList.add("quarto-dark");
        bodyEl.classList.remove("quarto-light");
      } else {
        bodyEl.classList.add("quarto-light");
        bodyEl.classList.remove("quarto-dark");
      }
    }
    const toggleBodyColorPrimary = () => {
      const bsSheetEl = window.document.querySelector("link#quarto-bootstrap:not([rel=disabled-stylesheet])");
      if (bsSheetEl) {
        toggleBodyColorMode(bsSheetEl);
      }
    }
    const setColorSchemeToggle = (alternate) => {
      const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
      for (let i=0; i < toggles.length; i++) {
        const toggle = toggles[i];
        if (toggle) {
          if (alternate) {
            toggle.classList.add("alternate");
          } else {
            toggle.classList.remove("alternate");
          }
        }
      }
    };
    const toggleColorMode = (alternate) => {
      // Switch the stylesheets
      const primaryStylesheets = window.document.querySelectorAll('link.quarto-color-scheme:not(.quarto-color-alternate)');
      const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
      manageTransitions('#quarto-margin-sidebar .nav-link', false);
      if (alternate) {
        // note: dark is layered on light, we don't disable primary!
        enableStylesheet(alternateStylesheets);
        for (const sheetNode of alternateStylesheets) {
          if (sheetNode.id === "quarto-bootstrap") {
            toggleBodyColorMode(sheetNode);
          }
        }
      } else {
        disableStylesheet(alternateStylesheets);
        enableStylesheet(primaryStylesheets)
        toggleBodyColorPrimary();
      }
      manageTransitions('#quarto-margin-sidebar .nav-link', true);
      // Switch the toggles
      setColorSchemeToggle(alternate)
      // Hack to workaround the fact that safari doesn't
      // properly recolor the scrollbar when toggling (#1455)
      if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
        manageTransitions("body", false);
        window.scrollTo(0, 1);
        setTimeout(() => {
          window.scrollTo(0, 0);
          manageTransitions("body", true);
        }, 40);
      }
    }
    const disableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        stylesheet.rel = 'disabled-stylesheet';
      }
    }
    const enableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        if(stylesheet.rel !== 'stylesheet') { // for Chrome, which will still FOUC without this check
          stylesheet.rel = 'stylesheet';
        }
      }
    }
    const manageTransitions = (selector, allowTransitions) => {
      const els = window.document.querySelectorAll(selector);
      for (let i=0; i < els.length; i++) {
        const el = els[i];
        if (allowTransitions) {
          el.classList.remove('notransition');
        } else {
          el.classList.add('notransition');
        }
      }
    }
    const isFileUrl = () => {
      return window.location.protocol === 'file:';
    }
    const hasAlternateSentinel = () => {
      let styleSentinel = getColorSchemeSentinel();
      if (styleSentinel !== null) {
        return styleSentinel === "alternate";
      } else {
        return false;
      }
    }
    const setStyleSentinel = (alternate) => {
      const value = alternate ? "alternate" : "default";
      if (!isFileUrl()) {
        window.localStorage.setItem("quarto-color-scheme", value);
      } else {
        localAlternateSentinel = value;
      }
    }
    const getColorSchemeSentinel = () => {
      if (!isFileUrl()) {
        const storageValue = window.localStorage.getItem("quarto-color-scheme");
        return storageValue != null ? storageValue : localAlternateSentinel;
      } else {
        return localAlternateSentinel;
      }
    }
    const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
      const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
      const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
      let newTheme = '';
      if(authorPrefersDark) {
        newTheme = isAlternate ? baseTheme : alternateTheme;
      } else {
        newTheme = isAlternate ? alternateTheme : baseTheme;
      }
      const changeGiscusTheme = () => {
        // From: https://github.com/giscus/giscus/issues/336
        const sendMessage = (message) => {
          const iframe = document.querySelector('iframe.giscus-frame');
          if (!iframe) return;
          iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
        }
        sendMessage({
          setConfig: {
            theme: newTheme
          }
        });
      }
      const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
      if (isGiscussLoaded) {
        changeGiscusTheme();
      }
    };
    const authorPrefersDark = true;
    const queryPrefersDark = window.matchMedia('(prefers-color-scheme: dark)');
    const darkModeDefault = queryPrefersDark.matches;
    let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
    // Dark / light mode switch
    window.quartoToggleColorScheme = () => {
      // Read the current dark / light value
      let toAlternate = !hasAlternateSentinel();
      toggleColorMode(toAlternate);
      setStyleSentinel(toAlternate);
      toggleGiscusIfUsed(toAlternate, darkModeDefault);
      window.dispatchEvent(new Event('resize'));
    };
    queryPrefersDark.addEventListener("change", e => {
      if(window.localStorage.getItem("quarto-color-scheme") !== null)
        return;
      const alternate = e.matches
      toggleColorMode(alternate);
      localAlternateSentinel = e.matches ? 'alternate' : 'default'; // this is used alongside local storage!
      toggleGiscusIfUsed(alternate, darkModeDefault);
    });
    // Switch to dark mode if need be
    if (hasAlternateSentinel()) {
      toggleColorMode(true);
    } else {
      toggleColorMode(false);
    }
  </script>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../00-intro-AI-HPC/index.html">[00] Intro to AI and HPC</a></li><li class="breadcrumb-item"><a href="../../00-intro-AI-HPC/7-statistical-learning/index.html">[7] Statistical Learning</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="../../index.html" class="sidebar-logo-link">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">Intro to HPC Bootcamp 2025</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="../../index.html" title="Home" class="quarto-navigation-tool px-1" aria-label="Home"><i class="bi bi-home"></i></a>
    <a href="https://github.com/saforem2/intro-hpc-bootcamp-2025" title="GitHub" class="quarto-navigation-tool px-1" aria-label="GitHub"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../00-intro-AI-HPC/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[00] Intro to AI and HPC</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../00-intro-AI-HPC/0-compute-systems/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[0] Compute Systems</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../00-intro-AI-HPC/1-shared-resources/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[1] Shared Resources</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../00-intro-AI-HPC/2-jupyter-notebooks/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[2] Jupyter Notebooks</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../00-intro-AI-HPC/3-python/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[3] Using Python</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../00-intro-AI-HPC/4-data/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[4] Working with Data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../00-intro-AI-HPC/5-mcmc-example/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[5] MCMC Example</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../00-intro-AI-HPC/6-linear-regression/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[6] Linear Regression</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../00-intro-AI-HPC/7-statistical-learning/index.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">[7] Statistical Learning</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../00-intro-AI-HPC/8-clustering/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[8] Clustering</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../01-neural-networks/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[01] Neural Networks</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../01-neural-networks/0-intro/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[0] Intro to NNs</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../01-neural-networks/1-mnist/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[1] MNIST Example</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../01-neural-networks/1-mnist-ipynb/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[1] MNIST Example (ipynb)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../01-neural-networks/2-advanced/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[2] Advanced</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../01-neural-networks/3-conv-nets/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[3] Conv. Nets</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../01-neural-networks/4-representation-learning/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[4] Representation Learning</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../01-neural-networks/5-distributed-training/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[5] Distributed Training</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../02-llms/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[02] Large Language Models</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../02-llms/00-intro-to-llms/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[0] Intro to LLMs</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../02-llms/01-hands-on-llms/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[1] Hands-on LLMs</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../02-llms/06-parallel-training/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[6] Parallel Training</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../02-llms/08-shakespeare-example-colab/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">[8] Shakespeare Example (Colab)</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#learning-to-cluster-data" id="toc-learning-to-cluster-data" class="nav-link active" data-scroll-target="#learning-to-cluster-data">Learning to Cluster Data</a>
  <ul class="collapse">
  <li><a href="#toy-dataset-2d-blobs" id="toc-toy-dataset-2d-blobs" class="nav-link" data-scroll-target="#toy-dataset-2d-blobs">Toy Dataset (2D Blobs)</a></li>
  </ul></li>
  <li><a href="#k-means-clustering" id="toc-k-means-clustering" class="nav-link" data-scroll-target="#k-means-clustering">K-means Clustering</a>
  <ul class="collapse">
  <li><a href="#step-1-pick-random-starting-centroids-from-data" id="toc-step-1-pick-random-starting-centroids-from-data" class="nav-link" data-scroll-target="#step-1-pick-random-starting-centroids-from-data">Step 1: Pick random starting centroids from data</a></li>
  <li><a href="#step-2a-calculate-distance-to-each-centroid" id="toc-step-2a-calculate-distance-to-each-centroid" class="nav-link" data-scroll-target="#step-2a-calculate-distance-to-each-centroid">Step 2a: Calculate distance to each centroid</a></li>
  <li><a href="#step-2b-find-nearest-cluster-for-each-point" id="toc-step-2b-find-nearest-cluster-for-each-point" class="nav-link" data-scroll-target="#step-2b-find-nearest-cluster-for-each-point">Step 2b: Find nearest cluster for each point</a></li>
  <li><a href="#step-3-calculate-new-centroids" id="toc-step-3-calculate-new-centroids" class="nav-link" data-scroll-target="#step-3-calculate-new-centroids">Step 3: Calculate new centroids</a></li>
  <li><a href="#step-4-repeat-until-convergence" id="toc-step-4-repeat-until-convergence" class="nav-link" data-scroll-target="#step-4-repeat-until-convergence">Step 4: Repeat until convergence</a></li>
  </ul></li>
  <li><a href="#run-example" id="toc-run-example" class="nav-link" data-scroll-target="#run-example">Run Example</a></li>
  <li><a href="#k-means-on-a-breast-cancer-dataset" id="toc-k-means-on-a-breast-cancer-dataset" class="nav-link" data-scroll-target="#k-means-on-a-breast-cancer-dataset">K-Means on a Breast Cancer Dataset</a></li>
  <li><a href="#self-organizing-maps" id="toc-self-organizing-maps" class="nav-link" data-scroll-target="#self-organizing-maps">Self-organizing maps</a>
  <ul class="collapse">
  <li><a href="#generating-color-data-and-normalizing" id="toc-generating-color-data-and-normalizing" class="nav-link" data-scroll-target="#generating-color-data-and-normalizing">Generating “Color” data and normalizing</a></li>
  <li><a href="#defining-som" id="toc-defining-som" class="nav-link" data-scroll-target="#defining-som">Defining SOM</a></li>
  <li><a href="#initial-state-of-som-color-network" id="toc-initial-state-of-som-color-network" class="nav-link" data-scroll-target="#initial-state-of-som-color-network">Initial state of SOM color network</a></li>
  <li><a href="#training-som" id="toc-training-som" class="nav-link" data-scroll-target="#training-som">Training SOM</a></li>
  <li><a href="#visualization-of-trained-colormap-som" id="toc-visualization-of-trained-colormap-som" class="nav-link" data-scroll-target="#visualization-of-trained-colormap-som">Visualization of trained colormap SOM</a></li>
  <li><a href="#som-on-cancer-data" id="toc-som-on-cancer-data" class="nav-link" data-scroll-target="#som-on-cancer-data">SOM on Cancer data</a></li>
  <li><a href="#training-som-on-cancer-data" id="toc-training-som-on-cancer-data" class="nav-link" data-scroll-target="#training-som-on-cancer-data">Training SOM on Cancer data</a></li>
  <li><a href="#visualization-of-trained-som" id="toc-visualization-of-trained-som" class="nav-link" data-scroll-target="#visualization-of-trained-som">Visualization of trained SOM</a></li>
  </ul></li>
  <li><a href="#keep-learning" id="toc-keep-learning" class="nav-link" data-scroll-target="#keep-learning">Keep learning</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/saforem2/intro-hpc-bootcamp-2025/blob/main/00-intro-AI-HPC/7-statistical-learning/index.qmd" class="toc-action"><i class="bi bi-github"></i>View source</a></li><li><a href="https://github.com/saforem2/intro-hpc-bootcamp-2025/edit/main/00-intro-AI-HPC/7-statistical-learning/index.qmd" class="toc-action"><i class="bi empty"></i>Edit this page</a></li><li><a href="https://github.com/saforem2/intro-hpc-bootcamp-2025/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div><div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="index.ipynb" download="index.ipynb"><i class="bi bi-journal-code"></i>Jupyter</a></li><li><a href="README.md"><i class="bi bi-file-code"></i>Github (GFM)</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<!-- Google Tag Manager (noscript) -->
<noscript><iframe src="https://www.googletagmanager.com/ns.html?id=GTM-TC329HJ" height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>
<!-- End Google Tag Manager (noscript) -->

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../00-intro-AI-HPC/index.html">[00] Intro to AI and HPC</a></li><li class="breadcrumb-item"><a href="../../00-intro-AI-HPC/7-statistical-learning/index.html">[7] Statistical Learning</a></li></ol></nav>
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Statistical Learning</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>

<div>
  <div class="description">
    Statistical learning is a framework for understanding and modeling complex data patterns, combining statistical methods with machine learning algorithms to build predictive models.
  </div>
</div>

<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Authors</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Sam Foreman <a href="mailto:foremans@anl.gov" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0002-9981-0876" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="https://alcf.anl.gov/about/people/sam-foreman">
            </a><a href="https://www.anl.gov/">ANL</a>
            
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Huihuo Zheng <a href="mailto:huihuo.zheng@anl.gov" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="https://alcf.anl.gov/about/people/huihuo-zheng">
            </a><a href="https://www.anl.gov/">ANL</a>
            
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">July 25, 2025</p>
    </div>
  </div>
  
    <div>
    <div class="quarto-title-meta-heading">Modified</div>
    <div class="quarto-title-meta-contents">
      <p class="date-modified">August 11, 2025</p>
    </div>
  </div>
    
  </div>
  


</header>


<p><a href="https://colab.research.google.com/github/saforem2/intro-hpc-bootcamp-2025/blob/main/docs/00-intro-AI-HPC/7-statistical-learning/index.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" class="img-fluid"></a> <a href="https://github.com/saforem2/intro-hpc-bootcamp-2025/blob/main/docs/00-intro-AI-HPC/7-statistical-learning/README.md"><img src="https://img.shields.io/badge/-View%20on%20GitHub-333333?style=flat&amp;logo=github&amp;labelColor=gray.png" class="img-fluid"></a></p>
<div id="4e0f8677" class="cell" data-execution_count="2">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1"></a><span class="op">%</span>load_ext autoreload</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="op">%</span>autoreload <span class="dv">2</span></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="im">import</span> matplotlib_inline.backend_inline</span>
<span id="cb1-4"><a href="#cb1-4"></a>matplotlib_inline.backend_inline.set_matplotlib_formats(<span class="st">'retina'</span>, <span class="st">'svg'</span>, <span class="st">'png'</span>)</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="im">import</span> os</span>
<span id="cb1-6"><a href="#cb1-6"></a>os.environ[<span class="st">"TRUECOLOR"</span>] <span class="op">=</span> <span class="st">"1"</span></span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="im">import</span> matplotlib <span class="im">as</span> mpl</span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="co"># mpl.rcParams['figure.dpi'] = 400</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<section id="learning-to-cluster-data" class="level2">
<h2 class="anchored" data-anchor-id="learning-to-cluster-data">Learning to Cluster Data</h2>
<p>If we have data with distinct groupings, the objective is to devise a method for labeling our data by its group in an automated way.</p>
<p>We will demonstrate this, first, on a toy dataset that we <em>design</em> to have a lower inherent dimensionality, then we move to a higher dimensional dataset.</p>
<p><a href="https://colab.research.google.com/github/saforem2/intro-hpc-bootcamp-2025/blob/main/docs/00-intro-AI-HPC/7-statistical-learning/index.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" class="img-fluid"></a></p>
<section id="toy-dataset-2d-blobs" class="level3">
<h3 class="anchored" data-anchor-id="toy-dataset-2d-blobs">Toy Dataset (2D Blobs)</h3>
<p>Define 3 blobs on a 2D plane:</p>
<div id="15421320" class="cell" data-execution_count="3">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1"></a><span class="im">import</span> time</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="im">import</span> os</span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="im">import</span> sklearn</span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="im">from</span> sklearn <span class="im">import</span> datasets</span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb2-6"><a href="#cb2-6"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb2-7"><a href="#cb2-7"></a></span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="im">from</span> sklearn.preprocessing <span class="im">import</span> (</span>
<span id="cb2-9"><a href="#cb2-9"></a>    StandardScaler,</span>
<span id="cb2-10"><a href="#cb2-10"></a>    MinMaxScaler,</span>
<span id="cb2-11"><a href="#cb2-11"></a>    RobustScaler</span>
<span id="cb2-12"><a href="#cb2-12"></a>)</span>
<span id="cb2-13"><a href="#cb2-13"></a><span class="im">from</span> sklearn.cluster <span class="im">import</span> KMeans</span>
<span id="cb2-14"><a href="#cb2-14"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="im">import</span> ambivalent</span>
<span id="cb2-16"><a href="#cb2-16"></a>plt.style.use(ambivalent.STYLES[<span class="st">"ambivalent"</span>])</span>
<span id="cb2-17"><a href="#cb2-17"></a></span>
<span id="cb2-18"><a href="#cb2-18"></a><span class="im">from</span> matplotlib.colors <span class="im">import</span> ListedColormap</span>
<span id="cb2-19"><a href="#cb2-19"></a><span class="im">from</span> bootcamp.plots <span class="im">import</span> scatter, COLORS</span>
<span id="cb2-20"><a href="#cb2-20"></a></span>
<span id="cb2-21"><a href="#cb2-21"></a><span class="co"># set random seed for reproducibility</span></span>
<span id="cb2-22"><a href="#cb2-22"></a>SEED <span class="op">=</span> <span class="dv">42</span></span>
<span id="cb2-23"><a href="#cb2-23"></a>DFIGSIZE <span class="op">=</span> plt.rcParamsDefault[<span class="st">'figure.figsize'</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">[<span style="color: #7f7f7f; text-decoration-color: #7f7f7f">2025-08-04 11:08:05,329776</span>][<span style="color: #008000; text-decoration-color: #008000">I</span>][<span style="color: #008080; text-decoration-color: #008080; font-style: italic">ezpz</span>/<span style="color: #000080; text-decoration-color: #000080">__init__</span><span style="color: #0000ff; text-decoration-color: #0000ff">:</span><span style="color: #800080; text-decoration-color: #800080">265</span><span style="color: #0000ff; text-decoration-color: #0000ff">:</span><span style="color: #00ff00; text-decoration-color: #00ff00; font-style: italic">ezpz</span>]<span style="color: #c0c0c0; text-decoration-color: #c0c0c0"> </span>Setting logging level to <span style="color: #008000; text-decoration-color: #008000">'INFO'</span> on <span style="color: #008000; text-decoration-color: #008000">'RANK == 0'</span>
</pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">[<span style="color: #7f7f7f; text-decoration-color: #7f7f7f">2025-08-04 11:08:05,331927</span>][<span style="color: #008000; text-decoration-color: #008000">I</span>][<span style="color: #008080; text-decoration-color: #008080; font-style: italic">ezpz</span>/<span style="color: #000080; text-decoration-color: #000080">__init__</span><span style="color: #0000ff; text-decoration-color: #0000ff">:</span><span style="color: #800080; text-decoration-color: #800080">266</span><span style="color: #0000ff; text-decoration-color: #0000ff">:</span><span style="color: #00ff00; text-decoration-color: #00ff00; font-style: italic">ezpz</span>]<span style="color: #c0c0c0; text-decoration-color: #c0c0c0"> </span>Setting logging level to <span style="color: #008000; text-decoration-color: #008000">'CRITICAL'</span> on all others <span style="color: #008000; text-decoration-color: #008000">'RANK != 0'</span>
</pre>
</div>
</div>
<div id="cf95cca2" class="cell" data-execution_count="4">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1"></a>n_samples <span class="op">=</span> <span class="dv">1000</span> <span class="co"># 300 2D data points</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>n_features <span class="op">=</span> <span class="dv">2</span>  <span class="co"># 2D data</span></span>
<span id="cb3-3"><a href="#cb3-3"></a>n_clusters <span class="op">=</span> <span class="dv">4</span>  <span class="co"># 3 unique blobs</span></span>
<span id="cb3-4"><a href="#cb3-4"></a></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co"># https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_blobs.html</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="co"># -- Returns --------------------------------------------------------</span></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="co"># x (n_samples, n_features): The generated samples.</span></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="co"># y (n_samples,): Int labels for cluster membership of each sample.</span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="co"># -------------------------------------------------------------------</span></span>
<span id="cb3-10"><a href="#cb3-10"></a>cmap <span class="op">=</span> ListedColormap(<span class="bu">list</span>(COLORS.values()))</span>
<span id="cb3-11"><a href="#cb3-11"></a>x, y <span class="op">=</span> datasets.make_blobs(</span>
<span id="cb3-12"><a href="#cb3-12"></a>    n_samples<span class="op">=</span>n_samples,</span>
<span id="cb3-13"><a href="#cb3-13"></a>    n_features<span class="op">=</span>n_features,</span>
<span id="cb3-14"><a href="#cb3-14"></a>    centers<span class="op">=</span>n_clusters,</span>
<span id="cb3-15"><a href="#cb3-15"></a>    random_state<span class="op">=</span><span class="dv">42</span></span>
<span id="cb3-16"><a href="#cb3-16"></a>)</span>
<span id="cb3-17"><a href="#cb3-17"></a>scatter_kwargs <span class="op">=</span> {</span>
<span id="cb3-18"><a href="#cb3-18"></a>    <span class="st">'xlabel'</span>: <span class="st">'x0'</span>,</span>
<span id="cb3-19"><a href="#cb3-19"></a>    <span class="st">'ylabel'</span>: <span class="st">'x1'</span>,</span>
<span id="cb3-20"><a href="#cb3-20"></a>    <span class="st">'cmap'</span>: cmap,</span>
<span id="cb3-21"><a href="#cb3-21"></a>    <span class="st">'plot_kwargs'</span>: {</span>
<span id="cb3-22"><a href="#cb3-22"></a>        <span class="st">'alpha'</span>: <span class="fl">0.8</span>,</span>
<span id="cb3-23"><a href="#cb3-23"></a>        <span class="st">'edgecolor'</span>: <span class="st">'#222'</span>,</span>
<span id="cb3-24"><a href="#cb3-24"></a>    }</span>
<span id="cb3-25"><a href="#cb3-25"></a>}</span>
<span id="cb3-26"><a href="#cb3-26"></a>_ <span class="op">=</span> scatter(x, y, <span class="op">**</span>scatter_kwargs)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="index_files/figure-html/cell-4-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img src="index_files/figure-html/cell-4-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="688d95f4" class="cell" data-execution_count="5">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># Normalize features</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="co"># https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html</span></span>
<span id="cb4-3"><a href="#cb4-3"></a>x_sc <span class="op">=</span> StandardScaler().fit_transform(x)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="dcf46b5a" class="cell" data-execution_count="6">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1"></a>fig, (ax1, ax2) <span class="op">=</span> plt.subplots(</span>
<span id="cb5-2"><a href="#cb5-2"></a>    figsize<span class="op">=</span>(<span class="fl">1.5</span> <span class="op">*</span> DFIGSIZE[<span class="dv">0</span>], <span class="fl">0.8</span> <span class="op">*</span> DFIGSIZE[<span class="dv">1</span>]),</span>
<span id="cb5-3"><a href="#cb5-3"></a>    ncols<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb5-4"><a href="#cb5-4"></a>    subplot_kw<span class="op">=</span>{</span>
<span id="cb5-5"><a href="#cb5-5"></a>        <span class="st">'aspect'</span>: <span class="st">'equal'</span>,</span>
<span id="cb5-6"><a href="#cb5-6"></a>    },</span>
<span id="cb5-7"><a href="#cb5-7"></a>    </span>
<span id="cb5-8"><a href="#cb5-8"></a>)</span>
<span id="cb5-9"><a href="#cb5-9"></a>fig, ax1 <span class="op">=</span> scatter(</span>
<span id="cb5-10"><a href="#cb5-10"></a>    x,</span>
<span id="cb5-11"><a href="#cb5-11"></a>    y,</span>
<span id="cb5-12"><a href="#cb5-12"></a>    fig<span class="op">=</span>fig,</span>
<span id="cb5-13"><a href="#cb5-13"></a>    ax<span class="op">=</span>ax1,</span>
<span id="cb5-14"><a href="#cb5-14"></a>    title<span class="op">=</span><span class="st">'Original'</span>,</span>
<span id="cb5-15"><a href="#cb5-15"></a>    <span class="op">**</span>scatter_kwargs</span>
<span id="cb5-16"><a href="#cb5-16"></a>)</span>
<span id="cb5-17"><a href="#cb5-17"></a>fig.subplots_adjust(wspace<span class="op">=</span><span class="fl">0.2</span>)</span>
<span id="cb5-18"><a href="#cb5-18"></a>fig, ax2 <span class="op">=</span> scatter(</span>
<span id="cb5-19"><a href="#cb5-19"></a>    x_sc,</span>
<span id="cb5-20"><a href="#cb5-20"></a>    y,</span>
<span id="cb5-21"><a href="#cb5-21"></a>    fig<span class="op">=</span>fig,</span>
<span id="cb5-22"><a href="#cb5-22"></a>    ax<span class="op">=</span>ax2,</span>
<span id="cb5-23"><a href="#cb5-23"></a>    title<span class="op">=</span><span class="st">'Normalized'</span>,</span>
<span id="cb5-24"><a href="#cb5-24"></a>    <span class="op">**</span>scatter_kwargs</span>
<span id="cb5-25"><a href="#cb5-25"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="index_files/figure-html/cell-6-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="index_files/figure-html/cell-6-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="k-means-clustering" class="level2">
<h2 class="anchored" data-anchor-id="k-means-clustering">K-means Clustering</h2>
<p>K-means clustering aims to partition n observations into k clusters in which each observation belongs to the cluster with the nearest mean, serving as a prototype of the cluster. This results in a partitioning of the data space into Voronoi cells.</p>
<ol type="1">
<li>Pick random starting centroids from data
<ul>
<li><span class="math inline">k</span> initial “means” (in this case, <span class="math inline">k=3</span>) are randomly generated within the data domain (shown in color)</li>
</ul></li>
<li>Calculate distance to each centroid
<ul>
<li><span class="math inline">k</span> clusters are created by associating every observation with the nearest mean. The partitions here represent the Voroni diagram generated by the means</li>
<li>Find nearest cluster for each point</li>
</ul></li>
<li>Calculate new centroids
<ul>
<li>The <strong>centroid</strong> of each of the <span class="math inline">k</span> clusters becomes the new mean</li>
</ul></li>
<li>Steps 2 and 3 are repeated until convergence has been reached</li>
</ol>
<div id="fig-kmeans" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-kmeans-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="flex-container">
<p><a href="./assets/k-means0.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-3" title="Figure&nbsp;1: Illustration of the K-means algorithm steps"><img src="./assets/k-means0.svg" class="img-fluid figure-img"></a></p>
<p><a href="./assets/k-means1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-4" title="Figure&nbsp;1: Illustration of the K-means algorithm steps"><img src="./assets/k-means1.svg" class="img-fluid figure-img"></a></p>
<p><a href="./assets/k-means2.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-5" title="Figure&nbsp;1: Illustration of the K-means algorithm steps"><img src="./assets/k-means2.svg" class="img-fluid figure-img"></a></p>
<p><a href="./assets/k-means3.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-6" title="Figure&nbsp;1: Illustration of the K-means algorithm steps"><img src="./assets/k-means3.svg" class="img-fluid figure-img"></a></p>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-kmeans-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Illustration of the K-means algorithm steps
</figcaption>
</figure>
</div>
<section id="step-1-pick-random-starting-centroids-from-data" class="level3">
<h3 class="anchored" data-anchor-id="step-1-pick-random-starting-centroids-from-data">Step 1: Pick random starting centroids from data</h3>
<ul>
<li><span class="math inline">k</span> initial “means” (in this case, <span class="math inline">k=3</span>) are randomly generated within the data domain (shown in color)</li>
</ul>
<div id="fig-kmeans-step1" class="quarto-float quarto-figure quarto-figure-center anchored" style="width:40%">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-kmeans-step1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="./assets/k-means0.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-7" title="Figure&nbsp;2: "><img src="./assets/k-means0.svg" class="img-fluid figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-kmeans-step1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2
</figcaption>
</figure>
</div>
<div id="ebfba10d" class="cell" data-execution_count="7">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1"></a><span class="kw">def</span> initialize_centroids(</span>
<span id="cb6-2"><a href="#cb6-2"></a>        x: np.ndarray,</span>
<span id="cb6-3"><a href="#cb6-3"></a>        n_clusters: <span class="bu">int</span>,</span>
<span id="cb6-4"><a href="#cb6-4"></a>        seed: <span class="bu">int</span> <span class="op">=</span> <span class="dv">123</span></span>
<span id="cb6-5"><a href="#cb6-5"></a>) <span class="op">-&gt;</span> np.ndarray:</span>
<span id="cb6-6"><a href="#cb6-6"></a>    <span class="co">"""Initialize centroids.</span></span>
<span id="cb6-7"><a href="#cb6-7"></a><span class="co">    </span></span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="co">    Inputs:</span></span>
<span id="cb6-9"><a href="#cb6-9"></a><span class="co">      - x (np.ndarray): Data of shape (num_points, num_features)</span></span>
<span id="cb6-10"><a href="#cb6-10"></a><span class="co">      - n_clusters (int): Number of clusters to use</span></span>
<span id="cb6-11"><a href="#cb6-11"></a><span class="co">      - seed (int): Random seed.</span></span>
<span id="cb6-12"><a href="#cb6-12"></a><span class="co">        </span></span>
<span id="cb6-13"><a href="#cb6-13"></a><span class="co">    Outputs:</span></span>
<span id="cb6-14"><a href="#cb6-14"></a><span class="co">      - centroids (np.ndarray): Randomly chosen from the data</span></span>
<span id="cb6-15"><a href="#cb6-15"></a><span class="co">        with shape: (num_clusters, num_features).</span></span>
<span id="cb6-16"><a href="#cb6-16"></a><span class="co">    """</span></span>
<span id="cb6-17"><a href="#cb6-17"></a>    np.random.RandomState(seed)</span>
<span id="cb6-18"><a href="#cb6-18"></a>    <span class="co"># 1. Randomly permute data points</span></span>
<span id="cb6-19"><a href="#cb6-19"></a>    <span class="co"># 2. From this, pick the first `n_clusters` indices</span></span>
<span id="cb6-20"><a href="#cb6-20"></a>    <span class="co"># 3. Return these as our initial centroids</span></span>
<span id="cb6-21"><a href="#cb6-21"></a>    random_idx <span class="op">=</span> np.random.permutation(x.shape[<span class="dv">0</span>])</span>
<span id="cb6-22"><a href="#cb6-22"></a>    centroids <span class="op">=</span> x[random_idx[:n_clusters]]</span>
<span id="cb6-23"><a href="#cb6-23"></a>    <span class="cf">return</span> centroids</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="step-2a-calculate-distance-to-each-centroid" class="level3">
<h3 class="anchored" data-anchor-id="step-2a-calculate-distance-to-each-centroid">Step 2a: Calculate distance to each centroid</h3>
<ul>
<li>Calculate distance to each centroid
<ul>
<li><span class="math inline">k</span> clusters are created by associating every observation with the nearest mean. The partitions here represent the Voroni diagram generated by the means.</li>
<li>Find nearest cluster for each point</li>
</ul></li>
</ul>
<div id="fig-kmeans-step2a" class="quarto-float quarto-figure quarto-figure-center anchored" style="width:40%;">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-kmeans-step2a-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="./assets/k-means1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-8" title="Figure&nbsp;3: "><img src="./assets/k-means1.svg" class="img-fluid figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-kmeans-step2a-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3
</figcaption>
</figure>
</div>
<div id="524def99" class="cell" data-execution_count="8">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">def</span> compute_distance(</span>
<span id="cb7-2"><a href="#cb7-2"></a>        x: np.ndarray,</span>
<span id="cb7-3"><a href="#cb7-3"></a>        centroids: np.ndarray,</span>
<span id="cb7-4"><a href="#cb7-4"></a>        n_clusters: <span class="bu">int</span></span>
<span id="cb7-5"><a href="#cb7-5"></a>) <span class="op">-&gt;</span> np.ndarray:</span>
<span id="cb7-6"><a href="#cb7-6"></a>    <span class="co">"""Compute distance.</span></span>
<span id="cb7-7"><a href="#cb7-7"></a></span>
<span id="cb7-8"><a href="#cb7-8"></a><span class="co">    Inputs:</span></span>
<span id="cb7-9"><a href="#cb7-9"></a><span class="co">      - x (np.ndarray): Input data of shape</span></span>
<span id="cb7-10"><a href="#cb7-10"></a><span class="co">        (num_points, num_features)</span></span>
<span id="cb7-11"><a href="#cb7-11"></a><span class="co">      - centroids (np.ndarray): Cluster centroids  with shape </span></span>
<span id="cb7-12"><a href="#cb7-12"></a><span class="co">        (num_clusters, num_features)</span></span>
<span id="cb7-13"><a href="#cb7-13"></a><span class="co">      - n_clusters (int): Number of clusters being used.</span></span>
<span id="cb7-14"><a href="#cb7-14"></a></span>
<span id="cb7-15"><a href="#cb7-15"></a><span class="co">    Outputs:</span></span>
<span id="cb7-16"><a href="#cb7-16"></a><span class="co">      - distance (np.ndarray): Distance of each point </span></span>
<span id="cb7-17"><a href="#cb7-17"></a><span class="co">        to each centroid with shape </span></span>
<span id="cb7-18"><a href="#cb7-18"></a><span class="co">        (num_points, num_clusters)</span></span>
<span id="cb7-19"><a href="#cb7-19"></a><span class="co">    """</span></span>
<span id="cb7-20"><a href="#cb7-20"></a>    <span class="co"># distance vector</span></span>
<span id="cb7-21"><a href="#cb7-21"></a>    distance <span class="op">=</span> np.zeros((x.shape[<span class="dv">0</span>], n_clusters))</span>
<span id="cb7-22"><a href="#cb7-22"></a>    <span class="co"># loop over each centroid</span></span>
<span id="cb7-23"><a href="#cb7-23"></a>    <span class="cf">for</span> k <span class="kw">in</span> <span class="bu">range</span>(n_clusters):</span>
<span id="cb7-24"><a href="#cb7-24"></a>        <span class="co"># calculate distance for each point from centroid</span></span>
<span id="cb7-25"><a href="#cb7-25"></a>        kcentroid_distance <span class="op">=</span> x <span class="op">-</span> centroids[k, :]</span>
<span id="cb7-26"><a href="#cb7-26"></a>        <span class="co"># apply normalization for stability</span></span>
<span id="cb7-27"><a href="#cb7-27"></a>        row_norm <span class="op">=</span> np.linalg.norm(kcentroid_distance, axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb7-28"><a href="#cb7-28"></a>        <span class="co"># return distance squared</span></span>
<span id="cb7-29"><a href="#cb7-29"></a>        distance[:, k] <span class="op">=</span> np.square(row_norm) </span>
<span id="cb7-30"><a href="#cb7-30"></a></span>
<span id="cb7-31"><a href="#cb7-31"></a>    <span class="cf">return</span> distance</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="step-2b-find-nearest-cluster-for-each-point" class="level3">
<h3 class="anchored" data-anchor-id="step-2b-find-nearest-cluster-for-each-point">Step 2b: Find nearest cluster for each point</h3>
<div id="9fc946f4" class="cell" data-execution_count="9">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw">def</span> find_closest_centroid(</span>
<span id="cb8-2"><a href="#cb8-2"></a>        distance: np.ndarray</span>
<span id="cb8-3"><a href="#cb8-3"></a>) <span class="op">-&gt;</span> np.ndarray:</span>
<span id="cb8-4"><a href="#cb8-4"></a>    <span class="co">"""Find closest centroid.</span></span>
<span id="cb8-5"><a href="#cb8-5"></a></span>
<span id="cb8-6"><a href="#cb8-6"></a><span class="co">    Inputs:</span></span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="co">      - distance (np.ndarray): Distance of each point to each centroid with shape</span></span>
<span id="cb8-8"><a href="#cb8-8"></a><span class="co">        (num_points, num_clusters)</span></span>
<span id="cb8-9"><a href="#cb8-9"></a></span>
<span id="cb8-10"><a href="#cb8-10"></a><span class="co">    Outputs:</span></span>
<span id="cb8-11"><a href="#cb8-11"></a><span class="co">      - nearest_centroid_indices (np.ndarray): Index of nearest centroid with shape</span></span>
<span id="cb8-12"><a href="#cb8-12"></a><span class="co">        (num_points,)</span></span>
<span id="cb8-13"><a href="#cb8-13"></a><span class="co">    """</span></span>
<span id="cb8-14"><a href="#cb8-14"></a>    nearest_centroid_indices <span class="op">=</span> np.argmin(distance, axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb8-15"><a href="#cb8-15"></a>    <span class="cf">return</span> nearest_centroid_indices</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="step-3-calculate-new-centroids" class="level3">
<h3 class="anchored" data-anchor-id="step-3-calculate-new-centroids">Step 3: Calculate new centroids</h3>
<ol start="3" type="1">
<li>Calculate new centroids
<ul>
<li>The <strong>centroid</strong> of each of the <span class="math inline">k</span> clusters becomes the new mean</li>
</ul></li>
</ol>
<div id="fig-kmeans-step3" class="quarto-float quarto-figure quarto-figure-center anchored" style="width:40%;">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-kmeans-step3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="./assets/k-means2.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-9" title="Figure&nbsp;4: "><img src="./assets/k-means2.svg" class="img-fluid figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-kmeans-step3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4
</figcaption>
</figure>
</div>
<div id="ad219841" class="cell" data-execution_count="10">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1"></a><span class="kw">def</span> compute_centroids(</span>
<span id="cb9-2"><a href="#cb9-2"></a>        x: np.ndarray,</span>
<span id="cb9-3"><a href="#cb9-3"></a>        nearest_centroid_indices: np.ndarray,</span>
<span id="cb9-4"><a href="#cb9-4"></a>        n_clusters: <span class="bu">int</span></span>
<span id="cb9-5"><a href="#cb9-5"></a>) <span class="op">-&gt;</span> np.ndarray:</span>
<span id="cb9-6"><a href="#cb9-6"></a>    <span class="co">"""Compute centroids.</span></span>
<span id="cb9-7"><a href="#cb9-7"></a><span class="co">    </span></span>
<span id="cb9-8"><a href="#cb9-8"></a><span class="co">    Inputs:</span></span>
<span id="cb9-9"><a href="#cb9-9"></a><span class="co">      - x (np.ndarray): Input data of shape</span></span>
<span id="cb9-10"><a href="#cb9-10"></a><span class="co">        (num_points, num_features)</span></span>
<span id="cb9-11"><a href="#cb9-11"></a><span class="co">      - nearest_centroid_indices: Index of nearest centroid of shape</span></span>
<span id="cb9-12"><a href="#cb9-12"></a><span class="co">        (num_points,)</span></span>
<span id="cb9-13"><a href="#cb9-13"></a><span class="co">      - n_clusters (int): Number of clusters being used</span></span>
<span id="cb9-14"><a href="#cb9-14"></a><span class="co">      </span></span>
<span id="cb9-15"><a href="#cb9-15"></a><span class="co">    Outputs:</span></span>
<span id="cb9-16"><a href="#cb9-16"></a><span class="co">      - centroids (np.ndarray): Cluster centroids with shape</span></span>
<span id="cb9-17"><a href="#cb9-17"></a><span class="co">        (num_clusters, num_features)</span></span>
<span id="cb9-18"><a href="#cb9-18"></a><span class="co">    """</span></span>
<span id="cb9-19"><a href="#cb9-19"></a>    <span class="co"># new centroids vector</span></span>
<span id="cb9-20"><a href="#cb9-20"></a>    centroids <span class="op">=</span> np.zeros((n_clusters, x.shape[<span class="dv">1</span>]))</span>
<span id="cb9-21"><a href="#cb9-21"></a>    <span class="co"># loop over each centroids</span></span>
<span id="cb9-22"><a href="#cb9-22"></a>    <span class="cf">for</span> k <span class="kw">in</span> <span class="bu">range</span>(n_clusters):</span>
<span id="cb9-23"><a href="#cb9-23"></a>        <span class="co"># calculate the mean of all points assigned to this centroid</span></span>
<span id="cb9-24"><a href="#cb9-24"></a>        centroids[k, :] <span class="op">=</span> np.mean(</span>
<span id="cb9-25"><a href="#cb9-25"></a>            x[nearest_centroid_indices <span class="op">==</span> k, :],</span>
<span id="cb9-26"><a href="#cb9-26"></a>            axis<span class="op">=</span><span class="dv">0</span></span>
<span id="cb9-27"><a href="#cb9-27"></a>        )</span>
<span id="cb9-28"><a href="#cb9-28"></a>    <span class="cf">return</span> centroids</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="step-4-repeat-until-convergence" class="level3">
<h3 class="anchored" data-anchor-id="step-4-repeat-until-convergence">Step 4: Repeat until convergence</h3>
<ol start="4" type="1">
<li>Repeat steps 2 and 3 until convergence has been reached</li>
</ol>
<div id="fig-kmeans-step4" class="quarto-float quarto-figure quarto-figure-center anchored" style="width:40%;">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-kmeans-step4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="./assets/k-means3.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-10" title="Figure&nbsp;5: "><img src="./assets/k-means3.svg" class="img-fluid figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-kmeans-step4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5
</figcaption>
</figure>
</div>
<div id="3aeca31a" class="cell" data-execution_count="11">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1"></a><span class="im">from</span> __future__ <span class="im">import</span> absolute_import, annotations, division, print_function</span>
<span id="cb10-2"><a href="#cb10-2"></a></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="im">from</span> typing <span class="im">import</span> Optional</span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="im">import</span> IPython.display <span class="im">as</span> ipydis</span>
<span id="cb10-6"><a href="#cb10-6"></a></span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="im">from</span> bootcamp.plots <span class="im">import</span> plot_kmeans_points</span>
<span id="cb10-8"><a href="#cb10-8"></a></span>
<span id="cb10-9"><a href="#cb10-9"></a><span class="kw">def</span> apply_kmeans(</span>
<span id="cb10-10"><a href="#cb10-10"></a>    x: np.ndarray,</span>
<span id="cb10-11"><a href="#cb10-11"></a>    n_clusters: <span class="bu">int</span>,</span>
<span id="cb10-12"><a href="#cb10-12"></a>    iterations: <span class="bu">int</span> <span class="op">=</span> <span class="dv">100</span>,</span>
<span id="cb10-13"><a href="#cb10-13"></a>    seed: <span class="bu">int</span> <span class="op">=</span> <span class="dv">123</span>,</span>
<span id="cb10-14"><a href="#cb10-14"></a>    cmap: Optional[<span class="bu">str</span>] <span class="op">=</span> <span class="va">None</span>,</span>
<span id="cb10-15"><a href="#cb10-15"></a>) <span class="op">-&gt;</span> <span class="bu">tuple</span>[np.ndarray, np.ndarray]:</span>
<span id="cb10-16"><a href="#cb10-16"></a>    <span class="co">"""Returns (centroids, cluster_id)."""</span></span>
<span id="cb10-17"><a href="#cb10-17"></a>    <span class="co"># ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓</span></span>
<span id="cb10-18"><a href="#cb10-18"></a>    <span class="co"># ┃ initialize centroids:               ┃</span></span>
<span id="cb10-19"><a href="#cb10-19"></a>    <span class="co"># ┃   - shape: (n_clusters, x.shape[1]) ┃</span></span>
<span id="cb10-20"><a href="#cb10-20"></a>    <span class="co"># ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛</span></span>
<span id="cb10-21"><a href="#cb10-21"></a>    centroids <span class="op">=</span> initialize_centroids(</span>
<span id="cb10-22"><a href="#cb10-22"></a>        x,  <span class="co"># (n_points, n_features)</span></span>
<span id="cb10-23"><a href="#cb10-23"></a>        n_clusters<span class="op">=</span>n_clusters,</span>
<span id="cb10-24"><a href="#cb10-24"></a>        seed<span class="op">=</span>seed,</span>
<span id="cb10-25"><a href="#cb10-25"></a>    )</span>
<span id="cb10-26"><a href="#cb10-26"></a>    <span class="co"># ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓</span></span>
<span id="cb10-27"><a href="#cb10-27"></a>    <span class="co"># ┃ -- Iteratively improve centroid location ----------------- ┃</span></span>
<span id="cb10-28"><a href="#cb10-28"></a>    <span class="co"># ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛</span></span>
<span id="cb10-29"><a href="#cb10-29"></a>    <span class="co"># ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓</span></span>
<span id="cb10-30"><a href="#cb10-30"></a>    <span class="co"># ┃ 1. Compute the distance from entries in x to each centroid ┃</span></span>
<span id="cb10-31"><a href="#cb10-31"></a>    <span class="co"># ┃    - distance.shape: (n_points, n_clusters)                ┃</span></span>
<span id="cb10-32"><a href="#cb10-32"></a>    <span class="co"># ┃ 2. Return the closest cluster (0, 1, ..., n_clusters-1)    ┃</span></span>
<span id="cb10-33"><a href="#cb10-33"></a>    <span class="co"># ┃    - cluster_id.shape: (n_points)                          ┃</span></span>
<span id="cb10-34"><a href="#cb10-34"></a>    <span class="co"># ┃ 3. Calculate the mean position of each labeled cluster     ┃</span></span>
<span id="cb10-35"><a href="#cb10-35"></a>    <span class="co"># ┃    - centroids.shape: (n_clusters, n_features)             ┃</span></span>
<span id="cb10-36"><a href="#cb10-36"></a>    <span class="co"># ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛</span></span>
<span id="cb10-37"><a href="#cb10-37"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(iterations):</span>
<span id="cb10-38"><a href="#cb10-38"></a>        <span class="co"># save old centroids</span></span>
<span id="cb10-39"><a href="#cb10-39"></a>        old_centroids <span class="op">=</span> centroids</span>
<span id="cb10-40"><a href="#cb10-40"></a>        distance <span class="op">=</span> compute_distance(x, old_centroids, n_clusters)</span>
<span id="cb10-41"><a href="#cb10-41"></a>        cluster_id <span class="op">=</span> find_closest_centroid(distance)</span>
<span id="cb10-42"><a href="#cb10-42"></a>        centroids <span class="op">=</span> compute_centroids(x, cluster_id, n_clusters)</span>
<span id="cb10-43"><a href="#cb10-43"></a></span>
<span id="cb10-44"><a href="#cb10-44"></a>        <span class="co"># plotting for visual comprehension</span></span>
<span id="cb10-45"><a href="#cb10-45"></a>        ipydis.clear_output(<span class="st">"wait"</span>)</span>
<span id="cb10-46"><a href="#cb10-46"></a>        <span class="bu">print</span>(<span class="ss">f"Iteration: </span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb10-47"><a href="#cb10-47"></a>        plot_kmeans_points(x, centroids, cluster_id, cmap<span class="op">=</span>cmap)</span>
<span id="cb10-48"><a href="#cb10-48"></a>        time.sleep(<span class="fl">0.5</span>)</span>
<span id="cb10-49"><a href="#cb10-49"></a></span>
<span id="cb10-50"><a href="#cb10-50"></a>        <span class="co"># if our points are the same as the old centroids, then we can stop</span></span>
<span id="cb10-51"><a href="#cb10-51"></a>        <span class="cf">if</span> np.<span class="bu">all</span>(old_centroids <span class="op">==</span> centroids):</span>
<span id="cb10-52"><a href="#cb10-52"></a>            <span class="bu">print</span>(<span class="ss">f"No change in centroids! Exiting!"</span>)</span>
<span id="cb10-53"><a href="#cb10-53"></a>            <span class="cf">break</span></span>
<span id="cb10-54"><a href="#cb10-54"></a></span>
<span id="cb10-55"><a href="#cb10-55"></a>    <span class="cf">return</span> centroids, cluster_id</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
</section>
<section id="run-example" class="level2">
<h2 class="anchored" data-anchor-id="run-example">Run Example</h2>
<div id="a078e163" class="cell" data-execution_count="12">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1"></a><span class="im">from</span> bootcamp.plots <span class="im">import</span> COLORS</span>
<span id="cb11-2"><a href="#cb11-2"></a></span>
<span id="cb11-3"><a href="#cb11-3"></a>cmap <span class="op">=</span> ListedColormap(<span class="bu">list</span>(COLORS.values()))</span>
<span id="cb11-4"><a href="#cb11-4"></a></span>
<span id="cb11-5"><a href="#cb11-5"></a>n_samples <span class="op">=</span> <span class="dv">500</span>  <span class="co"># 300 2D data points</span></span>
<span id="cb11-6"><a href="#cb11-6"></a>n_features <span class="op">=</span> <span class="dv">2</span>  <span class="co"># 2D data</span></span>
<span id="cb11-7"><a href="#cb11-7"></a>n_clusters_true <span class="op">=</span> <span class="dv">5</span>  <span class="co"># unique blobs</span></span>
<span id="cb11-8"><a href="#cb11-8"></a>n_clusters_guess <span class="op">=</span> <span class="dv">5</span></span>
<span id="cb11-9"><a href="#cb11-9"></a>SEED <span class="op">=</span> <span class="dv">456</span></span>
<span id="cb11-10"><a href="#cb11-10"></a>iterations <span class="op">=</span> <span class="dv">50</span></span>
<span id="cb11-11"><a href="#cb11-11"></a></span>
<span id="cb11-12"><a href="#cb11-12"></a><span class="co"># https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_blobs.html</span></span>
<span id="cb11-13"><a href="#cb11-13"></a>x, y <span class="op">=</span> datasets.make_blobs(</span>
<span id="cb11-14"><a href="#cb11-14"></a>    n_samples<span class="op">=</span>n_samples,</span>
<span id="cb11-15"><a href="#cb11-15"></a>    n_features<span class="op">=</span>n_features,</span>
<span id="cb11-16"><a href="#cb11-16"></a>    centers<span class="op">=</span>n_clusters_true,</span>
<span id="cb11-17"><a href="#cb11-17"></a>    random_state<span class="op">=</span>SEED,</span>
<span id="cb11-18"><a href="#cb11-18"></a>)</span>
<span id="cb11-19"><a href="#cb11-19"></a></span>
<span id="cb11-20"><a href="#cb11-20"></a>x_sc <span class="op">=</span> StandardScaler().fit_transform(x)</span>
<span id="cb11-21"><a href="#cb11-21"></a></span>
<span id="cb11-22"><a href="#cb11-22"></a>centroids, cluster_id <span class="op">=</span> apply_kmeans(</span>
<span id="cb11-23"><a href="#cb11-23"></a>    x_sc,</span>
<span id="cb11-24"><a href="#cb11-24"></a>    n_clusters_guess,</span>
<span id="cb11-25"><a href="#cb11-25"></a>    iterations,</span>
<span id="cb11-26"><a href="#cb11-26"></a>    SEED,</span>
<span id="cb11-27"><a href="#cb11-27"></a>    cmap<span class="op">=</span>cmap,</span>
<span id="cb11-28"><a href="#cb11-28"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Iteration: 9</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="index_files/figure-html/cell-12-output-2.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="index_files/figure-html/cell-12-output-2.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>No change in centroids! Exiting!</code></pre>
</div>
</div>
</section>
<section id="k-means-on-a-breast-cancer-dataset" class="level2">
<h2 class="anchored" data-anchor-id="k-means-on-a-breast-cancer-dataset">K-Means on a Breast Cancer Dataset</h2>
<p>Now we use more realistic data that cannot be easily plotted on a 2D grid. This dataset has 30 features (columns) for 569 patients (rows). In addition, there is a <em>target</em> feature that indicates if the cancer was <em>malignant</em> (0) or <em>benign</em> (1). In the ideal case, our 30 features would provide easy deliniation between these two classes.</p>
<p>Let’s extract our data into <span class="math inline">x</span> and our truth labels into <span class="math inline">y</span></p>
<div id="303e66cc" class="cell" data-execution_count="13">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb14-2"><a href="#cb14-2"></a></span>
<span id="cb14-3"><a href="#cb14-3"></a></span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="kw">def</span> load_cancer_data() <span class="op">-&gt;</span> <span class="bu">dict</span>:</span>
<span id="cb14-5"><a href="#cb14-5"></a>    <span class="co">"""Return cancer dataset (unscaled)."""</span></span>
<span id="cb14-6"><a href="#cb14-6"></a>    <span class="im">from</span> sklearn <span class="im">import</span> datasets</span>
<span id="cb14-7"><a href="#cb14-7"></a></span>
<span id="cb14-8"><a href="#cb14-8"></a>    data <span class="op">=</span> datasets.load_breast_cancer()</span>
<span id="cb14-9"><a href="#cb14-9"></a>    <span class="cf">return</span> data</span>
<span id="cb14-10"><a href="#cb14-10"></a></span>
<span id="cb14-11"><a href="#cb14-11"></a></span>
<span id="cb14-12"><a href="#cb14-12"></a><span class="kw">def</span> sort_cancer_data(data: <span class="bu">dict</span>) <span class="op">-&gt;</span> <span class="bu">tuple</span>[pd.DataFrame, pd.Series]:</span>
<span id="cb14-13"><a href="#cb14-13"></a>    <span class="co"># Get features and target</span></span>
<span id="cb14-14"><a href="#cb14-14"></a>    x <span class="op">=</span> pd.DataFrame(</span>
<span id="cb14-15"><a href="#cb14-15"></a>        data[<span class="st">"data"</span>],</span>
<span id="cb14-16"><a href="#cb14-16"></a>        columns<span class="op">=</span>data[<span class="st">"feature_names"</span>],</span>
<span id="cb14-17"><a href="#cb14-17"></a>    )</span>
<span id="cb14-18"><a href="#cb14-18"></a>    x <span class="op">=</span> x[<span class="bu">sorted</span>(x.columns)]</span>
<span id="cb14-19"><a href="#cb14-19"></a>    y <span class="op">=</span> data[<span class="st">"target"</span>]</span>
<span id="cb14-20"><a href="#cb14-20"></a>    <span class="cf">return</span> x, y</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="19442d0f" class="cell" data-execution_count="14">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1"></a>data <span class="op">=</span> load_cancer_data()</span>
<span id="cb15-2"><a href="#cb15-2"></a>x, y <span class="op">=</span> sort_cancer_data(data)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="018af91d" class="cell" data-execution_count="15">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1"></a><span class="bu">print</span>(data.keys())</span>
<span id="cb16-2"><a href="#cb16-2"></a><span class="bu">print</span>(<span class="st">"data size:"</span>, <span class="bu">len</span>(data[<span class="st">"data"</span>]))</span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="bu">print</span>(<span class="st">"number of features:"</span>, <span class="bu">len</span>(data[<span class="st">"feature_names"</span>]))</span>
<span id="cb16-4"><a href="#cb16-4"></a><span class="bu">print</span>(data[<span class="st">"feature_names"</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])
data size: 569
number of features: 30
['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']</code></pre>
</div>
</div>
<div id="de50d29a" class="cell" data-execution_count="16">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1"></a><span class="co"># wrap data in pandas DataFrame</span></span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="co"># y = data['target'] # value: 0 = 'Malignant' 1 = 'Benign'</span></span>
<span id="cb18-3"><a href="#cb18-3"></a>x <span class="op">=</span> data[<span class="st">"data"</span>]</span>
<span id="cb18-4"><a href="#cb18-4"></a>y <span class="op">=</span> data[<span class="st">"target"</span>]</span>
<span id="cb18-5"><a href="#cb18-5"></a><span class="bu">print</span>(<span class="ss">f"(malignant, benign): </span><span class="sc">{</span>np<span class="sc">.</span>bincount(y)<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb18-6"><a href="#cb18-6"></a><span class="bu">print</span>(<span class="ss">f"x.shape: </span><span class="sc">{</span>x<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb18-7"><a href="#cb18-7"></a><span class="bu">print</span>(<span class="ss">f"y.shape: </span><span class="sc">{</span>y<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>(malignant, benign): [212 357]
x.shape: (569, 30)
y.shape: (569,)</code></pre>
</div>
</div>
<div id="39efb3f9" class="cell" data-execution_count="17">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1"></a><span class="co"># normalize</span></span>
<span id="cb20-2"><a href="#cb20-2"></a>x_sc <span class="op">=</span> StandardScaler().fit_transform(x)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Plot histogram of number of true class labels (malignant/benign) and number of k-means cluster labels.</p>
<p>This gives an indication of how well we did in clustering our data, but is not a “correctness” or “accuracy” metric.</p>
<div class="callout callout-style-simple callout-note">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-body-container">
<p>Keep in mind, k-means is not a classifier.</p>
<p>Cluster label 0 given from k-means, does not correspond to cluster label 0 in the truth.</p>
</div>
</div>
</div>
<div id="13b63e27" class="cell" data-execution_count="18">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1"></a><span class="im">from</span> bootcamp.plots <span class="im">import</span> plot_hists</span>
<span id="cb21-2"><a href="#cb21-2"></a></span>
<span id="cb21-3"><a href="#cb21-3"></a><span class="co"># for seed in range(10):</span></span>
<span id="cb21-4"><a href="#cb21-4"></a>CSEED <span class="op">=</span> <span class="dv">42</span></span>
<span id="cb21-5"><a href="#cb21-5"></a>kmeans <span class="op">=</span> KMeans(n_clusters<span class="op">=</span><span class="dv">2</span>, random_state<span class="op">=</span>CSEED)</span>
<span id="cb21-6"><a href="#cb21-6"></a><span class="co"># fit the data</span></span>
<span id="cb21-7"><a href="#cb21-7"></a>kfit <span class="op">=</span> kmeans.fit(x_sc)</span>
<span id="cb21-8"><a href="#cb21-8"></a>cluster_ids <span class="op">=</span> kmeans.labels_</span>
<span id="cb21-9"><a href="#cb21-9"></a><span class="bu">print</span>(</span>
<span id="cb21-10"><a href="#cb21-10"></a>    <span class="ss">f"seed: </span><span class="sc">{</span>CSEED<span class="sc">}</span><span class="ch">\n</span><span class="ss">Number of samples in each cluster: </span><span class="sc">{</span>np<span class="sc">.</span>bincount(kmeans.labels_)<span class="sc">}</span><span class="ss">"</span></span>
<span id="cb21-11"><a href="#cb21-11"></a>)</span>
<span id="cb21-12"><a href="#cb21-12"></a>k_means_bins <span class="op">=</span> np.bincount(cluster_ids)</span>
<span id="cb21-13"><a href="#cb21-13"></a>y_bins <span class="op">=</span> np.bincount(y)</span>
<span id="cb21-14"><a href="#cb21-14"></a>plt.rcParams[<span class="st">"figure.figsize"</span>] <span class="op">=</span> [<span class="fl">1.5</span> <span class="op">*</span> DFIGSIZE[<span class="dv">0</span>], <span class="fl">0.8</span> <span class="op">*</span> DFIGSIZE[<span class="dv">1</span>]]</span>
<span id="cb21-15"><a href="#cb21-15"></a>plot_hists(k_means_bins, y_bins, xlabels<span class="op">=</span>[<span class="st">"Malignant"</span>, <span class="st">"Benign"</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>seed: 42
Number of samples in each cluster: [188 381]</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="index_files/figure-html/cell-18-output-2.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="index_files/figure-html/cell-18-output-2.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>This plot shows the normalized inertia as we vary the number <code>n_clusters</code> used in our k-means fit to the breast cancer data. This value essentially indicates the mean distance of a point to the cluster centroids. Obviously more clusters result in data points being nearer to a centroid.</p>
<p>If our dataset easily split into 2 clusters, we would see a step function behavior going from 1 to 2, then only very minor improvements above 2. What we see here tells us our data does not easily cluster.</p>
<div id="3ecb0be9" class="cell" data-execution_count="19">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1"></a><span class="im">from</span> bootcamp.plots <span class="im">import</span> plot_kmeans_obj</span>
<span id="cb23-2"><a href="#cb23-2"></a></span>
<span id="cb23-3"><a href="#cb23-3"></a>plt.rcParams[<span class="st">"figure.figsize"</span>] <span class="op">=</span> [DFIGSIZE[<span class="dv">0</span>], <span class="fl">0.6</span> <span class="op">*</span> DFIGSIZE[<span class="dv">1</span>]]</span>
<span id="cb23-4"><a href="#cb23-4"></a>_ <span class="op">=</span> plot_kmeans_obj(x_sc, nclusters<span class="op">=</span><span class="dv">10</span>, plot_points<span class="op">=</span><span class="va">False</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="index_files/figure-html/cell-19-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img src="index_files/figure-html/cell-19-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>For example, if we return to our blob data with 2 clusters, it become clear.</p>
<div id="c79b22fd" class="cell" data-execution_count="20">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1"></a>n_samples <span class="op">=</span> <span class="dv">300</span>  <span class="co"># 300 2D data points</span></span>
<span id="cb24-2"><a href="#cb24-2"></a>n_features <span class="op">=</span> <span class="dv">2</span>  <span class="co"># 2D data</span></span>
<span id="cb24-3"><a href="#cb24-3"></a>n_clusters <span class="op">=</span> <span class="dv">2</span>  <span class="co"># unique blobs</span></span>
<span id="cb24-4"><a href="#cb24-4"></a>seed <span class="op">=</span> <span class="dv">456</span></span>
<span id="cb24-5"><a href="#cb24-5"></a>iterations <span class="op">=</span> <span class="dv">25</span></span>
<span id="cb24-6"><a href="#cb24-6"></a>x, y <span class="op">=</span> datasets.make_blobs(</span>
<span id="cb24-7"><a href="#cb24-7"></a>    n_samples<span class="op">=</span>n_samples, n_features<span class="op">=</span>n_features, centers<span class="op">=</span>n_clusters, random_state<span class="op">=</span>seed</span>
<span id="cb24-8"><a href="#cb24-8"></a>)</span>
<span id="cb24-9"><a href="#cb24-9"></a>plot_kmeans_obj(x, nclusters<span class="op">=</span><span class="dv">10</span>, plot_points<span class="op">=</span><span class="va">False</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display" data-execution_count="19">
<pre><code>array([71.2677332 ,  2.06848282,  1.69706679,  1.2870742 ,  1.09504851,
        0.89354269,  0.9207715 ,  0.6767132 ,  0.62271334])</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="index_files/figure-html/cell-20-output-2.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img src="index_files/figure-html/cell-20-output-2.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="self-organizing-maps" class="level2">
<h2 class="anchored" data-anchor-id="self-organizing-maps"><a href="https://en.wikipedia.org/wiki/Self-organizing_map">Self-organizing maps</a></h2>
<p>Self Organizing Maps (SOM) were proposed and became widespread in the 1980s, by a Finnish professor named Teuvo Kohonen and are also called ‘Kohonen maps’.</p>
<div id="dffe90d4" class="cell" data-execution_count="21">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1"></a><span class="co"># finding best matching unit</span></span>
<span id="cb26-2"><a href="#cb26-2"></a><span class="kw">def</span> find_bmu(t, net, n):</span>
<span id="cb26-3"><a href="#cb26-3"></a>    <span class="co">"""</span></span>
<span id="cb26-4"><a href="#cb26-4"></a><span class="co">    Find the best matching unit for a given vector, t, in the SOM</span></span>
<span id="cb26-5"><a href="#cb26-5"></a><span class="co">    Returns: a (bmu, bmu_idx) tuple where bmu is the high-dimensional BMU</span></span>
<span id="cb26-6"><a href="#cb26-6"></a><span class="co">             and bmu_idx is the index of this vector in the SOM</span></span>
<span id="cb26-7"><a href="#cb26-7"></a><span class="co">    """</span></span>
<span id="cb26-8"><a href="#cb26-8"></a>    bmu_idx <span class="op">=</span> np.array([<span class="dv">0</span>, <span class="dv">0</span>])</span>
<span id="cb26-9"><a href="#cb26-9"></a>    <span class="co"># set the initial minimum distance to a huge number</span></span>
<span id="cb26-10"><a href="#cb26-10"></a>    <span class="co"># min_dist = #np.iinfo(np.int).max</span></span>
<span id="cb26-11"><a href="#cb26-11"></a>    min_dist <span class="op">=</span> np.inf</span>
<span id="cb26-12"><a href="#cb26-12"></a>    <span class="co"># calculate the high-dimensional distance between each neuron and the input</span></span>
<span id="cb26-13"><a href="#cb26-13"></a>    <span class="cf">for</span> x <span class="kw">in</span> <span class="bu">range</span>(net.shape[<span class="dv">0</span>]):</span>
<span id="cb26-14"><a href="#cb26-14"></a>        <span class="cf">for</span> y <span class="kw">in</span> <span class="bu">range</span>(net.shape[<span class="dv">1</span>]):</span>
<span id="cb26-15"><a href="#cb26-15"></a>            w <span class="op">=</span> net[x, y, :].reshape(n, <span class="dv">1</span>)</span>
<span id="cb26-16"><a href="#cb26-16"></a>            <span class="co"># don't bother with actual Euclidean distance, to avoid expensive sqrt operation</span></span>
<span id="cb26-17"><a href="#cb26-17"></a>            sq_dist <span class="op">=</span> np.<span class="bu">sum</span>((w <span class="op">-</span> t) <span class="op">**</span> <span class="dv">2</span>)</span>
<span id="cb26-18"><a href="#cb26-18"></a>            <span class="cf">if</span> sq_dist <span class="op">&lt;</span> min_dist:</span>
<span id="cb26-19"><a href="#cb26-19"></a>                min_dist <span class="op">=</span> sq_dist</span>
<span id="cb26-20"><a href="#cb26-20"></a>                bmu_idx <span class="op">=</span> np.array([x, y])</span>
<span id="cb26-21"><a href="#cb26-21"></a>    <span class="co"># get vector corresponding to bmu_idx</span></span>
<span id="cb26-22"><a href="#cb26-22"></a>    bmu <span class="op">=</span> net[bmu_idx[<span class="dv">0</span>], bmu_idx[<span class="dv">1</span>], :].reshape(n, <span class="dv">1</span>)</span>
<span id="cb26-23"><a href="#cb26-23"></a>    <span class="co"># return the (bmu, bmu_idx) tuple</span></span>
<span id="cb26-24"><a href="#cb26-24"></a>    <span class="cf">return</span> (bmu, bmu_idx)</span>
<span id="cb26-25"><a href="#cb26-25"></a></span>
<span id="cb26-26"><a href="#cb26-26"></a></span>
<span id="cb26-27"><a href="#cb26-27"></a><span class="co"># Decaying radius of influence</span></span>
<span id="cb26-28"><a href="#cb26-28"></a><span class="kw">def</span> decay_radius(initial_radius, i, time_constant):</span>
<span id="cb26-29"><a href="#cb26-29"></a>    <span class="cf">return</span> initial_radius <span class="op">*</span> np.exp(<span class="op">-</span>i <span class="op">/</span> time_constant)</span>
<span id="cb26-30"><a href="#cb26-30"></a></span>
<span id="cb26-31"><a href="#cb26-31"></a></span>
<span id="cb26-32"><a href="#cb26-32"></a><span class="co"># Decaying learning rate</span></span>
<span id="cb26-33"><a href="#cb26-33"></a><span class="kw">def</span> decay_learning_rate(initial_learning_rate, i, n_iterations):</span>
<span id="cb26-34"><a href="#cb26-34"></a>    <span class="cf">return</span> initial_learning_rate <span class="op">*</span> np.exp(<span class="op">-</span>i <span class="op">/</span> n_iterations)</span>
<span id="cb26-35"><a href="#cb26-35"></a></span>
<span id="cb26-36"><a href="#cb26-36"></a></span>
<span id="cb26-37"><a href="#cb26-37"></a><span class="co"># Influence in 2D space</span></span>
<span id="cb26-38"><a href="#cb26-38"></a><span class="kw">def</span> calculate_influence(distance, radius):</span>
<span id="cb26-39"><a href="#cb26-39"></a>    <span class="cf">return</span> np.exp(<span class="op">-</span>distance <span class="op">/</span> (<span class="fl">2.0</span> <span class="op">*</span> (radius<span class="op">**</span><span class="dv">2</span>)))</span>
<span id="cb26-40"><a href="#cb26-40"></a></span>
<span id="cb26-41"><a href="#cb26-41"></a></span>
<span id="cb26-42"><a href="#cb26-42"></a><span class="co"># Update weights</span></span>
<span id="cb26-43"><a href="#cb26-43"></a><span class="kw">def</span> update_weights(net, bmu_idx, r, l):</span>
<span id="cb26-44"><a href="#cb26-44"></a>    wlen <span class="op">=</span> net.shape[<span class="dv">2</span>]</span>
<span id="cb26-45"><a href="#cb26-45"></a>    <span class="cf">for</span> x <span class="kw">in</span> <span class="bu">range</span>(net.shape[<span class="dv">0</span>]):</span>
<span id="cb26-46"><a href="#cb26-46"></a>        <span class="cf">for</span> y <span class="kw">in</span> <span class="bu">range</span>(net.shape[<span class="dv">1</span>]):</span>
<span id="cb26-47"><a href="#cb26-47"></a>            w <span class="op">=</span> net[x, y, :].reshape(wlen, <span class="dv">1</span>)</span>
<span id="cb26-48"><a href="#cb26-48"></a>            <span class="co"># get the 2-D distance (again, not the actual Euclidean distance)</span></span>
<span id="cb26-49"><a href="#cb26-49"></a>            w_dist <span class="op">=</span> np.<span class="bu">sum</span>((np.array([x, y]) <span class="op">-</span> bmu_idx) <span class="op">**</span> <span class="dv">2</span>)</span>
<span id="cb26-50"><a href="#cb26-50"></a>            <span class="co"># if the distance is within the current neighbourhood radius</span></span>
<span id="cb26-51"><a href="#cb26-51"></a>            <span class="cf">if</span> w_dist <span class="op">&lt;=</span> r<span class="op">**</span><span class="dv">2</span>:</span>
<span id="cb26-52"><a href="#cb26-52"></a>                <span class="co"># calculate the degree of influence (based on the 2-D distance)</span></span>
<span id="cb26-53"><a href="#cb26-53"></a>                influence <span class="op">=</span> calculate_influence(w_dist, r)</span>
<span id="cb26-54"><a href="#cb26-54"></a>                <span class="co"># now update the neuron's weight using the formula:</span></span>
<span id="cb26-55"><a href="#cb26-55"></a>                <span class="co"># new w = old w + (learning rate * influence * delta)</span></span>
<span id="cb26-56"><a href="#cb26-56"></a>                <span class="co"># where delta = input vector (t) - old w</span></span>
<span id="cb26-57"><a href="#cb26-57"></a>                new_w <span class="op">=</span> w <span class="op">+</span> (l <span class="op">*</span> influence <span class="op">*</span> (t <span class="op">-</span> w))</span>
<span id="cb26-58"><a href="#cb26-58"></a>                <span class="co"># commit the new weight</span></span>
<span id="cb26-59"><a href="#cb26-59"></a>                net[x, y, :] <span class="op">=</span> new_w.reshape(<span class="dv">1</span>, wlen)</span>
<span id="cb26-60"><a href="#cb26-60"></a>    <span class="cf">return</span> net</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>The idea behind a SOM is that you’re mapping high-dimensional vectors onto a smaller dimensional (typically 2D) space. Vectors that are close in the high-dimensional space also end up being mapped to nodes that are close in 2D space thus preserving the “topology” of the original data.</p>
<section id="generating-color-data-and-normalizing" class="level3">
<h3 class="anchored" data-anchor-id="generating-color-data-and-normalizing">Generating “Color” data and normalizing</h3>
<div id="719fc57a" class="cell" data-execution_count="22">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1"></a>raw_data <span class="op">=</span> np.random.randint(<span class="dv">0</span>, <span class="dv">255</span>, (<span class="dv">10000</span>, <span class="dv">3</span>))  <span class="co"># JPEG like data</span></span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="co"># raw_data = np.random.uniform(size=(500, 3))</span></span>
<span id="cb27-3"><a href="#cb27-3"></a>data <span class="op">=</span> StandardScaler().fit_transform(raw_data)  <span class="co"># Normalized</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="defining-som" class="level3">
<h3 class="anchored" data-anchor-id="defining-som">Defining SOM</h3>
<p>Defining network size, number of iterations and learning rate</p>
<div id="be39e20a" class="cell" data-execution_count="23">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1"></a>network_dimensions <span class="op">=</span> np.array([<span class="dv">50</span>, <span class="dv">50</span>])</span>
<span id="cb28-2"><a href="#cb28-2"></a>n_iterations <span class="op">=</span> <span class="dv">2500</span></span>
<span id="cb28-3"><a href="#cb28-3"></a>init_learning_rate <span class="op">=</span> <span class="fl">0.075</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Establish size variables based on data</p>
<div id="31ad9c60" class="cell" data-execution_count="24">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1"></a>m <span class="op">=</span> data.shape[<span class="dv">0</span>]</span>
<span id="cb29-2"><a href="#cb29-2"></a>n <span class="op">=</span> data.shape[<span class="dv">1</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Weight matrix (i.e.&nbsp;the SOM) needs to be one n-dimensional vector for each neuron in the SOM</p>
<div id="eebb500b" class="cell" data-execution_count="25">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1"></a>net <span class="op">=</span> np.random.random(</span>
<span id="cb30-2"><a href="#cb30-2"></a>    (network_dimensions[<span class="dv">0</span>], network_dimensions[<span class="dv">1</span>], n)</span>
<span id="cb30-3"><a href="#cb30-3"></a>)  <span class="co"># 25 neurons each with a 3D vector</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Initial neighbourhood radius and decay parameter</p>
<div id="e1e0b503" class="cell" data-execution_count="26">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1"></a>init_radius <span class="op">=</span> <span class="bu">max</span>(network_dimensions[<span class="dv">0</span>], network_dimensions[<span class="dv">1</span>]) <span class="op">/</span> <span class="fl">2.0</span></span>
<span id="cb31-2"><a href="#cb31-2"></a>time_constant <span class="op">=</span> n_iterations <span class="op">/</span> np.log(init_radius)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="initial-state-of-som-color-network" class="level3">
<h3 class="anchored" data-anchor-id="initial-state-of-som-color-network">Initial state of SOM color network</h3>
<div id="9269287c" class="cell" data-execution_count="27">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1"></a>plt.rcParams[<span class="st">"image.cmap"</span>] <span class="op">=</span> <span class="st">"rainbow"</span></span>
<span id="cb32-2"><a href="#cb32-2"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">4</span>, <span class="dv">4</span>))</span>
<span id="cb32-3"><a href="#cb32-3"></a>image <span class="op">=</span> ax.matshow(net)</span>
<span id="cb32-4"><a href="#cb32-4"></a><span class="co"># _ = fig.colorbar(image)</span></span>
<span id="cb32-5"><a href="#cb32-5"></a>_ <span class="op">=</span> plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="index_files/figure-html/cell-27-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img src="index_files/figure-html/cell-27-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="training-som" class="level3">
<h3 class="anchored" data-anchor-id="training-som">Training SOM</h3>
<div id="86fa35b9" class="cell" data-execution_count="28">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1"></a><span class="kw">def</span> normalize(x: np.ndarray) <span class="op">-&gt;</span> np.ndarray:</span>
<span id="cb33-2"><a href="#cb33-2"></a>    <span class="cf">return</span> (x <span class="op">-</span> np.<span class="bu">min</span>(x)) <span class="op">/</span> (np.<span class="bu">max</span>(x) <span class="op">-</span> np.<span class="bu">min</span>(x))</span>
<span id="cb33-3"><a href="#cb33-3"></a></span>
<span id="cb33-4"><a href="#cb33-4"></a></span>
<span id="cb33-5"><a href="#cb33-5"></a>plt.rcParams[<span class="st">"figure.figsize"</span>] <span class="op">=</span> [</span>
<span id="cb33-6"><a href="#cb33-6"></a>    i <span class="op">/</span> <span class="fl">2.0</span> <span class="cf">for</span> i <span class="kw">in</span> plt.rcParamsDefault[<span class="st">"figure.figsize"</span>]</span>
<span id="cb33-7"><a href="#cb33-7"></a>]</span>
<span id="cb33-8"><a href="#cb33-8"></a>init_learning_rate <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb33-9"><a href="#cb33-9"></a><span class="co"># net = np.random.uniform(size=net.shape)</span></span>
<span id="cb33-10"><a href="#cb33-10"></a><span class="co"># net = (255 * ((net + 1.) / 2.))</span></span>
<span id="cb33-11"><a href="#cb33-11"></a>images <span class="op">=</span> []</span>
<span id="cb33-12"><a href="#cb33-12"></a><span class="cf">for</span> iteration <span class="kw">in</span> <span class="bu">range</span>(n_iterations):</span>
<span id="cb33-13"><a href="#cb33-13"></a>    <span class="co"># select a training example at random - shape of 1x3</span></span>
<span id="cb33-14"><a href="#cb33-14"></a>    t <span class="op">=</span> data[np.random.randint(<span class="dv">0</span>, m), :].reshape(np.array([n, <span class="dv">1</span>]))</span>
<span id="cb33-15"><a href="#cb33-15"></a>    <span class="co"># find its Best Matching Unit</span></span>
<span id="cb33-16"><a href="#cb33-16"></a>    bmu, bmu_idx <span class="op">=</span> find_bmu(t, net, n)  <span class="co"># Gives the row, column of the best neuron</span></span>
<span id="cb33-17"><a href="#cb33-17"></a>    <span class="co"># decay the SOM parameters</span></span>
<span id="cb33-18"><a href="#cb33-18"></a>    r <span class="op">=</span> decay_radius(init_radius, iteration, time_constant)</span>
<span id="cb33-19"><a href="#cb33-19"></a>    l <span class="op">=</span> decay_learning_rate(init_learning_rate, iteration, n_iterations)</span>
<span id="cb33-20"><a href="#cb33-20"></a>    <span class="co"># Update SOM weights</span></span>
<span id="cb33-21"><a href="#cb33-21"></a>    net <span class="op">=</span> update_weights(net, bmu_idx, r, l)</span>
<span id="cb33-22"><a href="#cb33-22"></a>    <span class="cf">if</span> iteration <span class="op">%</span> <span class="dv">50</span> <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb33-23"><a href="#cb33-23"></a>        <span class="bu">print</span>(<span class="ss">f"Iteration: </span><span class="sc">{</span>iteration<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb33-24"><a href="#cb33-24"></a>        ipydis.clear_output(<span class="st">"wait"</span>)</span>
<span id="cb33-25"><a href="#cb33-25"></a>        fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">4</span>, <span class="dv">4</span>))</span>
<span id="cb33-26"><a href="#cb33-26"></a>        net_img <span class="op">=</span> normalize(net)</span>
<span id="cb33-27"><a href="#cb33-27"></a>        _ <span class="op">=</span> fig.suptitle(<span class="ss">f"Iteration: </span><span class="sc">{</span>iteration<span class="sc">}</span><span class="ss">"</span>, y<span class="op">=</span><span class="fl">1.0</span>)</span>
<span id="cb33-28"><a href="#cb33-28"></a>        plt.tight_layout()</span>
<span id="cb33-29"><a href="#cb33-29"></a>        image <span class="op">=</span> ax.matshow(net_img, cmap<span class="op">=</span><span class="st">"rainbow"</span>)</span>
<span id="cb33-30"><a href="#cb33-30"></a>        <span class="cf">if</span> iteration <span class="op">%</span> <span class="dv">100</span> <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb33-31"><a href="#cb33-31"></a>            images.append(net_img)</span>
<span id="cb33-32"><a href="#cb33-32"></a>        _ <span class="op">=</span> plt.show()</span>
<span id="cb33-33"><a href="#cb33-33"></a>        time.sleep(<span class="fl">0.5</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="index_files/figure-html/cell-28-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img src="index_files/figure-html/cell-28-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="visualization-of-trained-colormap-som" class="level3">
<h3 class="anchored" data-anchor-id="visualization-of-trained-colormap-som">Visualization of trained colormap SOM</h3>
<div id="b061e041" class="cell" data-execution_count="29">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1"></a><span class="bu">len</span>(images)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display" data-execution_count="28">
<pre><code>25</code></pre>
</div>
</div>
<div id="4a0c02c1" class="cell" data-execution_count="30">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb36"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1"></a>fig, axes <span class="op">=</span> plt.subplots(</span>
<span id="cb36-2"><a href="#cb36-2"></a>    figsize<span class="op">=</span>(<span class="fl">2.0</span> <span class="op">*</span> DFIGSIZE[<span class="dv">0</span>], <span class="fl">1.5</span> <span class="op">*</span> DFIGSIZE[<span class="dv">1</span>]),</span>
<span id="cb36-3"><a href="#cb36-3"></a>    nrows<span class="op">=</span><span class="dv">5</span>,</span>
<span id="cb36-4"><a href="#cb36-4"></a>    ncols<span class="op">=</span><span class="dv">5</span>,</span>
<span id="cb36-5"><a href="#cb36-5"></a>    tight_layout<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb36-6"><a href="#cb36-6"></a>)</span>
<span id="cb36-7"><a href="#cb36-7"></a>fig.subplots_adjust(wspace<span class="op">=</span><span class="fl">0.1</span>)</span>
<span id="cb36-8"><a href="#cb36-8"></a>axes <span class="op">=</span> axes.flatten()</span>
<span id="cb36-9"><a href="#cb36-9"></a><span class="cf">for</span> idx, (img, ax) <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">zip</span>(images, axes)):</span>
<span id="cb36-10"><a href="#cb36-10"></a>    _ <span class="op">=</span> ax.matshow(img, cmap<span class="op">=</span><span class="st">"rainbow"</span>)</span>
<span id="cb36-11"><a href="#cb36-11"></a>    _ <span class="op">=</span> ax.set_title(<span class="ss">f"Iteration: </span><span class="sc">{</span><span class="dv">100</span> <span class="op">*</span> idx<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb36-12"><a href="#cb36-12"></a>    _ <span class="op">=</span> ax.set_xticks([])</span>
<span id="cb36-13"><a href="#cb36-13"></a>    _ <span class="op">=</span> ax.set_xticklabels([])</span>
<span id="cb36-14"><a href="#cb36-14"></a>    _ <span class="op">=</span> ax.set_yticks([])</span>
<span id="cb36-15"><a href="#cb36-15"></a>    _ <span class="op">=</span> ax.set_yticklabels([])</span>
<span id="cb36-16"><a href="#cb36-16"></a></span>
<span id="cb36-17"><a href="#cb36-17"></a>_ <span class="op">=</span> plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="index_files/figure-html/cell-30-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-17"><img src="index_files/figure-html/cell-30-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="1c7c570f" class="cell" data-execution_count="31">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> cm</span>
<span id="cb37-2"><a href="#cb37-2"></a></span>
<span id="cb37-3"><a href="#cb37-3"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">4</span>, <span class="dv">4</span>))</span>
<span id="cb37-4"><a href="#cb37-4"></a>image <span class="op">=</span> ax.imshow(normalize(net), cmap<span class="op">=</span><span class="st">"rainbow"</span>)</span>
<span id="cb37-5"><a href="#cb37-5"></a>_ <span class="op">=</span> ax.set_title(<span class="st">"Final SOM"</span>)</span>
<span id="cb37-6"><a href="#cb37-6"></a>_ <span class="op">=</span> plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="index_files/figure-html/cell-31-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-18"><img src="index_files/figure-html/cell-31-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="som-on-cancer-data" class="level3">
<h3 class="anchored" data-anchor-id="som-on-cancer-data">SOM on Cancer data</h3>
<div id="18716855" class="cell" data-execution_count="32">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb38"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1"></a><span class="co"># Load data</span></span>
<span id="cb38-2"><a href="#cb38-2"></a>data <span class="op">=</span> datasets.load_breast_cancer()</span>
<span id="cb38-3"><a href="#cb38-3"></a><span class="co"># Get features and target</span></span>
<span id="cb38-4"><a href="#cb38-4"></a>X <span class="op">=</span> data[<span class="st">"data"</span>]  <span class="co"># pd.DataFrame(data['data'], columns=data['feature_names'])</span></span>
<span id="cb38-5"><a href="#cb38-5"></a>Y <span class="op">=</span> data[<span class="st">"target"</span>]</span>
<span id="cb38-6"><a href="#cb38-6"></a><span class="bu">print</span>(X.shape, Y.shape)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>(569, 30) (569,)</code></pre>
</div>
</div>
<div id="2f475143" class="cell" data-execution_count="33">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1"></a>network_dimensions <span class="op">=</span> np.array([<span class="dv">10</span>, <span class="dv">10</span>])</span>
<span id="cb40-2"><a href="#cb40-2"></a>n_iterations <span class="op">=</span> <span class="dv">2000</span></span>
<span id="cb40-3"><a href="#cb40-3"></a>init_learning_rate <span class="op">=</span> <span class="fl">0.01</span></span>
<span id="cb40-4"><a href="#cb40-4"></a><span class="co"># establish size variables based on data</span></span>
<span id="cb40-5"><a href="#cb40-5"></a>n_points <span class="op">=</span> X.shape[<span class="dv">0</span>]  <span class="co"># number of points</span></span>
<span id="cb40-6"><a href="#cb40-6"></a>n_features <span class="op">=</span> X.shape[<span class="dv">1</span>]  <span class="co"># 30 features per point</span></span>
<span id="cb40-7"><a href="#cb40-7"></a></span>
<span id="cb40-8"><a href="#cb40-8"></a><span class="co"># weight matrix (i.e. the SOM) needs to be one n-dimensional vector for each neuron in the SOM</span></span>
<span id="cb40-9"><a href="#cb40-9"></a>net <span class="op">=</span> np.random.random((network_dimensions[<span class="dv">0</span>], network_dimensions[<span class="dv">1</span>], n_features))</span>
<span id="cb40-10"><a href="#cb40-10"></a></span>
<span id="cb40-11"><a href="#cb40-11"></a><span class="co"># initial neighbourhood radius</span></span>
<span id="cb40-12"><a href="#cb40-12"></a>init_radius <span class="op">=</span> <span class="bu">max</span>(network_dimensions[<span class="dv">0</span>], network_dimensions[<span class="dv">1</span>]) <span class="op">/</span> <span class="dv">2</span></span>
<span id="cb40-13"><a href="#cb40-13"></a><span class="co"># radius decay parameter</span></span>
<span id="cb40-14"><a href="#cb40-14"></a>time_constant <span class="op">=</span> n_iterations <span class="op">/</span> np.log(init_radius)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="1358b686" class="cell" data-execution_count="34">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb41"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1"></a><span class="co"># convert the network to something we can visualize</span></span>
<span id="cb41-2"><a href="#cb41-2"></a>net_vis <span class="op">=</span> np.zeros(</span>
<span id="cb41-3"><a href="#cb41-3"></a>    shape<span class="op">=</span>(net.shape[<span class="dv">0</span>], net.shape[<span class="dv">1</span>], <span class="dv">3</span>),</span>
<span id="cb41-4"><a href="#cb41-4"></a>    dtype<span class="op">=</span>np.float32,</span>
<span id="cb41-5"><a href="#cb41-5"></a>)  <span class="co"># Array for SOM color map visualization</span></span>
<span id="cb41-6"><a href="#cb41-6"></a></span>
<span id="cb41-7"><a href="#cb41-7"></a><span class="cf">for</span> sample <span class="kw">in</span> <span class="bu">range</span>(n_points):</span>
<span id="cb41-8"><a href="#cb41-8"></a>    t <span class="op">=</span> X[sample, :].reshape(np.array([n_features, <span class="dv">1</span>]))</span>
<span id="cb41-9"><a href="#cb41-9"></a>    <span class="co"># find its Best Matching Unit for this data point</span></span>
<span id="cb41-10"><a href="#cb41-10"></a>    bmu, bmu_idx <span class="op">=</span> find_bmu(t, net, n_features)</span>
<span id="cb41-11"><a href="#cb41-11"></a>    <span class="co"># set that unit to the label of this data point</span></span>
<span id="cb41-12"><a href="#cb41-12"></a>    net_vis[bmu_idx[<span class="dv">0</span>], bmu_idx[<span class="dv">1</span>], <span class="dv">0</span>] <span class="op">=</span> Y[sample]  <span class="co"># Red if benign</span></span>
<span id="cb41-13"><a href="#cb41-13"></a></span>
<span id="cb41-14"><a href="#cb41-14"></a></span>
<span id="cb41-15"><a href="#cb41-15"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb41-16"><a href="#cb41-16"></a>im <span class="op">=</span> ax.matshow(normalize(net_vis))</span>
<span id="cb41-17"><a href="#cb41-17"></a><span class="co"># _ = plt.colorbar(im1, ax=ax1)</span></span>
<span id="cb41-18"><a href="#cb41-18"></a>plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="index_files/figure-html/cell-34-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-19"><img src="index_files/figure-html/cell-34-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="training-som-on-cancer-data" class="level3">
<h3 class="anchored" data-anchor-id="training-som-on-cancer-data">Training SOM on Cancer data</h3>
<div id="0875e7bd" class="cell" data-execution_count="35">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1"></a><span class="cf">for</span> iteration <span class="kw">in</span> <span class="bu">range</span>(n_iterations):</span>
<span id="cb42-2"><a href="#cb42-2"></a>    <span class="co"># select a training example at random - shape of 1x3</span></span>
<span id="cb42-3"><a href="#cb42-3"></a>    t <span class="op">=</span> X[np.random.randint(<span class="dv">0</span>, n_points), :].reshape(np.array([n_features, <span class="dv">1</span>]))</span>
<span id="cb42-4"><a href="#cb42-4"></a>    <span class="co"># find its Best Matching Unit</span></span>
<span id="cb42-5"><a href="#cb42-5"></a>    bmu, bmu_idx <span class="op">=</span> find_bmu(t, net, n_features)</span>
<span id="cb42-6"><a href="#cb42-6"></a>    <span class="co"># decay the SOM parameters</span></span>
<span id="cb42-7"><a href="#cb42-7"></a>    r <span class="op">=</span> decay_radius(init_radius, iteration, time_constant)</span>
<span id="cb42-8"><a href="#cb42-8"></a>    l <span class="op">=</span> decay_learning_rate(init_learning_rate, iteration, n_iterations)</span>
<span id="cb42-9"><a href="#cb42-9"></a>    <span class="co"># Update SOM weights</span></span>
<span id="cb42-10"><a href="#cb42-10"></a>    net <span class="op">=</span> update_weights(net, bmu_idx, r, l)</span>
<span id="cb42-11"><a href="#cb42-11"></a></span>
<span id="cb42-12"><a href="#cb42-12"></a>    <span class="cf">if</span> iteration <span class="op">%</span> <span class="dv">50</span> <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb42-13"><a href="#cb42-13"></a>        ipydis.clear_output(<span class="st">"wait"</span>)</span>
<span id="cb42-14"><a href="#cb42-14"></a>        net_vis <span class="op">=</span> np.zeros(</span>
<span id="cb42-15"><a href="#cb42-15"></a>            shape<span class="op">=</span>(np.shape(net)[<span class="dv">0</span>], np.shape(net)[<span class="dv">1</span>], <span class="dv">3</span>), dtype<span class="op">=</span><span class="st">"double"</span></span>
<span id="cb42-16"><a href="#cb42-16"></a>        )  <span class="co"># Array for SOM color map visualization</span></span>
<span id="cb42-17"><a href="#cb42-17"></a>        <span class="cf">for</span> sample <span class="kw">in</span> <span class="bu">range</span>(n_points):</span>
<span id="cb42-18"><a href="#cb42-18"></a>            t <span class="op">=</span> X[sample, :].reshape(np.array([n_features, <span class="dv">1</span>]))</span>
<span id="cb42-19"><a href="#cb42-19"></a>            <span class="co"># find its Best Matching Unit</span></span>
<span id="cb42-20"><a href="#cb42-20"></a>            bmu, bmu_idx <span class="op">=</span> find_bmu(t, net, n_features)</span>
<span id="cb42-21"><a href="#cb42-21"></a>            net_vis[bmu_idx[<span class="dv">0</span>], bmu_idx[<span class="dv">1</span>], <span class="dv">0</span>] <span class="op">=</span> Y[sample]  <span class="co"># Red if benign</span></span>
<span id="cb42-22"><a href="#cb42-22"></a></span>
<span id="cb42-23"><a href="#cb42-23"></a>        fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">3</span>, <span class="dv">3</span>))</span>
<span id="cb42-24"><a href="#cb42-24"></a>        _ <span class="op">=</span> ax.set_title(<span class="ss">f"Iteration: </span><span class="sc">{</span>iteration<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb42-25"><a href="#cb42-25"></a>        _ <span class="op">=</span> ax.imshow(normalize(net_vis), cmap<span class="op">=</span><span class="st">"rainbow"</span>)</span>
<span id="cb42-26"><a href="#cb42-26"></a>        _ <span class="op">=</span> plt.show()</span>
<span id="cb42-27"><a href="#cb42-27"></a>        time.sleep(<span class="fl">0.25</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="index_files/figure-html/cell-35-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-20"><img src="index_files/figure-html/cell-35-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="visualization-of-trained-som" class="level3">
<h3 class="anchored" data-anchor-id="visualization-of-trained-som">Visualization of trained SOM</h3>
<div id="a411943e" class="cell" data-execution_count="36">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1"></a>net_vis <span class="op">=</span> np.zeros(</span>
<span id="cb43-2"><a href="#cb43-2"></a>    shape<span class="op">=</span>(net.shape[<span class="dv">0</span>], net.shape[<span class="dv">1</span>], <span class="dv">3</span>), dtype<span class="op">=</span><span class="st">"double"</span></span>
<span id="cb43-3"><a href="#cb43-3"></a>)  <span class="co"># Array for SOM color map visualization</span></span>
<span id="cb43-4"><a href="#cb43-4"></a></span>
<span id="cb43-5"><a href="#cb43-5"></a><span class="cf">for</span> sample <span class="kw">in</span> <span class="bu">range</span>(n_points):</span>
<span id="cb43-6"><a href="#cb43-6"></a>    t <span class="op">=</span> X[sample, :].reshape(np.array([n_features, <span class="dv">1</span>]))</span>
<span id="cb43-7"><a href="#cb43-7"></a>    <span class="co"># find its Best Matching Unit</span></span>
<span id="cb43-8"><a href="#cb43-8"></a>    bmu, bmu_idx <span class="op">=</span> find_bmu(t, net, n_features)</span>
<span id="cb43-9"><a href="#cb43-9"></a>    net_vis[bmu_idx[<span class="dv">0</span>], bmu_idx[<span class="dv">1</span>], <span class="dv">0</span>] <span class="op">=</span> Y[sample]  <span class="co"># Red if benign</span></span>
<span id="cb43-10"><a href="#cb43-10"></a></span>
<span id="cb43-11"><a href="#cb43-11"></a>_ <span class="op">=</span> plt.imshow(normalize(net_vis))</span>
<span id="cb43-12"><a href="#cb43-12"></a>_ <span class="op">=</span> plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="index_files/figure-html/cell-36-output-1.svg" class="lightbox" data-gallery="quarto-lightbox-gallery-21"><img src="index_files/figure-html/cell-36-output-1.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="keep-learning" class="level2">
<h2 class="anchored" data-anchor-id="keep-learning">Keep learning</h2>
<ol type="1">
<li>&lt;lagunita.stanford.edu/courses/HumanitiesSciences/StatLearning/Winter2016/course/&gt;</li>
<li>&lt;www.coursera.org/learn/ml-clustering-and-retrieval/&gt;</li>
<li>&lt;www.coursera.org/learn/machine-learning/home/week/8&gt;</li>
</ol>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" id="quarto-citation"><h2 class="anchored quarto-appendix-heading">Citation</h2><div><div class="quarto-appendix-secondary-label">BibTeX citation:</div><pre class="sourceCode code-with-copy quarto-appendix-bibtex"><code class="sourceCode bibtex">@online{foreman2025,
  author = {Foreman, Sam and Foreman, Sam and Zheng, Huihuo},
  title = {Statistical {Learning}},
  date = {2025-07-25},
  url = {https://saforem2.github.io/hpc-bootcamp-2025/00-intro-AI-HPC/7-statistical-learning/},
  langid = {en}
}
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre><div class="quarto-appendix-secondary-label">For attribution, please cite this work as:</div><div id="ref-foreman2025" class="csl-entry quarto-appendix-citeas" role="listitem">
Foreman, Sam, Sam Foreman, and Huihuo Zheng. 2025. <span>“Statistical
Learning.”</span> July 25, 2025. <a href="https://saforem2.github.io/hpc-bootcamp-2025/00-intro-AI-HPC/7-statistical-learning/">https://saforem2.github.io/hpc-bootcamp-2025/00-intro-AI-HPC/7-statistical-learning/</a>.
</div></div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    // Ensure there is a toggle, if there isn't float one in the top right
    if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
      const a = window.document.createElement('a');
      a.classList.add('top-right');
      a.classList.add('quarto-color-scheme-toggle');
      a.href = "";
      a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
      const i = window.document.createElement("i");
      i.classList.add('bi');
      a.appendChild(i);
      window.document.body.appendChild(a);
    }
    setColorSchemeToggle(hasAlternateSentinel())
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/saforem2\.github\.io\/hpc-bootcamp-2025\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
            // target, if specified
            link.setAttribute("target", "_blank");
            if (link.getAttribute("rel") === null) {
              link.setAttribute("rel", "noopener");
            }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<input type="hidden" id="giscus-base-theme" value="dark">
<input type="hidden" id="giscus-alt-theme" value="light">
<script>
  function loadGiscus() {
    // Function to get the theme based on body class
    const getTheme = () => {
      let baseTheme = document.getElementById('giscus-base-theme').value;
      let altTheme = document.getElementById('giscus-alt-theme').value;
      if (authorPrefersDark) {
          [baseTheme, altTheme] = [altTheme, baseTheme];
      }
      return document.body.classList.contains('quarto-dark') ? altTheme : baseTheme;
    };
    const script = document.createElement("script");
    script.src = "https://giscus.app/client.js";
    script.async = true;
    script.dataset.repo = "saforem2/intro-hpc-bootcamp-2025";
    script.dataset.repoId = "R_kgDOPQ7DhQ";
    script.dataset.category = "General";
    script.dataset.categoryId = "DIC_kwDOPQ7Dhc4CtXSR";
    script.dataset.mapping = "title";
    script.dataset.reactionsEnabled = "1";
    script.dataset.emitMetadata = "0";
    script.dataset.inputPosition = "top";
    script.dataset.theme = getTheme();
    script.dataset.lang = "en";
    script.crossOrigin = "anonymous";
    // Append the script to the desired div instead of at the end of the body
    document.getElementById("quarto-content").appendChild(script);
  }
  loadGiscus();
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../00-intro-AI-HPC/6-linear-regression/index.html" class="pagination-link" aria-label="[6] Linear Regression">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">[6] Linear Regression</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../00-intro-AI-HPC/8-clustering/index.html" class="pagination-link" aria-label="[8] Clustering">
        <span class="nav-page-text">[8] Clustering</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb44" data-shortcodes="false"><pre class="sourceCode numberSource markdown number-lines code-with-copy"><code class="sourceCode markdown"><span id="cb44-1"><a href="#cb44-1"></a><span class="co">---</span></span>
<span id="cb44-2"><a href="#cb44-2"></a><span class="an">title:</span><span class="co"> "Statistical Learning"</span></span>
<span id="cb44-3"><a href="#cb44-3"></a><span class="an">description:</span><span class="co"> "Statistical learning is a framework for understanding and modeling complex data patterns, combining statistical methods with machine learning algorithms to build predictive models."</span></span>
<span id="cb44-4"><a href="#cb44-4"></a><span class="an">date:</span><span class="co"> 2025-07-25</span></span>
<span id="cb44-5"><a href="#cb44-5"></a><span class="an">date-modified:</span><span class="co"> last-modified</span></span>
<span id="cb44-6"><a href="#cb44-6"></a><span class="an">execute:</span></span>
<span id="cb44-7"><a href="#cb44-7"></a><span class="co">  cache: true</span></span>
<span id="cb44-8"><a href="#cb44-8"></a><span class="co">  freeze: auto</span></span>
<span id="cb44-9"><a href="#cb44-9"></a><span class="an">format:</span></span>
<span id="cb44-10"><a href="#cb44-10"></a><span class="co">  ipynb:</span></span>
<span id="cb44-11"><a href="#cb44-11"></a><span class="co">    default-image-extension: svg</span></span>
<span id="cb44-12"><a href="#cb44-12"></a><span class="co">  html: default</span></span>
<span id="cb44-13"><a href="#cb44-13"></a><span class="co">  gfm:</span></span>
<span id="cb44-14"><a href="#cb44-14"></a><span class="co">    toc: true</span></span>
<span id="cb44-15"><a href="#cb44-15"></a><span class="an">author:</span></span>
<span id="cb44-16"><a href="#cb44-16"></a><span class="co">  - id: sf</span></span>
<span id="cb44-17"><a href="#cb44-17"></a><span class="co">    name: Sam Foreman</span></span>
<span id="cb44-18"><a href="#cb44-18"></a><span class="co">    orcid: 0000-0002-9981-0876</span></span>
<span id="cb44-19"><a href="#cb44-19"></a><span class="co">    email: foremans@anl.gov</span></span>
<span id="cb44-20"><a href="#cb44-20"></a><span class="co">    affiliation:</span></span>
<span id="cb44-21"><a href="#cb44-21"></a><span class="co">      - name: '[ANL](https://www.anl.gov/)'</span></span>
<span id="cb44-22"><a href="#cb44-22"></a><span class="co">        city: Lemont</span></span>
<span id="cb44-23"><a href="#cb44-23"></a><span class="co">        state: IL</span></span>
<span id="cb44-24"><a href="#cb44-24"></a><span class="co">        url: https://alcf.anl.gov/about/people/sam-foreman</span></span>
<span id="cb44-25"><a href="#cb44-25"></a><span class="co">  - id: hz</span></span>
<span id="cb44-26"><a href="#cb44-26"></a><span class="co">    name: Huihuo Zheng</span></span>
<span id="cb44-27"><a href="#cb44-27"></a><span class="co">    # orcid: 0000-0002-9981-0876</span></span>
<span id="cb44-28"><a href="#cb44-28"></a><span class="co">    email: huihuo.zheng@anl.gov</span></span>
<span id="cb44-29"><a href="#cb44-29"></a><span class="co">    affiliation:</span></span>
<span id="cb44-30"><a href="#cb44-30"></a><span class="co">      - name: '[ANL](https://www.anl.gov/)'</span></span>
<span id="cb44-31"><a href="#cb44-31"></a><span class="co">        city: Lemont</span></span>
<span id="cb44-32"><a href="#cb44-32"></a><span class="co">        state: IL</span></span>
<span id="cb44-33"><a href="#cb44-33"></a><span class="co">        url: https://alcf.anl.gov/about/people/huihuo-zheng</span></span>
<span id="cb44-34"><a href="#cb44-34"></a><span class="co">---</span></span>
<span id="cb44-35"><a href="#cb44-35"></a></span>
<span id="cb44-36"><a href="#cb44-36"></a><span class="co">[</span><span class="al">![](https://colab.research.google.com/assets/colab-badge.svg)</span><span class="co">](https://colab.research.google.com/github/saforem2/intro-hpc-bootcamp-2025/blob/main/docs/00-intro-AI-HPC/7-statistical-learning/index.ipynb)</span> <span class="co">[</span><span class="al">![](https://img.shields.io/badge/-View%20on%20GitHub-333333?style=flat&amp;logo=github&amp;labelColor=gray.png)</span><span class="co">](https://github.com/saforem2/intro-hpc-bootcamp-2025/blob/main/docs/00-intro-AI-HPC/7-statistical-learning/README.md)</span></span>
<span id="cb44-37"><a href="#cb44-37"></a></span>
<span id="cb44-40"><a href="#cb44-40"></a><span class="in">```{python}</span></span>
<span id="cb44-41"><a href="#cb44-41"></a><span class="op">%</span>load_ext autoreload</span>
<span id="cb44-42"><a href="#cb44-42"></a><span class="op">%</span>autoreload <span class="dv">2</span></span>
<span id="cb44-43"><a href="#cb44-43"></a><span class="im">import</span> matplotlib_inline.backend_inline</span>
<span id="cb44-44"><a href="#cb44-44"></a>matplotlib_inline.backend_inline.set_matplotlib_formats(<span class="st">'retina'</span>, <span class="st">'svg'</span>, <span class="st">'png'</span>)</span>
<span id="cb44-45"><a href="#cb44-45"></a><span class="im">import</span> os</span>
<span id="cb44-46"><a href="#cb44-46"></a>os.environ[<span class="st">"TRUECOLOR"</span>] <span class="op">=</span> <span class="st">"1"</span></span>
<span id="cb44-47"><a href="#cb44-47"></a><span class="im">import</span> matplotlib <span class="im">as</span> mpl</span>
<span id="cb44-48"><a href="#cb44-48"></a><span class="co"># mpl.rcParams['figure.dpi'] = 400</span></span>
<span id="cb44-49"><a href="#cb44-49"></a><span class="in">```</span></span>
<span id="cb44-50"><a href="#cb44-50"></a></span>
<span id="cb44-51"><a href="#cb44-51"></a><span class="fu">## Learning to Cluster Data</span></span>
<span id="cb44-52"><a href="#cb44-52"></a></span>
<span id="cb44-53"><a href="#cb44-53"></a>If we have data with distinct groupings, the objective is to devise a method</span>
<span id="cb44-54"><a href="#cb44-54"></a>for labeling our data by its group in an automated way.</span>
<span id="cb44-55"><a href="#cb44-55"></a></span>
<span id="cb44-56"><a href="#cb44-56"></a>We will demonstrate this, first, on a toy dataset that we _design_ to have a</span>
<span id="cb44-57"><a href="#cb44-57"></a>lower inherent dimensionality, then we move to a higher dimensional dataset.</span>
<span id="cb44-58"><a href="#cb44-58"></a></span>
<span id="cb44-59"><a href="#cb44-59"></a><span class="co">[</span><span class="al">![](https://colab.research.google.com/assets/colab-badge.svg)</span><span class="co">](https://colab.research.google.com/github/saforem2/intro-hpc-bootcamp-2025/blob/main/docs/00-intro-AI-HPC/7-statistical-learning/index.ipynb)</span></span>
<span id="cb44-60"><a href="#cb44-60"></a></span>
<span id="cb44-61"><a href="#cb44-61"></a><span class="fu">### Toy Dataset (2D Blobs)</span></span>
<span id="cb44-62"><a href="#cb44-62"></a></span>
<span id="cb44-63"><a href="#cb44-63"></a>Define 3 blobs on a 2D plane:</span>
<span id="cb44-64"><a href="#cb44-64"></a></span>
<span id="cb44-67"><a href="#cb44-67"></a><span class="in">```{python}</span></span>
<span id="cb44-68"><a href="#cb44-68"></a><span class="im">import</span> time</span>
<span id="cb44-69"><a href="#cb44-69"></a><span class="im">import</span> os</span>
<span id="cb44-70"><a href="#cb44-70"></a><span class="im">import</span> sklearn</span>
<span id="cb44-71"><a href="#cb44-71"></a><span class="im">from</span> sklearn <span class="im">import</span> datasets</span>
<span id="cb44-72"><a href="#cb44-72"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb44-73"><a href="#cb44-73"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb44-74"><a href="#cb44-74"></a></span>
<span id="cb44-75"><a href="#cb44-75"></a><span class="im">from</span> sklearn.preprocessing <span class="im">import</span> (</span>
<span id="cb44-76"><a href="#cb44-76"></a>    StandardScaler,</span>
<span id="cb44-77"><a href="#cb44-77"></a>    MinMaxScaler,</span>
<span id="cb44-78"><a href="#cb44-78"></a>    RobustScaler</span>
<span id="cb44-79"><a href="#cb44-79"></a>)</span>
<span id="cb44-80"><a href="#cb44-80"></a><span class="im">from</span> sklearn.cluster <span class="im">import</span> KMeans</span>
<span id="cb44-81"><a href="#cb44-81"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb44-82"><a href="#cb44-82"></a><span class="im">import</span> ambivalent</span>
<span id="cb44-83"><a href="#cb44-83"></a>plt.style.use(ambivalent.STYLES[<span class="st">"ambivalent"</span>])</span>
<span id="cb44-84"><a href="#cb44-84"></a></span>
<span id="cb44-85"><a href="#cb44-85"></a><span class="im">from</span> matplotlib.colors <span class="im">import</span> ListedColormap</span>
<span id="cb44-86"><a href="#cb44-86"></a><span class="im">from</span> bootcamp.plots <span class="im">import</span> scatter, COLORS</span>
<span id="cb44-87"><a href="#cb44-87"></a></span>
<span id="cb44-88"><a href="#cb44-88"></a><span class="co"># set random seed for reproducibility</span></span>
<span id="cb44-89"><a href="#cb44-89"></a>SEED <span class="op">=</span> <span class="dv">42</span></span>
<span id="cb44-90"><a href="#cb44-90"></a>DFIGSIZE <span class="op">=</span> plt.rcParamsDefault[<span class="st">'figure.figsize'</span>]</span>
<span id="cb44-91"><a href="#cb44-91"></a><span class="in">```</span></span>
<span id="cb44-92"><a href="#cb44-92"></a></span>
<span id="cb44-95"><a href="#cb44-95"></a><span class="in">```{python}</span></span>
<span id="cb44-96"><a href="#cb44-96"></a>n_samples <span class="op">=</span> <span class="dv">1000</span> <span class="co"># 300 2D data points</span></span>
<span id="cb44-97"><a href="#cb44-97"></a>n_features <span class="op">=</span> <span class="dv">2</span>  <span class="co"># 2D data</span></span>
<span id="cb44-98"><a href="#cb44-98"></a>n_clusters <span class="op">=</span> <span class="dv">4</span>  <span class="co"># 3 unique blobs</span></span>
<span id="cb44-99"><a href="#cb44-99"></a></span>
<span id="cb44-100"><a href="#cb44-100"></a><span class="co"># https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_blobs.html</span></span>
<span id="cb44-101"><a href="#cb44-101"></a><span class="co"># -- Returns --------------------------------------------------------</span></span>
<span id="cb44-102"><a href="#cb44-102"></a><span class="co"># x (n_samples, n_features): The generated samples.</span></span>
<span id="cb44-103"><a href="#cb44-103"></a><span class="co"># y (n_samples,): Int labels for cluster membership of each sample.</span></span>
<span id="cb44-104"><a href="#cb44-104"></a><span class="co"># -------------------------------------------------------------------</span></span>
<span id="cb44-105"><a href="#cb44-105"></a>cmap <span class="op">=</span> ListedColormap(<span class="bu">list</span>(COLORS.values()))</span>
<span id="cb44-106"><a href="#cb44-106"></a>x, y <span class="op">=</span> datasets.make_blobs(</span>
<span id="cb44-107"><a href="#cb44-107"></a>    n_samples<span class="op">=</span>n_samples,</span>
<span id="cb44-108"><a href="#cb44-108"></a>    n_features<span class="op">=</span>n_features,</span>
<span id="cb44-109"><a href="#cb44-109"></a>    centers<span class="op">=</span>n_clusters,</span>
<span id="cb44-110"><a href="#cb44-110"></a>    random_state<span class="op">=</span><span class="dv">42</span></span>
<span id="cb44-111"><a href="#cb44-111"></a>)</span>
<span id="cb44-112"><a href="#cb44-112"></a>scatter_kwargs <span class="op">=</span> {</span>
<span id="cb44-113"><a href="#cb44-113"></a>    <span class="st">'xlabel'</span>: <span class="st">'x0'</span>,</span>
<span id="cb44-114"><a href="#cb44-114"></a>    <span class="st">'ylabel'</span>: <span class="st">'x1'</span>,</span>
<span id="cb44-115"><a href="#cb44-115"></a>    <span class="st">'cmap'</span>: cmap,</span>
<span id="cb44-116"><a href="#cb44-116"></a>    <span class="st">'plot_kwargs'</span>: {</span>
<span id="cb44-117"><a href="#cb44-117"></a>        <span class="st">'alpha'</span>: <span class="fl">0.8</span>,</span>
<span id="cb44-118"><a href="#cb44-118"></a>        <span class="st">'edgecolor'</span>: <span class="st">'#222'</span>,</span>
<span id="cb44-119"><a href="#cb44-119"></a>    }</span>
<span id="cb44-120"><a href="#cb44-120"></a>}</span>
<span id="cb44-121"><a href="#cb44-121"></a>_ <span class="op">=</span> scatter(x, y, <span class="op">**</span>scatter_kwargs)</span>
<span id="cb44-122"><a href="#cb44-122"></a><span class="in">```</span></span>
<span id="cb44-123"><a href="#cb44-123"></a></span>
<span id="cb44-126"><a href="#cb44-126"></a><span class="in">```{python}</span></span>
<span id="cb44-127"><a href="#cb44-127"></a><span class="co"># Normalize features</span></span>
<span id="cb44-128"><a href="#cb44-128"></a><span class="co"># https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html</span></span>
<span id="cb44-129"><a href="#cb44-129"></a>x_sc <span class="op">=</span> StandardScaler().fit_transform(x)</span>
<span id="cb44-130"><a href="#cb44-130"></a><span class="in">```</span></span>
<span id="cb44-131"><a href="#cb44-131"></a></span>
<span id="cb44-134"><a href="#cb44-134"></a><span class="in">```{python}</span></span>
<span id="cb44-135"><a href="#cb44-135"></a>fig, (ax1, ax2) <span class="op">=</span> plt.subplots(</span>
<span id="cb44-136"><a href="#cb44-136"></a>    figsize<span class="op">=</span>(<span class="fl">1.5</span> <span class="op">*</span> DFIGSIZE[<span class="dv">0</span>], <span class="fl">0.8</span> <span class="op">*</span> DFIGSIZE[<span class="dv">1</span>]),</span>
<span id="cb44-137"><a href="#cb44-137"></a>    ncols<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb44-138"><a href="#cb44-138"></a>    subplot_kw<span class="op">=</span>{</span>
<span id="cb44-139"><a href="#cb44-139"></a>        <span class="st">'aspect'</span>: <span class="st">'equal'</span>,</span>
<span id="cb44-140"><a href="#cb44-140"></a>    },</span>
<span id="cb44-141"><a href="#cb44-141"></a>    </span>
<span id="cb44-142"><a href="#cb44-142"></a>)</span>
<span id="cb44-143"><a href="#cb44-143"></a>fig, ax1 <span class="op">=</span> scatter(</span>
<span id="cb44-144"><a href="#cb44-144"></a>    x,</span>
<span id="cb44-145"><a href="#cb44-145"></a>    y,</span>
<span id="cb44-146"><a href="#cb44-146"></a>    fig<span class="op">=</span>fig,</span>
<span id="cb44-147"><a href="#cb44-147"></a>    ax<span class="op">=</span>ax1,</span>
<span id="cb44-148"><a href="#cb44-148"></a>    title<span class="op">=</span><span class="st">'Original'</span>,</span>
<span id="cb44-149"><a href="#cb44-149"></a>    <span class="op">**</span>scatter_kwargs</span>
<span id="cb44-150"><a href="#cb44-150"></a>)</span>
<span id="cb44-151"><a href="#cb44-151"></a>fig.subplots_adjust(wspace<span class="op">=</span><span class="fl">0.2</span>)</span>
<span id="cb44-152"><a href="#cb44-152"></a>fig, ax2 <span class="op">=</span> scatter(</span>
<span id="cb44-153"><a href="#cb44-153"></a>    x_sc,</span>
<span id="cb44-154"><a href="#cb44-154"></a>    y,</span>
<span id="cb44-155"><a href="#cb44-155"></a>    fig<span class="op">=</span>fig,</span>
<span id="cb44-156"><a href="#cb44-156"></a>    ax<span class="op">=</span>ax2,</span>
<span id="cb44-157"><a href="#cb44-157"></a>    title<span class="op">=</span><span class="st">'Normalized'</span>,</span>
<span id="cb44-158"><a href="#cb44-158"></a>    <span class="op">**</span>scatter_kwargs</span>
<span id="cb44-159"><a href="#cb44-159"></a>)</span>
<span id="cb44-160"><a href="#cb44-160"></a><span class="in">```</span></span>
<span id="cb44-161"><a href="#cb44-161"></a></span>
<span id="cb44-162"><a href="#cb44-162"></a><span class="fu">## K-means Clustering</span></span>
<span id="cb44-163"><a href="#cb44-163"></a></span>
<span id="cb44-164"><a href="#cb44-164"></a>K-means clustering aims to partition n observations into k clusters in which</span>
<span id="cb44-165"><a href="#cb44-165"></a>each observation belongs to the cluster with the nearest mean, serving as a</span>
<span id="cb44-166"><a href="#cb44-166"></a>prototype of the cluster.</span>
<span id="cb44-167"><a href="#cb44-167"></a>This results in a partitioning of the data space into Voronoi cells.</span>
<span id="cb44-168"><a href="#cb44-168"></a></span>
<span id="cb44-169"><a href="#cb44-169"></a><span class="ss">1. </span>Pick random starting centroids from data</span>
<span id="cb44-170"><a href="#cb44-170"></a><span class="ss">   - </span>$k$ initial "means" (in this case, $k=3$) are randomly generated within</span>
<span id="cb44-171"><a href="#cb44-171"></a>     the data domain (shown in color)</span>
<span id="cb44-172"><a href="#cb44-172"></a><span class="ss">2. </span>Calculate distance to each centroid</span>
<span id="cb44-173"><a href="#cb44-173"></a><span class="ss">   - </span>$k$ clusters are created by associating  every observation with the</span>
<span id="cb44-174"><a href="#cb44-174"></a>     nearest mean.</span>
<span id="cb44-175"><a href="#cb44-175"></a>     The partitions here represent the Voroni diagram generated by the means</span>
<span id="cb44-176"><a href="#cb44-176"></a><span class="ss">   - </span>Find nearest cluster for each point</span>
<span id="cb44-177"><a href="#cb44-177"></a><span class="ss">3. </span>Calculate new centroids</span>
<span id="cb44-178"><a href="#cb44-178"></a><span class="ss">   - </span>The **centroid** of each of the $k$ clusters becomes the new mean</span>
<span id="cb44-179"><a href="#cb44-179"></a><span class="ss">4. </span>Steps 2 and 3 are repeated until convergence has been reached</span>
<span id="cb44-180"><a href="#cb44-180"></a></span>
<span id="cb44-181"><a href="#cb44-181"></a></span>
<span id="cb44-182"><a href="#cb44-182"></a>::: {#fig-kmeans}</span>
<span id="cb44-183"><a href="#cb44-183"></a></span>
<span id="cb44-184"><a href="#cb44-184"></a>::: {.flex-container}</span>
<span id="cb44-185"><a href="#cb44-185"></a><span class="al">![](./assets/k-means0.svg)</span></span>
<span id="cb44-186"><a href="#cb44-186"></a></span>
<span id="cb44-187"><a href="#cb44-187"></a><span class="al">![](./assets/k-means1.svg)</span></span>
<span id="cb44-188"><a href="#cb44-188"></a></span>
<span id="cb44-189"><a href="#cb44-189"></a><span class="al">![](./assets/k-means2.svg)</span></span>
<span id="cb44-190"><a href="#cb44-190"></a></span>
<span id="cb44-191"><a href="#cb44-191"></a><span class="al">![](./assets/k-means3.svg)</span></span>
<span id="cb44-192"><a href="#cb44-192"></a>:::</span>
<span id="cb44-193"><a href="#cb44-193"></a></span>
<span id="cb44-194"><a href="#cb44-194"></a>Illustration of the K-means algorithm steps</span>
<span id="cb44-195"><a href="#cb44-195"></a>:::</span>
<span id="cb44-196"><a href="#cb44-196"></a></span>
<span id="cb44-197"><a href="#cb44-197"></a><span class="fu">### Step 1: Pick random starting centroids from data</span></span>
<span id="cb44-198"><a href="#cb44-198"></a></span>
<span id="cb44-199"><a href="#cb44-199"></a><span class="ss">- </span>$k$ initial "means" (in this case, $k=3$) are randomly generated within</span>
<span id="cb44-200"><a href="#cb44-200"></a>  the data domain (shown in color)</span>
<span id="cb44-201"><a href="#cb44-201"></a></span>
<span id="cb44-202"><a href="#cb44-202"></a>:::{#fig-kmeans-step1 style="width:40%"}</span>
<span id="cb44-203"><a href="#cb44-203"></a></span>
<span id="cb44-204"><a href="#cb44-204"></a><span class="al">![](./assets/k-means0.svg)</span></span>
<span id="cb44-205"><a href="#cb44-205"></a></span>
<span id="cb44-206"><a href="#cb44-206"></a>:::</span>
<span id="cb44-207"><a href="#cb44-207"></a></span>
<span id="cb44-210"><a href="#cb44-210"></a><span class="in">```{python}</span></span>
<span id="cb44-211"><a href="#cb44-211"></a><span class="kw">def</span> initialize_centroids(</span>
<span id="cb44-212"><a href="#cb44-212"></a>        x: np.ndarray,</span>
<span id="cb44-213"><a href="#cb44-213"></a>        n_clusters: <span class="bu">int</span>,</span>
<span id="cb44-214"><a href="#cb44-214"></a>        seed: <span class="bu">int</span> <span class="op">=</span> <span class="dv">123</span></span>
<span id="cb44-215"><a href="#cb44-215"></a>) <span class="op">-&gt;</span> np.ndarray:</span>
<span id="cb44-216"><a href="#cb44-216"></a>    <span class="co">"""Initialize centroids.</span></span>
<span id="cb44-217"><a href="#cb44-217"></a><span class="co">    </span></span>
<span id="cb44-218"><a href="#cb44-218"></a><span class="co">    Inputs:</span></span>
<span id="cb44-219"><a href="#cb44-219"></a><span class="co">      - x (np.ndarray): Data of shape (num_points, num_features)</span></span>
<span id="cb44-220"><a href="#cb44-220"></a><span class="co">      - n_clusters (int): Number of clusters to use</span></span>
<span id="cb44-221"><a href="#cb44-221"></a><span class="co">      - seed (int): Random seed.</span></span>
<span id="cb44-222"><a href="#cb44-222"></a><span class="co">        </span></span>
<span id="cb44-223"><a href="#cb44-223"></a><span class="co">    Outputs:</span></span>
<span id="cb44-224"><a href="#cb44-224"></a><span class="co">      - centroids (np.ndarray): Randomly chosen from the data</span></span>
<span id="cb44-225"><a href="#cb44-225"></a><span class="co">        with shape: (num_clusters, num_features).</span></span>
<span id="cb44-226"><a href="#cb44-226"></a><span class="co">    """</span></span>
<span id="cb44-227"><a href="#cb44-227"></a>    np.random.RandomState(seed)</span>
<span id="cb44-228"><a href="#cb44-228"></a>    <span class="co"># 1. Randomly permute data points</span></span>
<span id="cb44-229"><a href="#cb44-229"></a>    <span class="co"># 2. From this, pick the first `n_clusters` indices</span></span>
<span id="cb44-230"><a href="#cb44-230"></a>    <span class="co"># 3. Return these as our initial centroids</span></span>
<span id="cb44-231"><a href="#cb44-231"></a>    random_idx <span class="op">=</span> np.random.permutation(x.shape[<span class="dv">0</span>])</span>
<span id="cb44-232"><a href="#cb44-232"></a>    centroids <span class="op">=</span> x[random_idx[:n_clusters]]</span>
<span id="cb44-233"><a href="#cb44-233"></a>    <span class="cf">return</span> centroids</span>
<span id="cb44-234"><a href="#cb44-234"></a><span class="in">```</span></span>
<span id="cb44-235"><a href="#cb44-235"></a></span>
<span id="cb44-236"><a href="#cb44-236"></a><span class="fu">### Step 2a: Calculate distance to each centroid</span></span>
<span id="cb44-237"><a href="#cb44-237"></a></span>
<span id="cb44-238"><a href="#cb44-238"></a><span class="ss">- </span>Calculate distance to each centroid</span>
<span id="cb44-239"><a href="#cb44-239"></a><span class="ss">  - </span>$k$ clusters are created by associating every observation with the</span>
<span id="cb44-240"><a href="#cb44-240"></a>    nearest mean.</span>
<span id="cb44-241"><a href="#cb44-241"></a>    The partitions here represent the Voroni diagram generated by the means.</span>
<span id="cb44-242"><a href="#cb44-242"></a><span class="ss">  - </span>Find nearest cluster for each point</span>
<span id="cb44-243"><a href="#cb44-243"></a></span>
<span id="cb44-244"><a href="#cb44-244"></a>::: {#fig-kmeans-step2a style="width:40%;"}</span>
<span id="cb44-245"><a href="#cb44-245"></a></span>
<span id="cb44-246"><a href="#cb44-246"></a><span class="al">![](./assets/k-means1.svg)</span></span>
<span id="cb44-247"><a href="#cb44-247"></a></span>
<span id="cb44-248"><a href="#cb44-248"></a>:::</span>
<span id="cb44-249"><a href="#cb44-249"></a></span>
<span id="cb44-252"><a href="#cb44-252"></a><span class="in">```{python}</span></span>
<span id="cb44-253"><a href="#cb44-253"></a><span class="kw">def</span> compute_distance(</span>
<span id="cb44-254"><a href="#cb44-254"></a>        x: np.ndarray,</span>
<span id="cb44-255"><a href="#cb44-255"></a>        centroids: np.ndarray,</span>
<span id="cb44-256"><a href="#cb44-256"></a>        n_clusters: <span class="bu">int</span></span>
<span id="cb44-257"><a href="#cb44-257"></a>) <span class="op">-&gt;</span> np.ndarray:</span>
<span id="cb44-258"><a href="#cb44-258"></a>    <span class="co">"""Compute distance.</span></span>
<span id="cb44-259"><a href="#cb44-259"></a></span>
<span id="cb44-260"><a href="#cb44-260"></a><span class="co">    Inputs:</span></span>
<span id="cb44-261"><a href="#cb44-261"></a><span class="co">      - x (np.ndarray): Input data of shape</span></span>
<span id="cb44-262"><a href="#cb44-262"></a><span class="co">        (num_points, num_features)</span></span>
<span id="cb44-263"><a href="#cb44-263"></a><span class="co">      - centroids (np.ndarray): Cluster centroids  with shape </span></span>
<span id="cb44-264"><a href="#cb44-264"></a><span class="co">        (num_clusters, num_features)</span></span>
<span id="cb44-265"><a href="#cb44-265"></a><span class="co">      - n_clusters (int): Number of clusters being used.</span></span>
<span id="cb44-266"><a href="#cb44-266"></a></span>
<span id="cb44-267"><a href="#cb44-267"></a><span class="co">    Outputs:</span></span>
<span id="cb44-268"><a href="#cb44-268"></a><span class="co">      - distance (np.ndarray): Distance of each point </span></span>
<span id="cb44-269"><a href="#cb44-269"></a><span class="co">        to each centroid with shape </span></span>
<span id="cb44-270"><a href="#cb44-270"></a><span class="co">        (num_points, num_clusters)</span></span>
<span id="cb44-271"><a href="#cb44-271"></a><span class="co">    """</span></span>
<span id="cb44-272"><a href="#cb44-272"></a>    <span class="co"># distance vector</span></span>
<span id="cb44-273"><a href="#cb44-273"></a>    distance <span class="op">=</span> np.zeros((x.shape[<span class="dv">0</span>], n_clusters))</span>
<span id="cb44-274"><a href="#cb44-274"></a>    <span class="co"># loop over each centroid</span></span>
<span id="cb44-275"><a href="#cb44-275"></a>    <span class="cf">for</span> k <span class="kw">in</span> <span class="bu">range</span>(n_clusters):</span>
<span id="cb44-276"><a href="#cb44-276"></a>        <span class="co"># calculate distance for each point from centroid</span></span>
<span id="cb44-277"><a href="#cb44-277"></a>        kcentroid_distance <span class="op">=</span> x <span class="op">-</span> centroids[k, :]</span>
<span id="cb44-278"><a href="#cb44-278"></a>        <span class="co"># apply normalization for stability</span></span>
<span id="cb44-279"><a href="#cb44-279"></a>        row_norm <span class="op">=</span> np.linalg.norm(kcentroid_distance, axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb44-280"><a href="#cb44-280"></a>        <span class="co"># return distance squared</span></span>
<span id="cb44-281"><a href="#cb44-281"></a>        distance[:, k] <span class="op">=</span> np.square(row_norm) </span>
<span id="cb44-282"><a href="#cb44-282"></a></span>
<span id="cb44-283"><a href="#cb44-283"></a>    <span class="cf">return</span> distance</span>
<span id="cb44-284"><a href="#cb44-284"></a><span class="in">```</span></span>
<span id="cb44-285"><a href="#cb44-285"></a></span>
<span id="cb44-286"><a href="#cb44-286"></a><span class="fu">### Step 2b: Find nearest cluster for each point</span></span>
<span id="cb44-287"><a href="#cb44-287"></a></span>
<span id="cb44-290"><a href="#cb44-290"></a><span class="in">```{python}</span></span>
<span id="cb44-291"><a href="#cb44-291"></a><span class="kw">def</span> find_closest_centroid(</span>
<span id="cb44-292"><a href="#cb44-292"></a>        distance: np.ndarray</span>
<span id="cb44-293"><a href="#cb44-293"></a>) <span class="op">-&gt;</span> np.ndarray:</span>
<span id="cb44-294"><a href="#cb44-294"></a>    <span class="co">"""Find closest centroid.</span></span>
<span id="cb44-295"><a href="#cb44-295"></a></span>
<span id="cb44-296"><a href="#cb44-296"></a><span class="co">    Inputs:</span></span>
<span id="cb44-297"><a href="#cb44-297"></a><span class="co">      - distance (np.ndarray): Distance of each point to each centroid with shape</span></span>
<span id="cb44-298"><a href="#cb44-298"></a><span class="co">        (num_points, num_clusters)</span></span>
<span id="cb44-299"><a href="#cb44-299"></a></span>
<span id="cb44-300"><a href="#cb44-300"></a><span class="co">    Outputs:</span></span>
<span id="cb44-301"><a href="#cb44-301"></a><span class="co">      - nearest_centroid_indices (np.ndarray): Index of nearest centroid with shape</span></span>
<span id="cb44-302"><a href="#cb44-302"></a><span class="co">        (num_points,)</span></span>
<span id="cb44-303"><a href="#cb44-303"></a><span class="co">    """</span></span>
<span id="cb44-304"><a href="#cb44-304"></a>    nearest_centroid_indices <span class="op">=</span> np.argmin(distance, axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb44-305"><a href="#cb44-305"></a>    <span class="cf">return</span> nearest_centroid_indices</span>
<span id="cb44-306"><a href="#cb44-306"></a></span>
<span id="cb44-307"><a href="#cb44-307"></a><span class="in">```</span></span>
<span id="cb44-308"><a href="#cb44-308"></a></span>
<span id="cb44-309"><a href="#cb44-309"></a><span class="fu">### Step 3: Calculate new centroids</span></span>
<span id="cb44-310"><a href="#cb44-310"></a></span>
<span id="cb44-311"><a href="#cb44-311"></a><span class="ss">3. </span>Calculate new centroids</span>
<span id="cb44-312"><a href="#cb44-312"></a><span class="ss">   - </span>The **centroid** of each of the $k$ clusters becomes the new mean</span>
<span id="cb44-313"><a href="#cb44-313"></a></span>
<span id="cb44-314"><a href="#cb44-314"></a>::: {#fig-kmeans-step3 style="width:40%;"}</span>
<span id="cb44-315"><a href="#cb44-315"></a><span class="al">![](./assets/k-means2.svg)</span></span>
<span id="cb44-316"><a href="#cb44-316"></a>:::</span>
<span id="cb44-317"><a href="#cb44-317"></a></span>
<span id="cb44-320"><a href="#cb44-320"></a><span class="in">```{python}</span></span>
<span id="cb44-321"><a href="#cb44-321"></a><span class="kw">def</span> compute_centroids(</span>
<span id="cb44-322"><a href="#cb44-322"></a>        x: np.ndarray,</span>
<span id="cb44-323"><a href="#cb44-323"></a>        nearest_centroid_indices: np.ndarray,</span>
<span id="cb44-324"><a href="#cb44-324"></a>        n_clusters: <span class="bu">int</span></span>
<span id="cb44-325"><a href="#cb44-325"></a>) <span class="op">-&gt;</span> np.ndarray:</span>
<span id="cb44-326"><a href="#cb44-326"></a>    <span class="co">"""Compute centroids.</span></span>
<span id="cb44-327"><a href="#cb44-327"></a><span class="co">    </span></span>
<span id="cb44-328"><a href="#cb44-328"></a><span class="co">    Inputs:</span></span>
<span id="cb44-329"><a href="#cb44-329"></a><span class="co">      - x (np.ndarray): Input data of shape</span></span>
<span id="cb44-330"><a href="#cb44-330"></a><span class="co">        (num_points, num_features)</span></span>
<span id="cb44-331"><a href="#cb44-331"></a><span class="co">      - nearest_centroid_indices: Index of nearest centroid of shape</span></span>
<span id="cb44-332"><a href="#cb44-332"></a><span class="co">        (num_points,)</span></span>
<span id="cb44-333"><a href="#cb44-333"></a><span class="co">      - n_clusters (int): Number of clusters being used</span></span>
<span id="cb44-334"><a href="#cb44-334"></a><span class="co">      </span></span>
<span id="cb44-335"><a href="#cb44-335"></a><span class="co">    Outputs:</span></span>
<span id="cb44-336"><a href="#cb44-336"></a><span class="co">      - centroids (np.ndarray): Cluster centroids with shape</span></span>
<span id="cb44-337"><a href="#cb44-337"></a><span class="co">        (num_clusters, num_features)</span></span>
<span id="cb44-338"><a href="#cb44-338"></a><span class="co">    """</span></span>
<span id="cb44-339"><a href="#cb44-339"></a>    <span class="co"># new centroids vector</span></span>
<span id="cb44-340"><a href="#cb44-340"></a>    centroids <span class="op">=</span> np.zeros((n_clusters, x.shape[<span class="dv">1</span>]))</span>
<span id="cb44-341"><a href="#cb44-341"></a>    <span class="co"># loop over each centroids</span></span>
<span id="cb44-342"><a href="#cb44-342"></a>    <span class="cf">for</span> k <span class="kw">in</span> <span class="bu">range</span>(n_clusters):</span>
<span id="cb44-343"><a href="#cb44-343"></a>        <span class="co"># calculate the mean of all points assigned to this centroid</span></span>
<span id="cb44-344"><a href="#cb44-344"></a>        centroids[k, :] <span class="op">=</span> np.mean(</span>
<span id="cb44-345"><a href="#cb44-345"></a>            x[nearest_centroid_indices <span class="op">==</span> k, :],</span>
<span id="cb44-346"><a href="#cb44-346"></a>            axis<span class="op">=</span><span class="dv">0</span></span>
<span id="cb44-347"><a href="#cb44-347"></a>        )</span>
<span id="cb44-348"><a href="#cb44-348"></a>    <span class="cf">return</span> centroids</span>
<span id="cb44-349"><a href="#cb44-349"></a><span class="in">```</span></span>
<span id="cb44-350"><a href="#cb44-350"></a></span>
<span id="cb44-351"><a href="#cb44-351"></a><span class="fu">### Step 4: Repeat until convergence</span></span>
<span id="cb44-352"><a href="#cb44-352"></a></span>
<span id="cb44-353"><a href="#cb44-353"></a><span class="ss">4. </span>Repeat steps 2 and 3 until convergence has been reached</span>
<span id="cb44-354"><a href="#cb44-354"></a></span>
<span id="cb44-355"><a href="#cb44-355"></a>::: {#fig-kmeans-step4 style="width:40%;"}</span>
<span id="cb44-356"><a href="#cb44-356"></a></span>
<span id="cb44-357"><a href="#cb44-357"></a><span class="al">![](./assets/k-means3.svg)</span></span>
<span id="cb44-358"><a href="#cb44-358"></a></span>
<span id="cb44-359"><a href="#cb44-359"></a>:::</span>
<span id="cb44-360"><a href="#cb44-360"></a></span>
<span id="cb44-363"><a href="#cb44-363"></a><span class="in">```{python}</span></span>
<span id="cb44-364"><a href="#cb44-364"></a><span class="im">from</span> __future__ <span class="im">import</span> absolute_import, annotations, division, print_function</span>
<span id="cb44-365"><a href="#cb44-365"></a></span>
<span id="cb44-366"><a href="#cb44-366"></a><span class="im">from</span> typing <span class="im">import</span> Optional</span>
<span id="cb44-367"><a href="#cb44-367"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb44-368"><a href="#cb44-368"></a><span class="im">import</span> IPython.display <span class="im">as</span> ipydis</span>
<span id="cb44-369"><a href="#cb44-369"></a></span>
<span id="cb44-370"><a href="#cb44-370"></a><span class="im">from</span> bootcamp.plots <span class="im">import</span> plot_kmeans_points</span>
<span id="cb44-371"><a href="#cb44-371"></a></span>
<span id="cb44-372"><a href="#cb44-372"></a><span class="kw">def</span> apply_kmeans(</span>
<span id="cb44-373"><a href="#cb44-373"></a>    x: np.ndarray,</span>
<span id="cb44-374"><a href="#cb44-374"></a>    n_clusters: <span class="bu">int</span>,</span>
<span id="cb44-375"><a href="#cb44-375"></a>    iterations: <span class="bu">int</span> <span class="op">=</span> <span class="dv">100</span>,</span>
<span id="cb44-376"><a href="#cb44-376"></a>    seed: <span class="bu">int</span> <span class="op">=</span> <span class="dv">123</span>,</span>
<span id="cb44-377"><a href="#cb44-377"></a>    cmap: Optional[<span class="bu">str</span>] <span class="op">=</span> <span class="va">None</span>,</span>
<span id="cb44-378"><a href="#cb44-378"></a>) <span class="op">-&gt;</span> <span class="bu">tuple</span>[np.ndarray, np.ndarray]:</span>
<span id="cb44-379"><a href="#cb44-379"></a>    <span class="co">"""Returns (centroids, cluster_id)."""</span></span>
<span id="cb44-380"><a href="#cb44-380"></a>    <span class="co"># ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓</span></span>
<span id="cb44-381"><a href="#cb44-381"></a>    <span class="co"># ┃ initialize centroids:               ┃</span></span>
<span id="cb44-382"><a href="#cb44-382"></a>    <span class="co"># ┃   - shape: (n_clusters, x.shape[1]) ┃</span></span>
<span id="cb44-383"><a href="#cb44-383"></a>    <span class="co"># ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛</span></span>
<span id="cb44-384"><a href="#cb44-384"></a>    centroids <span class="op">=</span> initialize_centroids(</span>
<span id="cb44-385"><a href="#cb44-385"></a>        x,  <span class="co"># (n_points, n_features)</span></span>
<span id="cb44-386"><a href="#cb44-386"></a>        n_clusters<span class="op">=</span>n_clusters,</span>
<span id="cb44-387"><a href="#cb44-387"></a>        seed<span class="op">=</span>seed,</span>
<span id="cb44-388"><a href="#cb44-388"></a>    )</span>
<span id="cb44-389"><a href="#cb44-389"></a>    <span class="co"># ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓</span></span>
<span id="cb44-390"><a href="#cb44-390"></a>    <span class="co"># ┃ -- Iteratively improve centroid location ----------------- ┃</span></span>
<span id="cb44-391"><a href="#cb44-391"></a>    <span class="co"># ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛</span></span>
<span id="cb44-392"><a href="#cb44-392"></a>    <span class="co"># ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓</span></span>
<span id="cb44-393"><a href="#cb44-393"></a>    <span class="co"># ┃ 1. Compute the distance from entries in x to each centroid ┃</span></span>
<span id="cb44-394"><a href="#cb44-394"></a>    <span class="co"># ┃    - distance.shape: (n_points, n_clusters)                ┃</span></span>
<span id="cb44-395"><a href="#cb44-395"></a>    <span class="co"># ┃ 2. Return the closest cluster (0, 1, ..., n_clusters-1)    ┃</span></span>
<span id="cb44-396"><a href="#cb44-396"></a>    <span class="co"># ┃    - cluster_id.shape: (n_points)                          ┃</span></span>
<span id="cb44-397"><a href="#cb44-397"></a>    <span class="co"># ┃ 3. Calculate the mean position of each labeled cluster     ┃</span></span>
<span id="cb44-398"><a href="#cb44-398"></a>    <span class="co"># ┃    - centroids.shape: (n_clusters, n_features)             ┃</span></span>
<span id="cb44-399"><a href="#cb44-399"></a>    <span class="co"># ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛</span></span>
<span id="cb44-400"><a href="#cb44-400"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(iterations):</span>
<span id="cb44-401"><a href="#cb44-401"></a>        <span class="co"># save old centroids</span></span>
<span id="cb44-402"><a href="#cb44-402"></a>        old_centroids <span class="op">=</span> centroids</span>
<span id="cb44-403"><a href="#cb44-403"></a>        distance <span class="op">=</span> compute_distance(x, old_centroids, n_clusters)</span>
<span id="cb44-404"><a href="#cb44-404"></a>        cluster_id <span class="op">=</span> find_closest_centroid(distance)</span>
<span id="cb44-405"><a href="#cb44-405"></a>        centroids <span class="op">=</span> compute_centroids(x, cluster_id, n_clusters)</span>
<span id="cb44-406"><a href="#cb44-406"></a></span>
<span id="cb44-407"><a href="#cb44-407"></a>        <span class="co"># plotting for visual comprehension</span></span>
<span id="cb44-408"><a href="#cb44-408"></a>        ipydis.clear_output(<span class="st">"wait"</span>)</span>
<span id="cb44-409"><a href="#cb44-409"></a>        <span class="bu">print</span>(<span class="ss">f"Iteration: </span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb44-410"><a href="#cb44-410"></a>        plot_kmeans_points(x, centroids, cluster_id, cmap<span class="op">=</span>cmap)</span>
<span id="cb44-411"><a href="#cb44-411"></a>        time.sleep(<span class="fl">0.5</span>)</span>
<span id="cb44-412"><a href="#cb44-412"></a></span>
<span id="cb44-413"><a href="#cb44-413"></a>        <span class="co"># if our points are the same as the old centroids, then we can stop</span></span>
<span id="cb44-414"><a href="#cb44-414"></a>        <span class="cf">if</span> np.<span class="bu">all</span>(old_centroids <span class="op">==</span> centroids):</span>
<span id="cb44-415"><a href="#cb44-415"></a>            <span class="bu">print</span>(<span class="ss">f"No change in centroids! Exiting!"</span>)</span>
<span id="cb44-416"><a href="#cb44-416"></a>            <span class="cf">break</span></span>
<span id="cb44-417"><a href="#cb44-417"></a></span>
<span id="cb44-418"><a href="#cb44-418"></a>    <span class="cf">return</span> centroids, cluster_id</span>
<span id="cb44-419"><a href="#cb44-419"></a><span class="in">```</span></span>
<span id="cb44-420"><a href="#cb44-420"></a></span>
<span id="cb44-421"><a href="#cb44-421"></a><span class="fu">## Run Example</span></span>
<span id="cb44-422"><a href="#cb44-422"></a></span>
<span id="cb44-425"><a href="#cb44-425"></a><span class="in">```{python}</span></span>
<span id="cb44-426"><a href="#cb44-426"></a><span class="im">from</span> bootcamp.plots <span class="im">import</span> COLORS</span>
<span id="cb44-427"><a href="#cb44-427"></a></span>
<span id="cb44-428"><a href="#cb44-428"></a>cmap <span class="op">=</span> ListedColormap(<span class="bu">list</span>(COLORS.values()))</span>
<span id="cb44-429"><a href="#cb44-429"></a></span>
<span id="cb44-430"><a href="#cb44-430"></a>n_samples <span class="op">=</span> <span class="dv">500</span>  <span class="co"># 300 2D data points</span></span>
<span id="cb44-431"><a href="#cb44-431"></a>n_features <span class="op">=</span> <span class="dv">2</span>  <span class="co"># 2D data</span></span>
<span id="cb44-432"><a href="#cb44-432"></a>n_clusters_true <span class="op">=</span> <span class="dv">5</span>  <span class="co"># unique blobs</span></span>
<span id="cb44-433"><a href="#cb44-433"></a>n_clusters_guess <span class="op">=</span> <span class="dv">5</span></span>
<span id="cb44-434"><a href="#cb44-434"></a>SEED <span class="op">=</span> <span class="dv">456</span></span>
<span id="cb44-435"><a href="#cb44-435"></a>iterations <span class="op">=</span> <span class="dv">50</span></span>
<span id="cb44-436"><a href="#cb44-436"></a></span>
<span id="cb44-437"><a href="#cb44-437"></a><span class="co"># https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_blobs.html</span></span>
<span id="cb44-438"><a href="#cb44-438"></a>x, y <span class="op">=</span> datasets.make_blobs(</span>
<span id="cb44-439"><a href="#cb44-439"></a>    n_samples<span class="op">=</span>n_samples,</span>
<span id="cb44-440"><a href="#cb44-440"></a>    n_features<span class="op">=</span>n_features,</span>
<span id="cb44-441"><a href="#cb44-441"></a>    centers<span class="op">=</span>n_clusters_true,</span>
<span id="cb44-442"><a href="#cb44-442"></a>    random_state<span class="op">=</span>SEED,</span>
<span id="cb44-443"><a href="#cb44-443"></a>)</span>
<span id="cb44-444"><a href="#cb44-444"></a></span>
<span id="cb44-445"><a href="#cb44-445"></a>x_sc <span class="op">=</span> StandardScaler().fit_transform(x)</span>
<span id="cb44-446"><a href="#cb44-446"></a></span>
<span id="cb44-447"><a href="#cb44-447"></a>centroids, cluster_id <span class="op">=</span> apply_kmeans(</span>
<span id="cb44-448"><a href="#cb44-448"></a>    x_sc,</span>
<span id="cb44-449"><a href="#cb44-449"></a>    n_clusters_guess,</span>
<span id="cb44-450"><a href="#cb44-450"></a>    iterations,</span>
<span id="cb44-451"><a href="#cb44-451"></a>    SEED,</span>
<span id="cb44-452"><a href="#cb44-452"></a>    cmap<span class="op">=</span>cmap,</span>
<span id="cb44-453"><a href="#cb44-453"></a>)</span>
<span id="cb44-454"><a href="#cb44-454"></a><span class="in">```</span></span>
<span id="cb44-455"><a href="#cb44-455"></a></span>
<span id="cb44-456"><a href="#cb44-456"></a></span>
<span id="cb44-457"><a href="#cb44-457"></a><span class="fu">## K-Means on a Breast Cancer Dataset</span></span>
<span id="cb44-458"><a href="#cb44-458"></a></span>
<span id="cb44-459"><a href="#cb44-459"></a>Now we use more realistic data that cannot be easily plotted on a 2D grid.</span>
<span id="cb44-460"><a href="#cb44-460"></a>This dataset has 30 features (columns) for 569 patients (rows).</span>
<span id="cb44-461"><a href="#cb44-461"></a>In addition, there is a _target_ feature that indicates if the cancer was</span>
<span id="cb44-462"><a href="#cb44-462"></a>_malignant_ (0) or _benign_ (1).</span>
<span id="cb44-463"><a href="#cb44-463"></a>In the ideal case, our 30 features would provide easy deliniation between these</span>
<span id="cb44-464"><a href="#cb44-464"></a>two classes.</span>
<span id="cb44-465"><a href="#cb44-465"></a></span>
<span id="cb44-466"><a href="#cb44-466"></a>Let's extract our data into $x$ and our truth labels into $y$</span>
<span id="cb44-467"><a href="#cb44-467"></a></span>
<span id="cb44-470"><a href="#cb44-470"></a><span class="in">```{python}</span></span>
<span id="cb44-471"><a href="#cb44-471"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb44-472"><a href="#cb44-472"></a></span>
<span id="cb44-473"><a href="#cb44-473"></a></span>
<span id="cb44-474"><a href="#cb44-474"></a><span class="kw">def</span> load_cancer_data() <span class="op">-&gt;</span> <span class="bu">dict</span>:</span>
<span id="cb44-475"><a href="#cb44-475"></a>    <span class="co">"""Return cancer dataset (unscaled)."""</span></span>
<span id="cb44-476"><a href="#cb44-476"></a>    <span class="im">from</span> sklearn <span class="im">import</span> datasets</span>
<span id="cb44-477"><a href="#cb44-477"></a></span>
<span id="cb44-478"><a href="#cb44-478"></a>    data <span class="op">=</span> datasets.load_breast_cancer()</span>
<span id="cb44-479"><a href="#cb44-479"></a>    <span class="cf">return</span> data</span>
<span id="cb44-480"><a href="#cb44-480"></a></span>
<span id="cb44-481"><a href="#cb44-481"></a></span>
<span id="cb44-482"><a href="#cb44-482"></a><span class="kw">def</span> sort_cancer_data(data: <span class="bu">dict</span>) <span class="op">-&gt;</span> <span class="bu">tuple</span>[pd.DataFrame, pd.Series]:</span>
<span id="cb44-483"><a href="#cb44-483"></a>    <span class="co"># Get features and target</span></span>
<span id="cb44-484"><a href="#cb44-484"></a>    x <span class="op">=</span> pd.DataFrame(</span>
<span id="cb44-485"><a href="#cb44-485"></a>        data[<span class="st">"data"</span>],</span>
<span id="cb44-486"><a href="#cb44-486"></a>        columns<span class="op">=</span>data[<span class="st">"feature_names"</span>],</span>
<span id="cb44-487"><a href="#cb44-487"></a>    )</span>
<span id="cb44-488"><a href="#cb44-488"></a>    x <span class="op">=</span> x[<span class="bu">sorted</span>(x.columns)]</span>
<span id="cb44-489"><a href="#cb44-489"></a>    y <span class="op">=</span> data[<span class="st">"target"</span>]</span>
<span id="cb44-490"><a href="#cb44-490"></a>    <span class="cf">return</span> x, y</span>
<span id="cb44-491"><a href="#cb44-491"></a><span class="in">```</span></span>
<span id="cb44-492"><a href="#cb44-492"></a></span>
<span id="cb44-495"><a href="#cb44-495"></a><span class="in">```{python}</span></span>
<span id="cb44-496"><a href="#cb44-496"></a>data <span class="op">=</span> load_cancer_data()</span>
<span id="cb44-497"><a href="#cb44-497"></a>x, y <span class="op">=</span> sort_cancer_data(data)</span>
<span id="cb44-498"><a href="#cb44-498"></a><span class="in">```</span></span>
<span id="cb44-499"><a href="#cb44-499"></a></span>
<span id="cb44-502"><a href="#cb44-502"></a><span class="in">```{python}</span></span>
<span id="cb44-503"><a href="#cb44-503"></a><span class="bu">print</span>(data.keys())</span>
<span id="cb44-504"><a href="#cb44-504"></a><span class="bu">print</span>(<span class="st">"data size:"</span>, <span class="bu">len</span>(data[<span class="st">"data"</span>]))</span>
<span id="cb44-505"><a href="#cb44-505"></a><span class="bu">print</span>(<span class="st">"number of features:"</span>, <span class="bu">len</span>(data[<span class="st">"feature_names"</span>]))</span>
<span id="cb44-506"><a href="#cb44-506"></a><span class="bu">print</span>(data[<span class="st">"feature_names"</span>])</span>
<span id="cb44-507"><a href="#cb44-507"></a><span class="in">```</span></span>
<span id="cb44-508"><a href="#cb44-508"></a></span>
<span id="cb44-511"><a href="#cb44-511"></a><span class="in">```{python}</span></span>
<span id="cb44-512"><a href="#cb44-512"></a><span class="co"># wrap data in pandas DataFrame</span></span>
<span id="cb44-513"><a href="#cb44-513"></a><span class="co"># y = data['target'] # value: 0 = 'Malignant' 1 = 'Benign'</span></span>
<span id="cb44-514"><a href="#cb44-514"></a>x <span class="op">=</span> data[<span class="st">"data"</span>]</span>
<span id="cb44-515"><a href="#cb44-515"></a>y <span class="op">=</span> data[<span class="st">"target"</span>]</span>
<span id="cb44-516"><a href="#cb44-516"></a><span class="bu">print</span>(<span class="ss">f"(malignant, benign): </span><span class="sc">{</span>np<span class="sc">.</span>bincount(y)<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb44-517"><a href="#cb44-517"></a><span class="bu">print</span>(<span class="ss">f"x.shape: </span><span class="sc">{</span>x<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb44-518"><a href="#cb44-518"></a><span class="bu">print</span>(<span class="ss">f"y.shape: </span><span class="sc">{</span>y<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb44-519"><a href="#cb44-519"></a><span class="in">```</span></span>
<span id="cb44-520"><a href="#cb44-520"></a></span>
<span id="cb44-523"><a href="#cb44-523"></a><span class="in">```{python}</span></span>
<span id="cb44-524"><a href="#cb44-524"></a><span class="co"># normalize</span></span>
<span id="cb44-525"><a href="#cb44-525"></a>x_sc <span class="op">=</span> StandardScaler().fit_transform(x)</span>
<span id="cb44-526"><a href="#cb44-526"></a><span class="in">```</span></span>
<span id="cb44-527"><a href="#cb44-527"></a></span>
<span id="cb44-528"><a href="#cb44-528"></a>Plot histogram of number of true class labels (malignant/benign) and number of k-means cluster labels.</span>
<span id="cb44-529"><a href="#cb44-529"></a></span>
<span id="cb44-530"><a href="#cb44-530"></a>This gives an indication of how well we did in clustering our data, but is not a "correctness" or "accuracy" metric.</span>
<span id="cb44-531"><a href="#cb44-531"></a></span>
<span id="cb44-532"><a href="#cb44-532"></a>::: {.callout-note}</span>
<span id="cb44-533"><a href="#cb44-533"></a></span>
<span id="cb44-534"><a href="#cb44-534"></a>Keep in mind, k-means is not a classifier.</span>
<span id="cb44-535"><a href="#cb44-535"></a></span>
<span id="cb44-536"><a href="#cb44-536"></a>Cluster label 0 given from k-means, does not correspond to cluster label 0 in the truth.</span>
<span id="cb44-537"><a href="#cb44-537"></a>:::</span>
<span id="cb44-538"><a href="#cb44-538"></a></span>
<span id="cb44-541"><a href="#cb44-541"></a><span class="in">```{python}</span></span>
<span id="cb44-542"><a href="#cb44-542"></a><span class="im">from</span> bootcamp.plots <span class="im">import</span> plot_hists</span>
<span id="cb44-543"><a href="#cb44-543"></a></span>
<span id="cb44-544"><a href="#cb44-544"></a><span class="co"># for seed in range(10):</span></span>
<span id="cb44-545"><a href="#cb44-545"></a>CSEED <span class="op">=</span> <span class="dv">42</span></span>
<span id="cb44-546"><a href="#cb44-546"></a>kmeans <span class="op">=</span> KMeans(n_clusters<span class="op">=</span><span class="dv">2</span>, random_state<span class="op">=</span>CSEED)</span>
<span id="cb44-547"><a href="#cb44-547"></a><span class="co"># fit the data</span></span>
<span id="cb44-548"><a href="#cb44-548"></a>kfit <span class="op">=</span> kmeans.fit(x_sc)</span>
<span id="cb44-549"><a href="#cb44-549"></a>cluster_ids <span class="op">=</span> kmeans.labels_</span>
<span id="cb44-550"><a href="#cb44-550"></a><span class="bu">print</span>(</span>
<span id="cb44-551"><a href="#cb44-551"></a>    <span class="ss">f"seed: </span><span class="sc">{</span>CSEED<span class="sc">}</span><span class="ch">\n</span><span class="ss">Number of samples in each cluster: </span><span class="sc">{</span>np<span class="sc">.</span>bincount(kmeans.labels_)<span class="sc">}</span><span class="ss">"</span></span>
<span id="cb44-552"><a href="#cb44-552"></a>)</span>
<span id="cb44-553"><a href="#cb44-553"></a>k_means_bins <span class="op">=</span> np.bincount(cluster_ids)</span>
<span id="cb44-554"><a href="#cb44-554"></a>y_bins <span class="op">=</span> np.bincount(y)</span>
<span id="cb44-555"><a href="#cb44-555"></a>plt.rcParams[<span class="st">"figure.figsize"</span>] <span class="op">=</span> [<span class="fl">1.5</span> <span class="op">*</span> DFIGSIZE[<span class="dv">0</span>], <span class="fl">0.8</span> <span class="op">*</span> DFIGSIZE[<span class="dv">1</span>]]</span>
<span id="cb44-556"><a href="#cb44-556"></a>plot_hists(k_means_bins, y_bins, xlabels<span class="op">=</span>[<span class="st">"Malignant"</span>, <span class="st">"Benign"</span>])</span>
<span id="cb44-557"><a href="#cb44-557"></a><span class="in">```</span></span>
<span id="cb44-558"><a href="#cb44-558"></a></span>
<span id="cb44-559"><a href="#cb44-559"></a>This plot shows the normalized inertia as we vary the number <span class="in">`n_clusters`</span> used</span>
<span id="cb44-560"><a href="#cb44-560"></a>in our k-means fit to the breast cancer data.</span>
<span id="cb44-561"><a href="#cb44-561"></a>This value essentially indicates the mean distance of a point to the cluster</span>
<span id="cb44-562"><a href="#cb44-562"></a>centroids.</span>
<span id="cb44-563"><a href="#cb44-563"></a>Obviously more clusters result in data points being nearer to a centroid.</span>
<span id="cb44-564"><a href="#cb44-564"></a></span>
<span id="cb44-565"><a href="#cb44-565"></a>If our dataset easily split into 2 clusters, we would see a step function</span>
<span id="cb44-566"><a href="#cb44-566"></a>behavior going from 1 to 2, then only very minor improvements above 2.</span>
<span id="cb44-567"><a href="#cb44-567"></a>What we see here tells us our data does not easily cluster.</span>
<span id="cb44-568"><a href="#cb44-568"></a></span>
<span id="cb44-571"><a href="#cb44-571"></a><span class="in">```{python}</span></span>
<span id="cb44-572"><a href="#cb44-572"></a><span class="im">from</span> bootcamp.plots <span class="im">import</span> plot_kmeans_obj</span>
<span id="cb44-573"><a href="#cb44-573"></a></span>
<span id="cb44-574"><a href="#cb44-574"></a>plt.rcParams[<span class="st">"figure.figsize"</span>] <span class="op">=</span> [DFIGSIZE[<span class="dv">0</span>], <span class="fl">0.6</span> <span class="op">*</span> DFIGSIZE[<span class="dv">1</span>]]</span>
<span id="cb44-575"><a href="#cb44-575"></a>_ <span class="op">=</span> plot_kmeans_obj(x_sc, nclusters<span class="op">=</span><span class="dv">10</span>, plot_points<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb44-576"><a href="#cb44-576"></a><span class="in">```</span></span>
<span id="cb44-577"><a href="#cb44-577"></a></span>
<span id="cb44-578"><a href="#cb44-578"></a>For example, if we return to our blob data with 2 clusters, it become clear.</span>
<span id="cb44-579"><a href="#cb44-579"></a></span>
<span id="cb44-582"><a href="#cb44-582"></a><span class="in">```{python}</span></span>
<span id="cb44-583"><a href="#cb44-583"></a>n_samples <span class="op">=</span> <span class="dv">300</span>  <span class="co"># 300 2D data points</span></span>
<span id="cb44-584"><a href="#cb44-584"></a>n_features <span class="op">=</span> <span class="dv">2</span>  <span class="co"># 2D data</span></span>
<span id="cb44-585"><a href="#cb44-585"></a>n_clusters <span class="op">=</span> <span class="dv">2</span>  <span class="co"># unique blobs</span></span>
<span id="cb44-586"><a href="#cb44-586"></a>seed <span class="op">=</span> <span class="dv">456</span></span>
<span id="cb44-587"><a href="#cb44-587"></a>iterations <span class="op">=</span> <span class="dv">25</span></span>
<span id="cb44-588"><a href="#cb44-588"></a>x, y <span class="op">=</span> datasets.make_blobs(</span>
<span id="cb44-589"><a href="#cb44-589"></a>    n_samples<span class="op">=</span>n_samples, n_features<span class="op">=</span>n_features, centers<span class="op">=</span>n_clusters, random_state<span class="op">=</span>seed</span>
<span id="cb44-590"><a href="#cb44-590"></a>)</span>
<span id="cb44-591"><a href="#cb44-591"></a>plot_kmeans_obj(x, nclusters<span class="op">=</span><span class="dv">10</span>, plot_points<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb44-592"><a href="#cb44-592"></a><span class="in">```</span></span>
<span id="cb44-593"><a href="#cb44-593"></a></span>
<span id="cb44-594"><a href="#cb44-594"></a><span class="fu">## [Self-organizing maps](https://en.wikipedia.org/wiki/Self-organizing_map)</span></span>
<span id="cb44-595"><a href="#cb44-595"></a></span>
<span id="cb44-596"><a href="#cb44-596"></a>Self Organizing Maps (SOM) were proposed and became widespread in the 1980s, by a Finnish professor named Teuvo Kohonen and are also called 'Kohonen maps'.</span>
<span id="cb44-597"><a href="#cb44-597"></a></span>
<span id="cb44-598"><a href="#cb44-598"></a></span>
<span id="cb44-601"><a href="#cb44-601"></a><span class="in">```{python}</span></span>
<span id="cb44-602"><a href="#cb44-602"></a><span class="co"># finding best matching unit</span></span>
<span id="cb44-603"><a href="#cb44-603"></a><span class="kw">def</span> find_bmu(t, net, n):</span>
<span id="cb44-604"><a href="#cb44-604"></a>    <span class="co">"""</span></span>
<span id="cb44-605"><a href="#cb44-605"></a><span class="co">    Find the best matching unit for a given vector, t, in the SOM</span></span>
<span id="cb44-606"><a href="#cb44-606"></a><span class="co">    Returns: a (bmu, bmu_idx) tuple where bmu is the high-dimensional BMU</span></span>
<span id="cb44-607"><a href="#cb44-607"></a><span class="co">             and bmu_idx is the index of this vector in the SOM</span></span>
<span id="cb44-608"><a href="#cb44-608"></a><span class="co">    """</span></span>
<span id="cb44-609"><a href="#cb44-609"></a>    bmu_idx <span class="op">=</span> np.array([<span class="dv">0</span>, <span class="dv">0</span>])</span>
<span id="cb44-610"><a href="#cb44-610"></a>    <span class="co"># set the initial minimum distance to a huge number</span></span>
<span id="cb44-611"><a href="#cb44-611"></a>    <span class="co"># min_dist = #np.iinfo(np.int).max</span></span>
<span id="cb44-612"><a href="#cb44-612"></a>    min_dist <span class="op">=</span> np.inf</span>
<span id="cb44-613"><a href="#cb44-613"></a>    <span class="co"># calculate the high-dimensional distance between each neuron and the input</span></span>
<span id="cb44-614"><a href="#cb44-614"></a>    <span class="cf">for</span> x <span class="kw">in</span> <span class="bu">range</span>(net.shape[<span class="dv">0</span>]):</span>
<span id="cb44-615"><a href="#cb44-615"></a>        <span class="cf">for</span> y <span class="kw">in</span> <span class="bu">range</span>(net.shape[<span class="dv">1</span>]):</span>
<span id="cb44-616"><a href="#cb44-616"></a>            w <span class="op">=</span> net[x, y, :].reshape(n, <span class="dv">1</span>)</span>
<span id="cb44-617"><a href="#cb44-617"></a>            <span class="co"># don't bother with actual Euclidean distance, to avoid expensive sqrt operation</span></span>
<span id="cb44-618"><a href="#cb44-618"></a>            sq_dist <span class="op">=</span> np.<span class="bu">sum</span>((w <span class="op">-</span> t) <span class="op">**</span> <span class="dv">2</span>)</span>
<span id="cb44-619"><a href="#cb44-619"></a>            <span class="cf">if</span> sq_dist <span class="op">&lt;</span> min_dist:</span>
<span id="cb44-620"><a href="#cb44-620"></a>                min_dist <span class="op">=</span> sq_dist</span>
<span id="cb44-621"><a href="#cb44-621"></a>                bmu_idx <span class="op">=</span> np.array([x, y])</span>
<span id="cb44-622"><a href="#cb44-622"></a>    <span class="co"># get vector corresponding to bmu_idx</span></span>
<span id="cb44-623"><a href="#cb44-623"></a>    bmu <span class="op">=</span> net[bmu_idx[<span class="dv">0</span>], bmu_idx[<span class="dv">1</span>], :].reshape(n, <span class="dv">1</span>)</span>
<span id="cb44-624"><a href="#cb44-624"></a>    <span class="co"># return the (bmu, bmu_idx) tuple</span></span>
<span id="cb44-625"><a href="#cb44-625"></a>    <span class="cf">return</span> (bmu, bmu_idx)</span>
<span id="cb44-626"><a href="#cb44-626"></a></span>
<span id="cb44-627"><a href="#cb44-627"></a></span>
<span id="cb44-628"><a href="#cb44-628"></a><span class="co"># Decaying radius of influence</span></span>
<span id="cb44-629"><a href="#cb44-629"></a><span class="kw">def</span> decay_radius(initial_radius, i, time_constant):</span>
<span id="cb44-630"><a href="#cb44-630"></a>    <span class="cf">return</span> initial_radius <span class="op">*</span> np.exp(<span class="op">-</span>i <span class="op">/</span> time_constant)</span>
<span id="cb44-631"><a href="#cb44-631"></a></span>
<span id="cb44-632"><a href="#cb44-632"></a></span>
<span id="cb44-633"><a href="#cb44-633"></a><span class="co"># Decaying learning rate</span></span>
<span id="cb44-634"><a href="#cb44-634"></a><span class="kw">def</span> decay_learning_rate(initial_learning_rate, i, n_iterations):</span>
<span id="cb44-635"><a href="#cb44-635"></a>    <span class="cf">return</span> initial_learning_rate <span class="op">*</span> np.exp(<span class="op">-</span>i <span class="op">/</span> n_iterations)</span>
<span id="cb44-636"><a href="#cb44-636"></a></span>
<span id="cb44-637"><a href="#cb44-637"></a></span>
<span id="cb44-638"><a href="#cb44-638"></a><span class="co"># Influence in 2D space</span></span>
<span id="cb44-639"><a href="#cb44-639"></a><span class="kw">def</span> calculate_influence(distance, radius):</span>
<span id="cb44-640"><a href="#cb44-640"></a>    <span class="cf">return</span> np.exp(<span class="op">-</span>distance <span class="op">/</span> (<span class="fl">2.0</span> <span class="op">*</span> (radius<span class="op">**</span><span class="dv">2</span>)))</span>
<span id="cb44-641"><a href="#cb44-641"></a></span>
<span id="cb44-642"><a href="#cb44-642"></a></span>
<span id="cb44-643"><a href="#cb44-643"></a><span class="co"># Update weights</span></span>
<span id="cb44-644"><a href="#cb44-644"></a><span class="kw">def</span> update_weights(net, bmu_idx, r, l):</span>
<span id="cb44-645"><a href="#cb44-645"></a>    wlen <span class="op">=</span> net.shape[<span class="dv">2</span>]</span>
<span id="cb44-646"><a href="#cb44-646"></a>    <span class="cf">for</span> x <span class="kw">in</span> <span class="bu">range</span>(net.shape[<span class="dv">0</span>]):</span>
<span id="cb44-647"><a href="#cb44-647"></a>        <span class="cf">for</span> y <span class="kw">in</span> <span class="bu">range</span>(net.shape[<span class="dv">1</span>]):</span>
<span id="cb44-648"><a href="#cb44-648"></a>            w <span class="op">=</span> net[x, y, :].reshape(wlen, <span class="dv">1</span>)</span>
<span id="cb44-649"><a href="#cb44-649"></a>            <span class="co"># get the 2-D distance (again, not the actual Euclidean distance)</span></span>
<span id="cb44-650"><a href="#cb44-650"></a>            w_dist <span class="op">=</span> np.<span class="bu">sum</span>((np.array([x, y]) <span class="op">-</span> bmu_idx) <span class="op">**</span> <span class="dv">2</span>)</span>
<span id="cb44-651"><a href="#cb44-651"></a>            <span class="co"># if the distance is within the current neighbourhood radius</span></span>
<span id="cb44-652"><a href="#cb44-652"></a>            <span class="cf">if</span> w_dist <span class="op">&lt;=</span> r<span class="op">**</span><span class="dv">2</span>:</span>
<span id="cb44-653"><a href="#cb44-653"></a>                <span class="co"># calculate the degree of influence (based on the 2-D distance)</span></span>
<span id="cb44-654"><a href="#cb44-654"></a>                influence <span class="op">=</span> calculate_influence(w_dist, r)</span>
<span id="cb44-655"><a href="#cb44-655"></a>                <span class="co"># now update the neuron's weight using the formula:</span></span>
<span id="cb44-656"><a href="#cb44-656"></a>                <span class="co"># new w = old w + (learning rate * influence * delta)</span></span>
<span id="cb44-657"><a href="#cb44-657"></a>                <span class="co"># where delta = input vector (t) - old w</span></span>
<span id="cb44-658"><a href="#cb44-658"></a>                new_w <span class="op">=</span> w <span class="op">+</span> (l <span class="op">*</span> influence <span class="op">*</span> (t <span class="op">-</span> w))</span>
<span id="cb44-659"><a href="#cb44-659"></a>                <span class="co"># commit the new weight</span></span>
<span id="cb44-660"><a href="#cb44-660"></a>                net[x, y, :] <span class="op">=</span> new_w.reshape(<span class="dv">1</span>, wlen)</span>
<span id="cb44-661"><a href="#cb44-661"></a>    <span class="cf">return</span> net</span>
<span id="cb44-662"><a href="#cb44-662"></a><span class="in">```</span></span>
<span id="cb44-663"><a href="#cb44-663"></a></span>
<span id="cb44-664"><a href="#cb44-664"></a>The idea behind a SOM is that you’re mapping high-dimensional vectors onto a</span>
<span id="cb44-665"><a href="#cb44-665"></a>smaller dimensional (typically 2D) space.</span>
<span id="cb44-666"><a href="#cb44-666"></a>Vectors that are close in the high-dimensional space also end up being mapped</span>
<span id="cb44-667"><a href="#cb44-667"></a>to nodes that are close in 2D space thus preserving the "topology" of the</span>
<span id="cb44-668"><a href="#cb44-668"></a>original data.</span>
<span id="cb44-669"><a href="#cb44-669"></a></span>
<span id="cb44-670"><a href="#cb44-670"></a><span class="fu">### Generating "Color" data and normalizing</span></span>
<span id="cb44-671"><a href="#cb44-671"></a></span>
<span id="cb44-674"><a href="#cb44-674"></a><span class="in">```{python}</span></span>
<span id="cb44-675"><a href="#cb44-675"></a>raw_data <span class="op">=</span> np.random.randint(<span class="dv">0</span>, <span class="dv">255</span>, (<span class="dv">10000</span>, <span class="dv">3</span>))  <span class="co"># JPEG like data</span></span>
<span id="cb44-676"><a href="#cb44-676"></a><span class="co"># raw_data = np.random.uniform(size=(500, 3))</span></span>
<span id="cb44-677"><a href="#cb44-677"></a>data <span class="op">=</span> StandardScaler().fit_transform(raw_data)  <span class="co"># Normalized</span></span>
<span id="cb44-678"><a href="#cb44-678"></a><span class="in">```</span></span>
<span id="cb44-679"><a href="#cb44-679"></a></span>
<span id="cb44-680"><a href="#cb44-680"></a><span class="fu">### Defining SOM </span></span>
<span id="cb44-681"><a href="#cb44-681"></a></span>
<span id="cb44-682"><a href="#cb44-682"></a>Defining network size, number of iterations and learning rate</span>
<span id="cb44-683"><a href="#cb44-683"></a></span>
<span id="cb44-686"><a href="#cb44-686"></a><span class="in">```{python}</span></span>
<span id="cb44-687"><a href="#cb44-687"></a>network_dimensions <span class="op">=</span> np.array([<span class="dv">50</span>, <span class="dv">50</span>])</span>
<span id="cb44-688"><a href="#cb44-688"></a>n_iterations <span class="op">=</span> <span class="dv">2500</span></span>
<span id="cb44-689"><a href="#cb44-689"></a>init_learning_rate <span class="op">=</span> <span class="fl">0.075</span></span>
<span id="cb44-690"><a href="#cb44-690"></a><span class="in">```</span></span>
<span id="cb44-691"><a href="#cb44-691"></a></span>
<span id="cb44-692"><a href="#cb44-692"></a>Establish size variables based on data</span>
<span id="cb44-693"><a href="#cb44-693"></a></span>
<span id="cb44-696"><a href="#cb44-696"></a><span class="in">```{python}</span></span>
<span id="cb44-697"><a href="#cb44-697"></a>m <span class="op">=</span> data.shape[<span class="dv">0</span>]</span>
<span id="cb44-698"><a href="#cb44-698"></a>n <span class="op">=</span> data.shape[<span class="dv">1</span>]</span>
<span id="cb44-699"><a href="#cb44-699"></a><span class="in">```</span></span>
<span id="cb44-700"><a href="#cb44-700"></a></span>
<span id="cb44-701"><a href="#cb44-701"></a>Weight matrix (i.e. the SOM) needs to be one n-dimensional vector for each</span>
<span id="cb44-702"><a href="#cb44-702"></a>neuron in the SOM</span>
<span id="cb44-703"><a href="#cb44-703"></a></span>
<span id="cb44-706"><a href="#cb44-706"></a><span class="in">```{python}</span></span>
<span id="cb44-707"><a href="#cb44-707"></a>net <span class="op">=</span> np.random.random(</span>
<span id="cb44-708"><a href="#cb44-708"></a>    (network_dimensions[<span class="dv">0</span>], network_dimensions[<span class="dv">1</span>], n)</span>
<span id="cb44-709"><a href="#cb44-709"></a>)  <span class="co"># 25 neurons each with a 3D vector</span></span>
<span id="cb44-710"><a href="#cb44-710"></a><span class="in">```</span></span>
<span id="cb44-711"><a href="#cb44-711"></a></span>
<span id="cb44-712"><a href="#cb44-712"></a>Initial neighbourhood radius and decay parameter</span>
<span id="cb44-713"><a href="#cb44-713"></a></span>
<span id="cb44-716"><a href="#cb44-716"></a><span class="in">```{python}</span></span>
<span id="cb44-717"><a href="#cb44-717"></a>init_radius <span class="op">=</span> <span class="bu">max</span>(network_dimensions[<span class="dv">0</span>], network_dimensions[<span class="dv">1</span>]) <span class="op">/</span> <span class="fl">2.0</span></span>
<span id="cb44-718"><a href="#cb44-718"></a>time_constant <span class="op">=</span> n_iterations <span class="op">/</span> np.log(init_radius)</span>
<span id="cb44-719"><a href="#cb44-719"></a><span class="in">```</span></span>
<span id="cb44-720"><a href="#cb44-720"></a></span>
<span id="cb44-721"><a href="#cb44-721"></a><span class="fu">### Initial state of SOM color network</span></span>
<span id="cb44-722"><a href="#cb44-722"></a></span>
<span id="cb44-725"><a href="#cb44-725"></a><span class="in">```{python}</span></span>
<span id="cb44-726"><a href="#cb44-726"></a>plt.rcParams[<span class="st">"image.cmap"</span>] <span class="op">=</span> <span class="st">"rainbow"</span></span>
<span id="cb44-727"><a href="#cb44-727"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">4</span>, <span class="dv">4</span>))</span>
<span id="cb44-728"><a href="#cb44-728"></a>image <span class="op">=</span> ax.matshow(net)</span>
<span id="cb44-729"><a href="#cb44-729"></a><span class="co"># _ = fig.colorbar(image)</span></span>
<span id="cb44-730"><a href="#cb44-730"></a>_ <span class="op">=</span> plt.show()</span>
<span id="cb44-731"><a href="#cb44-731"></a><span class="in">```</span></span>
<span id="cb44-732"><a href="#cb44-732"></a></span>
<span id="cb44-733"><a href="#cb44-733"></a><span class="fu">### Training SOM</span></span>
<span id="cb44-734"><a href="#cb44-734"></a></span>
<span id="cb44-737"><a href="#cb44-737"></a><span class="in">```{python}</span></span>
<span id="cb44-738"><a href="#cb44-738"></a><span class="kw">def</span> normalize(x: np.ndarray) <span class="op">-&gt;</span> np.ndarray:</span>
<span id="cb44-739"><a href="#cb44-739"></a>    <span class="cf">return</span> (x <span class="op">-</span> np.<span class="bu">min</span>(x)) <span class="op">/</span> (np.<span class="bu">max</span>(x) <span class="op">-</span> np.<span class="bu">min</span>(x))</span>
<span id="cb44-740"><a href="#cb44-740"></a></span>
<span id="cb44-741"><a href="#cb44-741"></a></span>
<span id="cb44-742"><a href="#cb44-742"></a>plt.rcParams[<span class="st">"figure.figsize"</span>] <span class="op">=</span> [</span>
<span id="cb44-743"><a href="#cb44-743"></a>    i <span class="op">/</span> <span class="fl">2.0</span> <span class="cf">for</span> i <span class="kw">in</span> plt.rcParamsDefault[<span class="st">"figure.figsize"</span>]</span>
<span id="cb44-744"><a href="#cb44-744"></a>]</span>
<span id="cb44-745"><a href="#cb44-745"></a>init_learning_rate <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb44-746"><a href="#cb44-746"></a><span class="co"># net = np.random.uniform(size=net.shape)</span></span>
<span id="cb44-747"><a href="#cb44-747"></a><span class="co"># net = (255 * ((net + 1.) / 2.))</span></span>
<span id="cb44-748"><a href="#cb44-748"></a>images <span class="op">=</span> []</span>
<span id="cb44-749"><a href="#cb44-749"></a><span class="cf">for</span> iteration <span class="kw">in</span> <span class="bu">range</span>(n_iterations):</span>
<span id="cb44-750"><a href="#cb44-750"></a>    <span class="co"># select a training example at random - shape of 1x3</span></span>
<span id="cb44-751"><a href="#cb44-751"></a>    t <span class="op">=</span> data[np.random.randint(<span class="dv">0</span>, m), :].reshape(np.array([n, <span class="dv">1</span>]))</span>
<span id="cb44-752"><a href="#cb44-752"></a>    <span class="co"># find its Best Matching Unit</span></span>
<span id="cb44-753"><a href="#cb44-753"></a>    bmu, bmu_idx <span class="op">=</span> find_bmu(t, net, n)  <span class="co"># Gives the row, column of the best neuron</span></span>
<span id="cb44-754"><a href="#cb44-754"></a>    <span class="co"># decay the SOM parameters</span></span>
<span id="cb44-755"><a href="#cb44-755"></a>    r <span class="op">=</span> decay_radius(init_radius, iteration, time_constant)</span>
<span id="cb44-756"><a href="#cb44-756"></a>    l <span class="op">=</span> decay_learning_rate(init_learning_rate, iteration, n_iterations)</span>
<span id="cb44-757"><a href="#cb44-757"></a>    <span class="co"># Update SOM weights</span></span>
<span id="cb44-758"><a href="#cb44-758"></a>    net <span class="op">=</span> update_weights(net, bmu_idx, r, l)</span>
<span id="cb44-759"><a href="#cb44-759"></a>    <span class="cf">if</span> iteration <span class="op">%</span> <span class="dv">50</span> <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb44-760"><a href="#cb44-760"></a>        <span class="bu">print</span>(<span class="ss">f"Iteration: </span><span class="sc">{</span>iteration<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb44-761"><a href="#cb44-761"></a>        ipydis.clear_output(<span class="st">"wait"</span>)</span>
<span id="cb44-762"><a href="#cb44-762"></a>        fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">4</span>, <span class="dv">4</span>))</span>
<span id="cb44-763"><a href="#cb44-763"></a>        net_img <span class="op">=</span> normalize(net)</span>
<span id="cb44-764"><a href="#cb44-764"></a>        _ <span class="op">=</span> fig.suptitle(<span class="ss">f"Iteration: </span><span class="sc">{</span>iteration<span class="sc">}</span><span class="ss">"</span>, y<span class="op">=</span><span class="fl">1.0</span>)</span>
<span id="cb44-765"><a href="#cb44-765"></a>        plt.tight_layout()</span>
<span id="cb44-766"><a href="#cb44-766"></a>        image <span class="op">=</span> ax.matshow(net_img, cmap<span class="op">=</span><span class="st">"rainbow"</span>)</span>
<span id="cb44-767"><a href="#cb44-767"></a>        <span class="cf">if</span> iteration <span class="op">%</span> <span class="dv">100</span> <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb44-768"><a href="#cb44-768"></a>            images.append(net_img)</span>
<span id="cb44-769"><a href="#cb44-769"></a>        _ <span class="op">=</span> plt.show()</span>
<span id="cb44-770"><a href="#cb44-770"></a>        time.sleep(<span class="fl">0.5</span>)</span>
<span id="cb44-771"><a href="#cb44-771"></a><span class="in">```</span></span>
<span id="cb44-772"><a href="#cb44-772"></a></span>
<span id="cb44-773"><a href="#cb44-773"></a><span class="fu">### Visualization of trained colormap SOM</span></span>
<span id="cb44-774"><a href="#cb44-774"></a></span>
<span id="cb44-777"><a href="#cb44-777"></a><span class="in">```{python}</span></span>
<span id="cb44-778"><a href="#cb44-778"></a><span class="bu">len</span>(images)</span>
<span id="cb44-779"><a href="#cb44-779"></a><span class="in">```</span></span>
<span id="cb44-780"><a href="#cb44-780"></a></span>
<span id="cb44-783"><a href="#cb44-783"></a><span class="in">```{python}</span></span>
<span id="cb44-784"><a href="#cb44-784"></a>fig, axes <span class="op">=</span> plt.subplots(</span>
<span id="cb44-785"><a href="#cb44-785"></a>    figsize<span class="op">=</span>(<span class="fl">2.0</span> <span class="op">*</span> DFIGSIZE[<span class="dv">0</span>], <span class="fl">1.5</span> <span class="op">*</span> DFIGSIZE[<span class="dv">1</span>]),</span>
<span id="cb44-786"><a href="#cb44-786"></a>    nrows<span class="op">=</span><span class="dv">5</span>,</span>
<span id="cb44-787"><a href="#cb44-787"></a>    ncols<span class="op">=</span><span class="dv">5</span>,</span>
<span id="cb44-788"><a href="#cb44-788"></a>    tight_layout<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb44-789"><a href="#cb44-789"></a>)</span>
<span id="cb44-790"><a href="#cb44-790"></a>fig.subplots_adjust(wspace<span class="op">=</span><span class="fl">0.1</span>)</span>
<span id="cb44-791"><a href="#cb44-791"></a>axes <span class="op">=</span> axes.flatten()</span>
<span id="cb44-792"><a href="#cb44-792"></a><span class="cf">for</span> idx, (img, ax) <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">zip</span>(images, axes)):</span>
<span id="cb44-793"><a href="#cb44-793"></a>    _ <span class="op">=</span> ax.matshow(img, cmap<span class="op">=</span><span class="st">"rainbow"</span>)</span>
<span id="cb44-794"><a href="#cb44-794"></a>    _ <span class="op">=</span> ax.set_title(<span class="ss">f"Iteration: </span><span class="sc">{</span><span class="dv">100</span> <span class="op">*</span> idx<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb44-795"><a href="#cb44-795"></a>    _ <span class="op">=</span> ax.set_xticks([])</span>
<span id="cb44-796"><a href="#cb44-796"></a>    _ <span class="op">=</span> ax.set_xticklabels([])</span>
<span id="cb44-797"><a href="#cb44-797"></a>    _ <span class="op">=</span> ax.set_yticks([])</span>
<span id="cb44-798"><a href="#cb44-798"></a>    _ <span class="op">=</span> ax.set_yticklabels([])</span>
<span id="cb44-799"><a href="#cb44-799"></a></span>
<span id="cb44-800"><a href="#cb44-800"></a>_ <span class="op">=</span> plt.show()</span>
<span id="cb44-801"><a href="#cb44-801"></a><span class="in">```</span></span>
<span id="cb44-802"><a href="#cb44-802"></a></span>
<span id="cb44-805"><a href="#cb44-805"></a><span class="in">```{python}</span></span>
<span id="cb44-806"><a href="#cb44-806"></a><span class="im">from</span> matplotlib <span class="im">import</span> cm</span>
<span id="cb44-807"><a href="#cb44-807"></a></span>
<span id="cb44-808"><a href="#cb44-808"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">4</span>, <span class="dv">4</span>))</span>
<span id="cb44-809"><a href="#cb44-809"></a>image <span class="op">=</span> ax.imshow(normalize(net), cmap<span class="op">=</span><span class="st">"rainbow"</span>)</span>
<span id="cb44-810"><a href="#cb44-810"></a>_ <span class="op">=</span> ax.set_title(<span class="st">"Final SOM"</span>)</span>
<span id="cb44-811"><a href="#cb44-811"></a>_ <span class="op">=</span> plt.show()</span>
<span id="cb44-812"><a href="#cb44-812"></a><span class="in">```</span></span>
<span id="cb44-813"><a href="#cb44-813"></a></span>
<span id="cb44-814"><a href="#cb44-814"></a></span>
<span id="cb44-815"><a href="#cb44-815"></a><span class="fu">### SOM on Cancer data</span></span>
<span id="cb44-816"><a href="#cb44-816"></a></span>
<span id="cb44-819"><a href="#cb44-819"></a><span class="in">```{python}</span></span>
<span id="cb44-820"><a href="#cb44-820"></a><span class="co"># Load data</span></span>
<span id="cb44-821"><a href="#cb44-821"></a>data <span class="op">=</span> datasets.load_breast_cancer()</span>
<span id="cb44-822"><a href="#cb44-822"></a><span class="co"># Get features and target</span></span>
<span id="cb44-823"><a href="#cb44-823"></a>X <span class="op">=</span> data[<span class="st">"data"</span>]  <span class="co"># pd.DataFrame(data['data'], columns=data['feature_names'])</span></span>
<span id="cb44-824"><a href="#cb44-824"></a>Y <span class="op">=</span> data[<span class="st">"target"</span>]</span>
<span id="cb44-825"><a href="#cb44-825"></a><span class="bu">print</span>(X.shape, Y.shape)</span>
<span id="cb44-826"><a href="#cb44-826"></a><span class="in">```</span></span>
<span id="cb44-827"><a href="#cb44-827"></a></span>
<span id="cb44-830"><a href="#cb44-830"></a><span class="in">```{python}</span></span>
<span id="cb44-831"><a href="#cb44-831"></a>network_dimensions <span class="op">=</span> np.array([<span class="dv">10</span>, <span class="dv">10</span>])</span>
<span id="cb44-832"><a href="#cb44-832"></a>n_iterations <span class="op">=</span> <span class="dv">2000</span></span>
<span id="cb44-833"><a href="#cb44-833"></a>init_learning_rate <span class="op">=</span> <span class="fl">0.01</span></span>
<span id="cb44-834"><a href="#cb44-834"></a><span class="co"># establish size variables based on data</span></span>
<span id="cb44-835"><a href="#cb44-835"></a>n_points <span class="op">=</span> X.shape[<span class="dv">0</span>]  <span class="co"># number of points</span></span>
<span id="cb44-836"><a href="#cb44-836"></a>n_features <span class="op">=</span> X.shape[<span class="dv">1</span>]  <span class="co"># 30 features per point</span></span>
<span id="cb44-837"><a href="#cb44-837"></a></span>
<span id="cb44-838"><a href="#cb44-838"></a><span class="co"># weight matrix (i.e. the SOM) needs to be one n-dimensional vector for each neuron in the SOM</span></span>
<span id="cb44-839"><a href="#cb44-839"></a>net <span class="op">=</span> np.random.random((network_dimensions[<span class="dv">0</span>], network_dimensions[<span class="dv">1</span>], n_features))</span>
<span id="cb44-840"><a href="#cb44-840"></a></span>
<span id="cb44-841"><a href="#cb44-841"></a><span class="co"># initial neighbourhood radius</span></span>
<span id="cb44-842"><a href="#cb44-842"></a>init_radius <span class="op">=</span> <span class="bu">max</span>(network_dimensions[<span class="dv">0</span>], network_dimensions[<span class="dv">1</span>]) <span class="op">/</span> <span class="dv">2</span></span>
<span id="cb44-843"><a href="#cb44-843"></a><span class="co"># radius decay parameter</span></span>
<span id="cb44-844"><a href="#cb44-844"></a>time_constant <span class="op">=</span> n_iterations <span class="op">/</span> np.log(init_radius)</span>
<span id="cb44-845"><a href="#cb44-845"></a><span class="in">```</span></span>
<span id="cb44-846"><a href="#cb44-846"></a></span>
<span id="cb44-849"><a href="#cb44-849"></a><span class="in">```{python}</span></span>
<span id="cb44-850"><a href="#cb44-850"></a><span class="co"># convert the network to something we can visualize</span></span>
<span id="cb44-851"><a href="#cb44-851"></a>net_vis <span class="op">=</span> np.zeros(</span>
<span id="cb44-852"><a href="#cb44-852"></a>    shape<span class="op">=</span>(net.shape[<span class="dv">0</span>], net.shape[<span class="dv">1</span>], <span class="dv">3</span>),</span>
<span id="cb44-853"><a href="#cb44-853"></a>    dtype<span class="op">=</span>np.float32,</span>
<span id="cb44-854"><a href="#cb44-854"></a>)  <span class="co"># Array for SOM color map visualization</span></span>
<span id="cb44-855"><a href="#cb44-855"></a></span>
<span id="cb44-856"><a href="#cb44-856"></a><span class="cf">for</span> sample <span class="kw">in</span> <span class="bu">range</span>(n_points):</span>
<span id="cb44-857"><a href="#cb44-857"></a>    t <span class="op">=</span> X[sample, :].reshape(np.array([n_features, <span class="dv">1</span>]))</span>
<span id="cb44-858"><a href="#cb44-858"></a>    <span class="co"># find its Best Matching Unit for this data point</span></span>
<span id="cb44-859"><a href="#cb44-859"></a>    bmu, bmu_idx <span class="op">=</span> find_bmu(t, net, n_features)</span>
<span id="cb44-860"><a href="#cb44-860"></a>    <span class="co"># set that unit to the label of this data point</span></span>
<span id="cb44-861"><a href="#cb44-861"></a>    net_vis[bmu_idx[<span class="dv">0</span>], bmu_idx[<span class="dv">1</span>], <span class="dv">0</span>] <span class="op">=</span> Y[sample]  <span class="co"># Red if benign</span></span>
<span id="cb44-862"><a href="#cb44-862"></a></span>
<span id="cb44-863"><a href="#cb44-863"></a></span>
<span id="cb44-864"><a href="#cb44-864"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb44-865"><a href="#cb44-865"></a>im <span class="op">=</span> ax.matshow(normalize(net_vis))</span>
<span id="cb44-866"><a href="#cb44-866"></a><span class="co"># _ = plt.colorbar(im1, ax=ax1)</span></span>
<span id="cb44-867"><a href="#cb44-867"></a>plt.show()</span>
<span id="cb44-868"><a href="#cb44-868"></a><span class="in">```</span></span>
<span id="cb44-869"><a href="#cb44-869"></a></span>
<span id="cb44-870"><a href="#cb44-870"></a><span class="fu">### Training SOM on Cancer data</span></span>
<span id="cb44-871"><a href="#cb44-871"></a></span>
<span id="cb44-874"><a href="#cb44-874"></a><span class="in">```{python}</span></span>
<span id="cb44-875"><a href="#cb44-875"></a><span class="cf">for</span> iteration <span class="kw">in</span> <span class="bu">range</span>(n_iterations):</span>
<span id="cb44-876"><a href="#cb44-876"></a>    <span class="co"># select a training example at random - shape of 1x3</span></span>
<span id="cb44-877"><a href="#cb44-877"></a>    t <span class="op">=</span> X[np.random.randint(<span class="dv">0</span>, n_points), :].reshape(np.array([n_features, <span class="dv">1</span>]))</span>
<span id="cb44-878"><a href="#cb44-878"></a>    <span class="co"># find its Best Matching Unit</span></span>
<span id="cb44-879"><a href="#cb44-879"></a>    bmu, bmu_idx <span class="op">=</span> find_bmu(t, net, n_features)</span>
<span id="cb44-880"><a href="#cb44-880"></a>    <span class="co"># decay the SOM parameters</span></span>
<span id="cb44-881"><a href="#cb44-881"></a>    r <span class="op">=</span> decay_radius(init_radius, iteration, time_constant)</span>
<span id="cb44-882"><a href="#cb44-882"></a>    l <span class="op">=</span> decay_learning_rate(init_learning_rate, iteration, n_iterations)</span>
<span id="cb44-883"><a href="#cb44-883"></a>    <span class="co"># Update SOM weights</span></span>
<span id="cb44-884"><a href="#cb44-884"></a>    net <span class="op">=</span> update_weights(net, bmu_idx, r, l)</span>
<span id="cb44-885"><a href="#cb44-885"></a></span>
<span id="cb44-886"><a href="#cb44-886"></a>    <span class="cf">if</span> iteration <span class="op">%</span> <span class="dv">50</span> <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb44-887"><a href="#cb44-887"></a>        ipydis.clear_output(<span class="st">"wait"</span>)</span>
<span id="cb44-888"><a href="#cb44-888"></a>        net_vis <span class="op">=</span> np.zeros(</span>
<span id="cb44-889"><a href="#cb44-889"></a>            shape<span class="op">=</span>(np.shape(net)[<span class="dv">0</span>], np.shape(net)[<span class="dv">1</span>], <span class="dv">3</span>), dtype<span class="op">=</span><span class="st">"double"</span></span>
<span id="cb44-890"><a href="#cb44-890"></a>        )  <span class="co"># Array for SOM color map visualization</span></span>
<span id="cb44-891"><a href="#cb44-891"></a>        <span class="cf">for</span> sample <span class="kw">in</span> <span class="bu">range</span>(n_points):</span>
<span id="cb44-892"><a href="#cb44-892"></a>            t <span class="op">=</span> X[sample, :].reshape(np.array([n_features, <span class="dv">1</span>]))</span>
<span id="cb44-893"><a href="#cb44-893"></a>            <span class="co"># find its Best Matching Unit</span></span>
<span id="cb44-894"><a href="#cb44-894"></a>            bmu, bmu_idx <span class="op">=</span> find_bmu(t, net, n_features)</span>
<span id="cb44-895"><a href="#cb44-895"></a>            net_vis[bmu_idx[<span class="dv">0</span>], bmu_idx[<span class="dv">1</span>], <span class="dv">0</span>] <span class="op">=</span> Y[sample]  <span class="co"># Red if benign</span></span>
<span id="cb44-896"><a href="#cb44-896"></a></span>
<span id="cb44-897"><a href="#cb44-897"></a>        fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">3</span>, <span class="dv">3</span>))</span>
<span id="cb44-898"><a href="#cb44-898"></a>        _ <span class="op">=</span> ax.set_title(<span class="ss">f"Iteration: </span><span class="sc">{</span>iteration<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb44-899"><a href="#cb44-899"></a>        _ <span class="op">=</span> ax.imshow(normalize(net_vis), cmap<span class="op">=</span><span class="st">"rainbow"</span>)</span>
<span id="cb44-900"><a href="#cb44-900"></a>        _ <span class="op">=</span> plt.show()</span>
<span id="cb44-901"><a href="#cb44-901"></a>        time.sleep(<span class="fl">0.25</span>)</span>
<span id="cb44-902"><a href="#cb44-902"></a><span class="in">```</span></span>
<span id="cb44-903"><a href="#cb44-903"></a></span>
<span id="cb44-904"><a href="#cb44-904"></a><span class="fu">### Visualization of trained SOM</span></span>
<span id="cb44-905"><a href="#cb44-905"></a></span>
<span id="cb44-906"><a href="#cb44-906"></a></span>
<span id="cb44-909"><a href="#cb44-909"></a><span class="in">```{python}</span></span>
<span id="cb44-910"><a href="#cb44-910"></a>net_vis <span class="op">=</span> np.zeros(</span>
<span id="cb44-911"><a href="#cb44-911"></a>    shape<span class="op">=</span>(net.shape[<span class="dv">0</span>], net.shape[<span class="dv">1</span>], <span class="dv">3</span>), dtype<span class="op">=</span><span class="st">"double"</span></span>
<span id="cb44-912"><a href="#cb44-912"></a>)  <span class="co"># Array for SOM color map visualization</span></span>
<span id="cb44-913"><a href="#cb44-913"></a></span>
<span id="cb44-914"><a href="#cb44-914"></a><span class="cf">for</span> sample <span class="kw">in</span> <span class="bu">range</span>(n_points):</span>
<span id="cb44-915"><a href="#cb44-915"></a>    t <span class="op">=</span> X[sample, :].reshape(np.array([n_features, <span class="dv">1</span>]))</span>
<span id="cb44-916"><a href="#cb44-916"></a>    <span class="co"># find its Best Matching Unit</span></span>
<span id="cb44-917"><a href="#cb44-917"></a>    bmu, bmu_idx <span class="op">=</span> find_bmu(t, net, n_features)</span>
<span id="cb44-918"><a href="#cb44-918"></a>    net_vis[bmu_idx[<span class="dv">0</span>], bmu_idx[<span class="dv">1</span>], <span class="dv">0</span>] <span class="op">=</span> Y[sample]  <span class="co"># Red if benign</span></span>
<span id="cb44-919"><a href="#cb44-919"></a></span>
<span id="cb44-920"><a href="#cb44-920"></a>_ <span class="op">=</span> plt.imshow(normalize(net_vis))</span>
<span id="cb44-921"><a href="#cb44-921"></a>_ <span class="op">=</span> plt.show()</span>
<span id="cb44-922"><a href="#cb44-922"></a><span class="in">```</span></span>
<span id="cb44-923"><a href="#cb44-923"></a></span>
<span id="cb44-924"><a href="#cb44-924"></a><span class="fu">## Keep learning</span></span>
<span id="cb44-925"><a href="#cb44-925"></a></span>
<span id="cb44-926"><a href="#cb44-926"></a><span class="ss">1. </span>&lt;lagunita.stanford.edu/courses/HumanitiesSciences/StatLearning/Winter2016/course/&gt;</span>
<span id="cb44-927"><a href="#cb44-927"></a><span class="ss">2. </span>&lt;www.coursera.org/learn/ml-clustering-and-retrieval/&gt;</span>
<span id="cb44-928"><a href="#cb44-928"></a><span class="ss">3. </span>&lt;www.coursera.org/learn/machine-learning/home/week/8&gt;</span>
</code></pre></div><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="https://saforem2.github.io/intro-hpc-bootcamp-2025/">
<p>saforem2.github.io/intro-hpc-bootcamp-2025</p>
</a>
  </li>  
</ul>
    </div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/saforem2/intro-hpc-bootcamp-2025/blob/main/00-intro-AI-HPC/7-statistical-learning/index.qmd" class="toc-action"><i class="bi bi-github"></i>View source</a></li><li><a href="https://github.com/saforem2/intro-hpc-bootcamp-2025/edit/main/00-intro-AI-HPC/7-statistical-learning/index.qmd" class="toc-action"><i class="bi empty"></i>Edit this page</a></li><li><a href="https://github.com/saforem2/intro-hpc-bootcamp-2025/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/saforem2">
      <i class="bi bi-twitter" role="img" aria-label="Sam Foreman Twitter">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link active" href="https://github.com/saforem2/intro-hpc-bootcamp-2025" aria-current="page">
      <i class="bi bi-github" role="img" aria-label="Sam Foreman GitHub">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>